---
title: "Build SCE Parse ovarian carcinosarcoma Downsampled"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/build_sce/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
library(Matrix)
library(DropletUtils)
library(tidyverse)
library(patchwork)
library(scuttle)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

When I extracted nuclei and fixed with the 10x Genomics kit the sample became increasingly clumped and ended up blocking the microfluidics chip. Fix with the Parse Bio version 2 kit. 

Additionally try a 10x Genomics protocol where I fix tumor pieces first then extract nuclei.

## Samples
Human ovarian carcinosarcoma PDX tissue pieces. Has been kept frozen as rice sized pieces.

I received 4 samples from Clare Scottâ€™s lab:

| Sample | Count preFix | Count postfix |
|--------|--------------|---------------|
| PDX117 | 780,000      | 308,250       |
| PDX233 | 1.54M        | 1.32M         |
| PDX105 | 373,000      | 237,750       |
| PDX264 | 782,000      | 726,750       |

I passed PDX1 = PDX117 and PDX3 = PDX105 to Zac Moore for processing

Link to the [ELN entry](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/NDE0Ljd8NjgxOTcvMzE5L1RyZWVOb2RlLzE2NzI5MzY5MzV8MTA1Mi43)

## Downsampling

I downsampled the reads to allow a fairer comparision with 10x Genomics v3.1 frsh nuclei.  
The script I used for downsampling is 

10x Genomics fresh nuclei capture

* 16181 cells
* 207,384,474 reads
* For a fair comparison to Parse downsample to 207,384,474 / 16,181 = 12,817 reads per cell

Parse Bio v2 mini

* Parse 8,948 cells total all 12 wells
* So at 12,817 reads per cell need 12,817 * 8,948 = 114,686,516 reads

This corresponded to 114,686,516 reads for Parse v2 mini vs 207,384,474 reads for 10x Genomics

## Sequencing metrics

| statistic | nuclei_pdx_1 | nuclei_pdx_2 | all-well |
|---|---|---|---|
| sequencing_saturation | 0.0554 | 0.0546 | 0.0563 |
| sample_well_count | 2 | 2 | 12 |
| number_of_cells | 1937 | 1746 | 9054 |
| hg38_number_of_cells | 1925 | 1696 | 8987 |
| mm10_number_of_cells | 0 | 1 | 1 |
| multiplet_number_of_cells | 12 | 49 | 66 |
| hg38_median_tscp_per_cell | 4033 | 3578 | 3059 |
| mm10_median_tscp_per_cell |  | 891 | 891 |
| hg38_median_tscp_at50 | 36382.5633 | 32739.3559 | 27184.6201 |
| hg38_median_genes_per_cell | 2468 | 2049 | 1909 |
| mean_reads_per_cell | 17640.3856 | 15534.7205 | 12666.9446 |
| number_of_reads | 34169426.9 | 27123621.96 | 114686516 |
| number_of_tscp | 18523470 | 14703893 | 62172311 |
| hg38_number_of_tscp | 18489812 | 14582184 | 61975798 |
| mm10_number_of_tscp | 33658 | 121709 | 196513 |
| valid_barcode_fraction | 0.7886 | 0.7886 | 0.7886 |
| transcriptome_map_fraction | 0.7284 | 0.7284 | 0.7284 |
| tso_fraction_in_read1 | 0.2877 | 0.2877 | 0.2877 |
| hg38_fraction_reads_in_cells | 0.8432 | 0.8407 | 0.7973 |
| mm10_fraction_reads_in_cells | 0 | 0.0073 | 0.0045 |
| hg38_fraction_tscp_in_cells | 0.8426 | 0.84 | 0.7967 |
| mm10_fraction_tscp_in_cells | 0 | 0.0073 | 0.0045 |
| hg38_fraction_exonic | 0.2339 | 0.2158 | 0.2942 |
| mm10_fraction_exonic | 0.305 | 0.2658 | 0.3108 |
| cell_tscp_cutoff | 1091 | 857 | 855 |
| cell_tscp_f01_slope | 6.0836 | 5.4728 | 2.9537 |

# Load Data

I follow the instructions on [Parse Bio support site](https://support.parsebiosciences.com/hc/en-us/articles/360053078092-Seurat-Tutorial-65k-PBMCs)

## PDX117

```{r read_pdx117}
sample <- "nuclei_pdx_1"

DGE_folder <- paste("/vast/scratch/users/brown.d/Parse_PDX_Zac/output_downsample/", sample, "/DGE_filtered/", sep = "")
mat <- readMM(paste0(DGE_folder, "DGE.mtx"))

cell_meta <- read.delim(paste0(DGE_folder, "cell_metadata.csv"),
                  stringsAsFactor = FALSE, sep = ",")
genes <- read.delim(paste0(DGE_folder, "all_genes.csv"),
                    stringsAsFactor = FALSE, sep = ",")

cell_meta$bc_wells <- make.unique(cell_meta$bc_wells, sep = "_dup")
rownames(cell_meta) <- cell_meta$bc_wells
genes$gene_name <- make.unique(genes$gene_name, sep = "_dup")

# Setting column and rownames to expression matrix
colnames(mat) <- genes$gene_name
rownames(mat) <- rownames(cell_meta)
mat_t <- t(mat)

# Remove empty rownames, if they exist
mat_t <- mat_t[(rownames(mat_t) != ""),]

pdx117 <- SingleCellExperiment(assay=list(counts=mat_t))
colData(pdx117) <- DataFrame(cell_meta)
rowData(pdx117) <- genes
pdx117$sample <- "PDX117"
```

## PDX105

```{r read_pdx105}
sample <- "nuclei_pdx_2"

DGE_folder <- paste("/vast/scratch/users/brown.d/Parse_PDX_Zac/output_downsample/", sample, "/DGE_filtered/", sep = "")
mat <- readMM(paste0(DGE_folder, "DGE.mtx"))

cell_meta <- read.delim(paste0(DGE_folder, "cell_metadata.csv"),
                  stringsAsFactor = FALSE, sep = ",")
genes <- read.delim(paste0(DGE_folder, "all_genes.csv"),
                    stringsAsFactor = FALSE, sep = ",")

cell_meta$bc_wells <- make.unique(cell_meta$bc_wells, sep = "_dup")
rownames(cell_meta) <- cell_meta$bc_wells
genes$gene_name <- make.unique(genes$gene_name, sep = "_dup")

# Setting column and rownames to expression matrix
colnames(mat) <- genes$gene_name
rownames(mat) <- rownames(cell_meta)
mat_t <- t(mat)

# Remove empty rownames, if they exist
mat_t <- mat_t[(rownames(mat_t) != ""),]

pdx105 <- SingleCellExperiment(assay=list(counts=mat_t))
colData(pdx105) <- DataFrame(cell_meta)
rowData(pdx105) <- genes
pdx105$sample <- "PDX105"
```

Combine into a single SCE object, split mouse counts into alt experiment, remove non expressed genes

```{r}
sce <- cbind(pdx105, pdx117)
unique(sce$species)

is.mouse <- grepl("^mm10", rowData(sce)$genome)
sce <- splitAltExps(sce, ifelse(is.mouse, "Mouse", "gene"))

ave_counts <- calculateAverage(sce)
to_keep <- ave_counts > 0
sce <- sce[to_keep, ]
```

```{r}

saveRDS(sce, here::here(
  "/stornext/Projects/score/Analyses/R010_multiplexing/SCEs/PDX", "downsample_pdx105_117.PARSE_basic-sce.rds"
))
```

## Session Info

```{r}
sessionInfo()
```
