---
title: "Quality control plots Parse vs 10x Genomics"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/Parse_tenX_compareQC/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
suppressPackageStartupMessages({
  library(tidyverse)
  library(patchwork)
  library(scater)
  library(scuttle)
  })

source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication(base_size = 18))
```

# Aim

When I extracted nuclei and fixed with the 10x Genomics kit the sample became increasingly clumped and ended up blocking the microfluidics chip. Fix with the Parse Bio version 2 kit. 

Additionally try a 10x Genomics protocol where I fix tumor pieces first then extract nuclei.

## Samples
Human ovarian carcinosarcoma PDX tissue pieces. Has been kept frozen as rice sized pieces.

I received 4 samples from Clare Scottâ€™s lab:

| Sample | Count preFix | Count postfix |
|--------|--------------|---------------|
| PDX117 | 780,000      | 308,250       |
| PDX105 | 373,000      | 237,750       |

I passed PDX1 = PDX117 and PDX3 = PDX105 to Zac Moore for processing

Link to the [ELN entry](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/NDE0Ljd8NjgxOTcvMzE5L1RyZWVOb2RlLzE2NzI5MzY5MzV8MTA1Mi43)

## Sequencing metrics

| statistic | nuclei_pdx_1 | nuclei_pdx_2 | all-well |
|---|---|---|---|
| sequencing_saturation | 0.0554 | 0.0546 | 0.0563 |
| sample_well_count | 2 | 2 | 12 |
| number_of_cells | 1937 | 1746 | 9054 |
| hg38_number_of_cells | 1925 | 1696 | 8987 |
| mm10_number_of_cells | 0 | 1 | 1 |
| multiplet_number_of_cells | 12 | 49 | 66 |
| hg38_median_tscp_per_cell | 4033 | 3578 | 3059 |
| mm10_median_tscp_per_cell |  | 891 | 891 |
| hg38_median_tscp_at50 | 36382.5633 | 32739.3559 | 27184.6201 |
| hg38_median_genes_per_cell | 2468 | 2049 | 1909 |
| mean_reads_per_cell | 17640.3856 | 15534.7205 | 12666.9446 |
| number_of_reads | 34169426.9 | 27123621.96 | 114686516 |
| number_of_tscp | 18523470 | 14703893 | 62172311 |
| hg38_number_of_tscp | 18489812 | 14582184 | 61975798 |
| mm10_number_of_tscp | 33658 | 121709 | 196513 |
| valid_barcode_fraction | 0.7886 | 0.7886 | 0.7886 |
| transcriptome_map_fraction | 0.7284 | 0.7284 | 0.7284 |
| tso_fraction_in_read1 | 0.2877 | 0.2877 | 0.2877 |
| hg38_fraction_reads_in_cells | 0.8432 | 0.8407 | 0.7973 |
| mm10_fraction_reads_in_cells | 0 | 0.0073 | 0.0045 |
| hg38_fraction_tscp_in_cells | 0.8426 | 0.84 | 0.7967 |
| mm10_fraction_tscp_in_cells | 0 | 0.0073 | 0.0045 |
| hg38_fraction_exonic | 0.2339 | 0.2158 | 0.2942 |
| mm10_fraction_exonic | 0.305 | 0.2658 | 0.3108 |
| cell_tscp_cutoff | 1091 | 857 | 855 |
| cell_tscp_f01_slope | 6.0836 | 5.4728 | 2.9537 |

# Load Data

## Parse Bio

Generated in the 1A_build_SCE notebook.

The Parse samples were cross-referenced with 10x Genomics unlabelled capture in **3A_recode-by_snp_donor.Rmd**

Recall the SNP donors begin with donor0

* Donor 2 is PDX105
* Donor 1 is PDX117

```{r loadData}
sce_p <- readRDS(here::here(
   "/stornext/Projects/score/Analyses/R010_multiplexing/SCEs/PDX", "downsample_pdx105_117.PARSE_basic-sce.rds"
))

sce_p$donor_id <- ""
sce_p$donor_id[sce_p$sample == "PDX117"] <- "donor1"
sce_p$donor_id[sce_p$sample == "PDX105"] <- "donor2"

is.mito <- grepl("^MT", rownames(sce_p))
sce_p <- addPerCellQCMetrics(sce_p, subsets=list(Mito=is.mito))

sce_p
```

## 10x Genomics v3.1

```{r load10x}
sce_t <- readRDS((
  "/stornext/Projects/score/Analyses/R010_multiplexing/SCEs/PDX/S000322.qcmetrics.sces.rds"
))[["Unlabelled"]]

keep_cells <- sce_t$donor_id %in% c("donor2", "donor1")

sce_t <- sce_t[,keep_cells]

```


The number of cells in each Parse capture

```{r}
tb_p <- as_tibble(colData(sce_p))
tb_p$Technology <- "Parse Bio v2"

tb_p %>% 
  dplyr::count(sample)
```

The number of cells in each 10x capture

```{r}
tb_t <- as_tibble(colData(sce_t))
tb_t$Technology <- "10x Genomics v3.1"

tb_t %>% 
  dplyr::count(donor_id)
```

# Mung the colData

```{r}
keep_cols <- intersect(
  colnames(tb_p),
  colnames(tb_t)
)

tb <- rbind(
  tb_p[,keep_cols],
  tb_t[,keep_cols]
)
```


# Quality control metrics {.tabset}

## Library size

```{r libSize_p, fig.cap="Library size in deduplicated UMIs"}
plt1 <- ggplot(tb, 
             aes(x = Technology, y= sum, fill = donor_id)) + 
  geom_boxplot(show.legend = TRUE, outlier.size = 0.5, outlier.alpha = 0.5) +  
  ylab("Library Size (UMIs)") + 
  xlab("") +
  scale_y_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "l") +
  scale_fill_brewer(palette = "Dark2") 

plt1
```

## Detected genes

```{r detected_genes_p, fig.cap="Number of detected genes"}
plt2 <- ggplot(tb, 
             aes(x = Technology, y= detected, fill = donor_id)) + 
  geom_boxplot(show.legend = TRUE, outlier.size = 0.5, outlier.alpha = 0.5) + 
  ylab("Detected genes") + 
  xlab("") +
  scale_y_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "l") +
  scale_fill_brewer(palette = "Dark2")

plt2
```

## Mitochondiral percentage

```{r mito_per_p, fig.cap="Mitochondrial percentage of total counts"}
plt3 <- ggplot(tb, 
             aes(x = Technology, y= subsets_Mito_percent, fill = donor_id)) + 
  geom_boxplot(show.legend = TRUE, outlier.size = 0.2, outlier.alpha = 0.25) + 
  ylab("Mitochondrial %") + 
  xlab("") +
  scale_fill_brewer(palette = "Dark2")

plt3
```

## Genes v UMIs

```{r gene_umi, fig.cap="UMIs versus genes"}
plt4 <- ggplot(tb, 
             aes(x = detected+1, y= sum+1, colour = Technology)) + 
  geom_point(show.legend = TRUE, alpha=0.2) + 
  ylab("Library size (UMIs)") + 
  xlab("Genes detected") +
  scale_y_continuous(trans='log10') + scale_x_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "bl") +
  scale_colour_brewer(palette = "Dark2")

plt4
```

# Summary

```{r summary, fig.height=12, fig.width=16, fig.cap="Summary"}
plt1 + plt2 + plt3 + plt4 + 
  plot_layout(guides = 'collect')
```

# Highly expressed genes

```{r high_expr}
hexpr_p <- plotHighestExprs(sce_p, n = 15, colour_cells_by = "donor_id") + xlim(0,40)

hexpr_t <- plotHighestExprs(sce_t, n = 15, colour_cells_by = "donor_id") + xlim(0,40)

hexpr_p + theme_Publication(base_size = 12) + 
  hexpr_t + theme_Publication(base_size = 12) + plot_layout(ncol=2, guides = "collect")
```

# Session Info

```{r}
sessionInfo()
```
