---
title: "Gene gene correlation Parse vs 10x Genomics"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/gene_gene_correlation/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
suppressPackageStartupMessages({
  library(tidyverse)
  library(patchwork)
  library(scater)
  library(scuttle)
  })

source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication(base_size = 18))
```

# Aim

When I extracted nuclei and fixed with the 10x Genomics kit the sample became increasingly clumped and ended up blocking the microfluidics chip. Fix with the Parse Bio version 2 kit. 

Additionally try a 10x Genomics protocol where I fix tumor pieces first then extract nuclei.

## Samples
Human ovarian carcinosarcoma PDX tissue pieces. Has been kept frozen as rice sized pieces.

I received 4 samples from Clare Scottâ€™s lab:

| Sample | Count preFix | Count postfix |
|--------|--------------|---------------|
| PDX117 | 780,000      | 308,250       |
| PDX105 | 373,000      | 237,750       |

I passed PDX1 = PDX117 and PDX3 = PDX105 to Zac Moore for processing

Link to the [ELN entry](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/NDE0Ljd8NjgxOTcvMzE5L1RyZWVOb2RlLzE2NzI5MzY5MzV8MTA1Mi43)

# Load Data donor 2

## Parse Bio

Generated in the 1A_build_SCE notebook.

The Parse samples were cross-referenced with 10x Genomics unlabelled capture in **3A_recode-by_snp_donor.Rmd**

Recall the SNP donors begin with donor0

* Donor 2 is PDX105
* Donor 1 is PDX117

Use PDX105 for this analysis

```{r loadData}
sce_p <- readRDS(here::here(
   "/stornext/Projects/score/Analyses/R010_multiplexing/SCEs/PDX", "downsample_pdx105_117.PARSE_basic-sce.rds"
))

sce_p$donor_id <- ""
sce_p$donor_id[sce_p$sample == "PDX117"] <- "donor1"
sce_p$donor_id[sce_p$sample == "PDX105"] <- "donor2"

keep_cells <- sce_p$donor_id %in% "donor2"

sce_p
```

## 10x Genomics v3.1

Subset only the same samples that were processed with Parse. 

* Donor 2 is the only one with appreciable MET, ZFPM2-AS1, FSTL5 expression
  + PDX105
* Donor 3 DACH1 and MECOM are the best markers
  + PDX117

```{r load10x}
sce_t <- readRDS((
  "/stornext/Projects/score/Analyses/R010_multiplexing/SCEs/PDX/S000322.qcmetrics.sces.rds"
))[["Unlabelled"]]

keep_cells <- sce_t$donor_id %in% "donor2"

sce_t <- sce_t[,keep_cells]

```

# Mung the count matrices

```{r rowSums}
sce_p <- addPerFeatureQC(sce_p)
rowData(sce_p)$sum <- rowSums2(counts(sce_p))
rowData(sce_p)$Technology <- "Parse Bio v2"

sce_t <- addPerFeatureQC(sce_t)
rowData(sce_t)$sum <- rowSums2(counts(sce_t))
rowData(sce_t)$Technology <- "10x Genomics v3.1"

keep_genes <- intersect(
  rownames(sce_p),
  rownames(sce_t)
)

rd_parse <- as_tibble(rowData(sce_p[keep_genes,]))
rd_tenX <- as_tibble(rowData(sce_t[keep_genes,]))
rd_tenX <- rename(rd_tenX, "Symbol" = "gene_name")
rd_tenX$gene_name <- gsub("GRCh38_", "", rd_tenX$gene_name)

keep_cols <- intersect(
  colnames(rd_parse),
  colnames(rd_tenX)
)
```

Convert to wide tibble

```{r}
gene_tb <- rbind(
  rd_parse[,keep_cols],
  rd_tenX[,keep_cols]
)

# Add 1 before logging
gene_tb$sum <- gene_tb$sum + 1

gene_tb_wide <- gene_tb %>%
  select(gene_name,sum,Technology) %>% 
  pivot_wider(names_from = Technology, values_from = sum)

```

Colour in the mitochondrial genes

```{r}
gene_tb_wide$Gene <- ""
gene_tb_wide$Gene[grep(pattern = "^RP", x=gene_tb_wide$gene_name)] <- "Ribosomal"
gene_tb_wide$Gene[grep(pattern = "^MT-", x=gene_tb_wide$gene_name)] <- "Mitochondrial"
```

# Plot the gene gene correlation

Decent correlation between technologies.  

Compute person correlation

```{r corr_gene}
pc <- cor(x = gene_tb_wide$`Parse Bio v2`, y = gene_tb_wide$`10x Genomics v3.1`, method="pearson")
sub_title <- paste("Pearson correlation coefficient = ",
  as.character(
  round(pc, digits = 2),
  sep=""))
```

The `r sub_title`

If I scramble the the donors i.e plot donor 2 vs donor 3 the correlation is similar which is consistent with there not being much variation in the data, mostly tumour related gene expression

```{r gene_gene_sum}
plt1 <- ggplot(gene_tb_wide, 
             aes(x = `10x Genomics v3.1`, y= `Parse Bio v2`, colour=Gene)) + 
  geom_point(alpha = 0.33, size=1) + 
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1))) +
  xlab("10x Genomics v3.1") + 
  ylab("Parse Bio v2") +
  scale_x_continuous(trans='log10') +
  scale_y_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "bl") +
  scale_color_manual(values = c("blue", "red", "orange"), 
                     labels = c('Other', "Mitochondrial", 'Ribosomal')) +
  theme_Publication(base_size = 20)

plt1
```

# Session Info

```{r}
sessionInfo()
```
