---
title: "QC dimred PDX117"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/PDX117/qc_dimred/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
suppressPackageStartupMessages({
  library(tidyverse)
  library(patchwork)
  library(scuttle)
  library(scater)
  library(scran)
  library(viridis)
  })

source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

Quality control and dimension reduction for PDX117

# Recap

When I extracted nuclei and fixed with the 10x Genomics kit the sample became increasingly clumped and ended up blocking the microfluidics chip. Fix with the Parse Bio version 2 kit. 

Additionally try a 10x Genomics protocol where I fix tumor pieces first then extract nuclei.

## Samples
Human ovarian carcinosarcoma PDX tissue pieces. Has been kept frozen as rice sized pieces.

I received 4 samples from Clare Scott’s lab:

| Sample | Count preFix | Count postfix |
|--------|--------------|---------------|
| PDX117 | 780,000      | 308,250       |
| PDX105 | 373,000      | 237,750       |

I passed PDX1 = PDX117 and PDX3 = PDX105 to Zac Moore for processing

Link to the [ELN entry](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/NDE0Ljd8NjgxOTcvMzE5L1RyZWVOb2RlLzE2NzI5MzY5MzV8MTA1Mi43)

# Read SCE

From the **../3A_recode-by_snp_donor** notebook.  
The number of reads was downsampled to the the same as 10x Genomics v3.1 fresh nuclei.

```{r readSCE}
sce <- readRDS(here::here(
   "data/genomics_data/SCEs", "PDX", "parse_downsample_basic-sce.rds"
))


is.mito <- grepl("^MT", rownames(sce))
sce <- addPerCellQCMetrics(sce, subsets=list(Mito=is.mito))

sce <- sce[,sce$sample == "PDX117"]
```

# Quality control

## Identifying outliers by each metric

Sine half the cells have been removed in subsetting this sample perform the low quality cell removal again.

Include removal of cells with too high library size. There is a group of cells with extremely high library size which are likely cell doublets or clumps. This was a very sticky sample.

```{r}
sce_orig <- sce

mito_drop <- isOutlier(
    metric = sce$subsets_Mito_percent, 
    nmads = 3, type = "higher")
```

Measure how many cells dropped.

```{r outlier_detect}
qc_df <- colData(sce)
reasons <- as.matrix(perCellQCFilters(
    x=qc_df, 
    sub.fields=c("subsets_Mito_percent", "sum")))

discard <- reasons
discard <- discard[,4]

sce <- sce[,!discard]

colnames(reasons)
colSums(reasons)
```

For some reason a few cells with high (50%) mitochondrial genes are retained. Remove this cell manually with a threshold of 10%.

```{r}
sce <- sce[,sce$subsets_Mito_percent <= 10]
```

 
Summarise how many cells left.

256 cells have been dropped. Spread equally across high library size and mitochondrial percentage.

```{r}
cell_drop_tb <- rbind(
  cbind(length(colnames(sce_orig)), length(colnames(sce)))
)

colnames(cell_drop_tb) <- c("Before_Filter", "After_filter")
cell_drop_tb
```

## Filtering out low-abundance genes

Low-abundance genes are problematic as zero or near-zero counts do not contain much information for reliable statistical inference.  
We drop another ~5000 genes combined to the overall Parse object

```{r average_count_p}
ave_counts <- calculateAverage(sce)
to_keep <- ave_counts > 0
sce <- sce[to_keep, ]

hist(
  x = log10(ave_counts), 
  breaks = 100, 
  main = "Parse Evercode mini v2", 
  col = "grey",
  xlab = expression(Log[10] ~ "average count"))
```


```{r filter_gene}
# Summarise how many genes left
gene_drop_tb <- rbind(
  cbind(length(rownames(sce_orig)), length(rownames(sce)))
)

colnames(gene_drop_tb) <- c("Before_Filter", "After_filter")
gene_drop_tb
```

# GEX Normalization

## Using the deconvolution method to deal with zero counts

We pool counts from many cells to increase the count size for accurate size factor estimation (A. T. Lun, Bach, and Marioni 2016). Pool-based size factors are then ‘deconvolved’ into cell-based factors for cell-specific normalization. This removes scaling biases associated with cell-specific differences in capture efficiency, sequencing depth and composition biases.

```{r normalise}
set.seed(666)

clusters <- quickCluster(sce)
sce <- computeSumFactors(sce, clusters = clusters, min.mean = 0.1)
sce <- logNormCounts(sce)
print(summary(sizeFactors(sce)))

```

# Selecting highly variable genes (HVGs)

## Fit variance

Interesting that compared to 10x Genomics there are fewer nusiance genes like ribosomal and mitochondrial

```{r var_genes_p}
var_fit <- modelGeneVarByPoisson(sce)
hvg <- getTopHVGs(var_fit, var.threshold = 0)
```

```{r}
p1 <- plotExpression(object = sce, features = hvg[1:10])
p1
```

# Dimension reduction {.tabset}

## PCA

```{r PCA}
set.seed(666)

sce <- denoisePCA(sce, var_fit, subset.row = hvg)
```

### Number of PCA dimensions to retain

Based on these percent variance plots retain 30 PCA dimensions.  
The variation is relatively low with PC1 being 4% only.

```{r}
# Percentage of variance explained is tucked away in the attributes.
percent.var <- attr(reducedDim(sce), "percentVar")

plot(percent.var, xlab="PC", ylab="Variance explained (%)")
```

### Visualise PCA

The classic V-shaped PCA driven by library size and genes detected

```{r pca_plts}
pca_plt <- plotReducedDim(sce, dimred="PCA", colour_by = "detected")

pca_plt + theme_Publication()
```

## Run UMAP

```{r umap}
set.seed(100)
sce <- runUMAP(sce, dimred = "PCA", n_dimred=30)
```

# Clustering

We build a shared nearest neighbour graph and use the Louvain algorithm to identify clusters. We build the graph using the principal components.

## Initial clustering

Smaller 'k' usually yields finer clusters.

```{r clustering_ks}
set.seed(911)

snn_gr_cmo <- buildSNNGraph(sce, use.dimred = "PCA", k = 5) 
clust.5 <- igraph::cluster_louvain(snn_gr_cmo)
table(clust.5$membership)

snn_gr_cmo <- buildSNNGraph(sce, use.dimred = "PCA", k = 10) 
clust.10 <- igraph::cluster_louvain(snn_gr_cmo)
table(clust.10$membership)

snn_gr_cmo <- buildSNNGraph(sce, use.dimred = "PCA", k = 15) 
clust.15 <- igraph::cluster_louvain(snn_gr_cmo)
table(clust.15$membership)

snn_gr_cmo <- buildSNNGraph(sce, use.dimred = "PCA", k = 20) 
clust.20 <- igraph::cluster_louvain(snn_gr_cmo)
table(clust.20$membership)
```

Based on trying k from 5 - 20 I use k of 15 as all clusters have more than 100 cells. k of 10 does reflect what thee UMAP looks like but cluster 8, the group at the bottom center of the plot has 45 cells.

```{r k_10}
set.seed(911)

snn_gr_cmo <- buildSNNGraph(sce, use.dimred = "PCA", k = 10) 
clusters <- igraph::cluster_louvain(snn_gr_cmo)
sce$cluster <- factor(clusters$membership)
plotReducedDim(sce, dimred="UMAP", colour_by="cluster") + ggtitle("k=10") + theme_Publication()
```

Of course the true test will be when ascribing biological meaning.

```{r clustering_k20}
snn_gr_cmo <- buildSNNGraph(sce, use.dimred = "PCA", k = 20) 
clusters <- igraph::cluster_louvain(snn_gr_cmo)
sce$cluster <- factor(clusters$membership)
```

# Visualise UMAP {.tabset}

## Cluster number only

```{r umap_cluster}
umap_cluster <- plotReducedDim(sce, dimred="UMAP", colour_by="cluster") + theme_Publication()

umap_cluster
```

## With QC parameters

Not a great deal of structure but also no obvious distortions to do with libarry size or mitochondrial gene percentage

```{r umap_plts, fig.dim=c(12,12)}
umap_sum <- plotReducedDim(sce, dimred="UMAP", colour_by="sum") + theme_Publication()
umap_detected <- plotReducedDim(sce, dimred="UMAP", colour_by="detected") + theme_Publication()
umap_cluster <- plotReducedDim(sce, dimred="UMAP", colour_by="cluster") + theme_Publication()
umap_mito <- plotReducedDim(sce, dimred="UMAP", colour_by="subsets_Mito_percent") + theme_Publication()


umap_cluster + umap_sum + umap_detected + umap_mito + plot_layout(ncol=2)
```

# Save SCE object

```{r saveSCEs}
saveRDS(sce,
  "/vast/scratch/users/brown.d/Parse_PDX_Zac/SCEs/downsample_parse_dimred.pdx117.rds"
)
```

## Session Info

```{r}
sessionInfo()
```
