---
title: "Cell calling ovarian carcinosarcoma"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/cell_calling/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
library(DropletUtils)
library(tidyverse)
library(patchwork)
library(scuttle)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim
I previous compared Total-Seq A hashtag, MULTI-Seq and CellPlex lipid on human PBMCs. All protocols worked pretty well but human PBMCs are the easiest sample to work with.

Then I tried mouse embryonic brain E18.5 which didn't work that well. CellPlex diluted 1:10 worked best here.

## This sample
Run a more challenging sample = Ovarian carcinosarcome PDX nuclei.  
Four PDXs were processed with 10x Genomics nuclei isolation kit. A capture was immdiately performed which is the unlabelled.  
Three more aliquots were labelled in singlicate with cell multiplexing reagents and captured later.
## Comparisons

* Fresh unlabelled nuclei after nuclei isolation with 10x Genomics kit
* Fixed nuclei nuclei isolation with 10x Genomics kit
* Fresh nuclei CellPlex 1:10 diluted
* Fresh nuclei custom MULTI-Seq CMO (Nextera mod)
* Fresh nuclei 1ug nucleus hashtag Mab414

# Load Data

Find the paths to the cell ranger output

```{r dataPaths}
UTDDataPath = "/stornext/Projects/score/GenomicsRnD/DB/S000331/cellranger_outs/G000225_PDX_UTD/outs/raw_feature_bc_matrix"

CMODataPath = "/stornext/Projects/score/GenomicsRnD/DB/S000322/cellranger_outs/G000225_PDX_CMO/multi/count/raw_feature_bc_matrix"

NXTDataPath = '/stornext/Projects/score/GenomicsRnD/DB/S000322/cellranger_outs/G000225_PDX_NXT/multi/count/raw_feature_bc_matrix'

HTODataPath = '/stornext/Projects/score/GenomicsRnD/DB/S000322/cellranger_outs/G000225_PDX_HTO/multi/count/raw_feature_bc_matrix'

```

Load data using DropletUtils

```{r loadData}
utdData = read10xCounts(UTDDataPath,version='3', col.names = T)
cmoData = read10xCounts(CMODataPath,version='3', col.names = T)
nxtData = read10xCounts(NXTDataPath,version='3', col.names = T)
htoData = read10xCounts(HTODataPath,version='3', col.names = T)

cmoData$Protocol = "CellPlex"
nxtData$Protocol = "MULTI-Seq CMO"
htoData$Protocol = "Hashtag Ab"
utdData$Protocol <- "Unlabelled"

cmoData <- splitAltExps(
  cmoData,
  rowData(cmoData)$Type,
  "Gene Expression")

nxtData <- splitAltExps(
  nxtData,
  rowData(nxtData)$Type,
  "Gene Expression")

htoData <- splitAltExps(
  htoData,
  rowData(htoData)$Type,
  "Gene Expression")

counts(altExp(htoData))
```

Convert to list of sces

```{r saveIntRDS}
sces <- list(cmoData, htoData, nxtData, utdData)
names(sces) <- c("CellPlex", "Hashtag Ab", "MULTI-Seq CMO" ,"Unlabelled")

saveRDS(sces, "/stornext/Projects/score/Analyses/R010_multiplexing/SCEs/PDX/PDX_raw_feature_matrix.sces.rds")

# sces <- readRDS("/stornext/Projects/score/Analyses/R010_multiplexing/SCEs/PDX/PDX_raw_feature_matrix.sces.rds")

sces_orig <- sces
```

# Droplet Utils cell calling

```{r bcr_all, fig.cap="Barcode plots all samples"}
par(mfrow = c(2, 2)) # We have 4 captures
par(mar=c(1,1,1,1))
for (s in levels(sces$capture)) {
  sce <- sces[[s]] 
  bcrank <- barcodeRanks(counts(sce), lower=50) 

  # Only showing unique points for plotting speed.
  uniq <- !duplicated(bcrank$rank)
  plot(
    x = bcrank$rank[uniq],
    y = bcrank$total[uniq],
    log = "xy",
    xlab = "Rank",
    ylab = "Total UMI count",
    main = s,
    cex.lab = 1.2,
    xlim = c(1, 500000),
    ylim = c(1, 200000))
  abline(h = metadata(bcrank)$inflection, col = "darkgreen", lty = 2)
  abline(h = metadata(bcrank)$knee, col = "dodgerblue", lty = 2)
}
```

We use the default parameters for empty drops

```{r emptyDrops}

cmo.empties <- emptyDrops(counts(cmoData))
cmoData <- cmoData[,which(cmo.empties$FDR <= 0.005)]

nxt.empties <- emptyDrops(counts(nxtData))
nxtData <- nxtData[,which(nxt.empties$FDR <= 0.005)]

hto.empties <- emptyDrops(counts(htoData))
htoData <- htoData[,which(hto.empties$FDR <= 0.005)]

utd.empties <- emptyDrops(counts(utdData))
utdData <- utdData[,which(utd.empties$FDR <= 0.001)]

capture_names = c("CellPlex", "Hashtag Ab", "MULTI-Seq CMO", "Unlabelled")
cell_counts = c(sum(cmo.empties$FDR <= 0.005, na.rm=TRUE),
                sum(nxt.empties$FDR <= 0.005, na.rm=TRUE),
                sum(hto.empties$FDR <= 0.005, na.rm=TRUE),
                sum(utd.empties$FDR <= 0.005, na.rm=TRUE))

empties <- list(cmo.empties, hto.empties, nxt.empties, utd.empties)
names(empties) <- capture_names
saveRDS(empties, "/vast/scratch/users/brown.d/S000322/S000322_empties.rds")

```

Summarise the empty drops calls.

I was targeting 20,000 cell containing droplets for 16,000 singlets in this experiment. In all captures this value is wildly different in empty drops calls.

```{r cell_num_table}
calldata = as.data.frame(cell_counts)
calldata = t(calldata)
colnames(calldata) = capture_names
rownames(calldata) = "Non-Empty Calls"
knitr::kable(calldata,
             caption = "Counts of non-empty droplets selected by emptyDrops().")
```

# Manual cell calling from droplets {.tabset}

Set manual UMI thresholds otherwise I retrieve too few cells for a useful analysis.

## Manual unlabelled

A 800 UMI threshold looks like a more conservative value to have more real cell containing droplets rather than DropletUtils 60,000

```{r unlabelled_knee}
c <- "Unlabelled"
umi_threshold_utd <- 600
sce <- sces[[c]] 

bcrank <- barcodeRanks(counts(sce), lower=50) 

uniq <- !duplicated(bcrank$rank)
tb <- as_tibble(bcrank[uniq,])
inflect <- metadata(bcrank)$inflection
knee <- metadata(bcrank)$knee

plt1 <- ggplot(tb, 
             aes(x = rank, y=total+1)) + 
  geom_point(size=0.5, alpha=0.75) + 
  xlab("Rank") + 
  ylab("Total UMI count") +
  ggtitle("Unlabelled capture") +
  geom_hline(yintercept=inflect, color = "darkgreen") +
  geom_hline(yintercept=knee, color = "dodgerblue") +
  geom_hline(yintercept=umi_threshold_utd, color = "red") +
  scale_y_continuous(trans='log10') +
  scale_x_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "bl")

plt1

```

## Manual CellPlex

I use a 100 UMI threshold here simply to retrieve suficient cells for an analysis

```{r cellplex_knee}
c <- "CellPlex"
umi_threshold_cmo <- 100
sce <- sces[[c]] 

bcrank <- barcodeRanks(counts(sce), lower=50) 

uniq <- !duplicated(bcrank$rank)
tb <- as_tibble(bcrank[uniq,])
inflect <- metadata(bcrank)$inflection
knee <- metadata(bcrank)$knee

plt2 <- ggplot(tb, 
             aes(x = rank, y=total+1)) + 
  geom_point(size=0.5, alpha=0.75) + 
  xlab("Rank") + 
  ylab("Total UMI count") +
  ggtitle("CellPlex capture") +
  geom_hline(yintercept=inflect, color = "darkgreen") +
  geom_hline(yintercept=knee, color = "dodgerblue") +
  geom_hline(yintercept=umi_threshold_cmo, color = "red") +
  scale_y_continuous(trans='log10') +
  scale_x_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "bl")

plt2

```

## Manual Hashtag Ab

```{r hashtag_knee}
c <- "Hashtag Ab"
umi_threshold_hto <- 250
sce <- sces[[c]] 

bcrank <- barcodeRanks(counts(sce), lower=50) 

uniq <- !duplicated(bcrank$rank)
tb <- as_tibble(bcrank[uniq,])
inflect <- metadata(bcrank)$inflection
knee <- metadata(bcrank)$knee

plt3 <- ggplot(tb, 
             aes(x = rank, y=total+1)) + 
  geom_point(size=0.5, alpha=0.75) + 
  xlab("Rank") + 
  ylab("Total UMI count") +
  ggtitle(paste(c, "capture", sep=" ")) +
  geom_hline(yintercept=inflect, color = "darkgreen") +
  geom_hline(yintercept=knee, color = "dodgerblue") +
  geom_hline(yintercept=umi_threshold_hto, color = "red") +
  scale_y_continuous(trans='log10') +
  scale_x_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "bl")

plt3

```

## Manual MULTI-Seq CMO

```{r MULTIseq_knee}
c <- "MULTI-Seq CMO"
umi_threshold_nxt <- 450
sce <- sces[[c]] 

bcrank <- barcodeRanks(counts(sce), lower=50) 

uniq <- !duplicated(bcrank$rank)
tb <- as_tibble(bcrank[uniq,])
inflect <- metadata(bcrank)$inflection
knee <- metadata(bcrank)$knee

plt4 <- ggplot(tb, 
             aes(x = rank, y=total+1)) + 
  geom_point(size=0.5, alpha=0.75) + 
  xlab("Rank") + 
  ylab("Total UMI count") +
  ggtitle(paste(c, "capture", sep=" ")) +
  geom_hline(yintercept=inflect, color = "darkgreen") +
  geom_hline(yintercept=knee, color = "dodgerblue") +
  geom_hline(yintercept=umi_threshold_nxt, color = "red") +
  scale_y_continuous(trans='log10') +
  scale_x_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "bl")

plt4

```

# Combined barcode rank plot

```{r combined_barcode_rank, fig.width=12, fig.height=12}
plt1 + plt2 + plt3 + plt4
```

# Filter sces by manual threshold

```{r}
sces_filter <- sces

names(sces)
umi_threshold <- list(umi_threshold_cmo, umi_threshold_hto, umi_threshold_nxt, umi_threshold_utd)
names(umi_threshold) <- names(sces)

for (n in names(sces)) {
  current <- sces[[n]]
  current$lib_size <- colSums2(counts(current))
  current <- current[,current$lib_size >= umi_threshold[[n]]]
  sces_filter[[n]] <- current
}

cell_counts_manual = c(length(colnames(sces_filter[[1]])),
                length(colnames(sces_filter[[2]])),
                length(colnames(sces_filter[[3]])),
                length(colnames(sces_filter[[4]]))
                )
```

## Compare the empty drops cell numbers vs the manual thresholds

Most captures have not been materially affected but the unlabelled capture is a more realistic value.

```{r}
cell_count <- rbind(cell_counts, cell_counts_manual)
colnames(cell_count) <- names(sces)
row.names(cell_count) <- c("emptyDrops", "manual_umi")
cell_count
```

## Write objects to file

SCE object

```{r}
saveRDS(sces_filter, "/vast/scratch/users/brown.d/S000322/S000322.manual_umi.sces.rds")
```

Write valid cell barcodes to file for cell snp lite

```{r, eval=FALSE}
for (n in names(sces_filter)) {
  current <- sces_filter[[n]]
  bc <- colnames(current)
  
  # For cell snplite need to remove the "-" separator
  bc_noHypen <- str_split(string = bc, pattern = "-", simplify = T)[,1]
  output <- paste("OvarianCarcinosarcoma/data/S000322/cell_bcs/", n, "_wellBCs.csv", sep="")
  write.table(x = bc_noHypen, file = output, quote=F, col.names = F, row.names = F)
  
  # Fo cell subsetting need to kep hypen
  output <- paste("OvarianCarcinosarcoma/data/S000322/cell_bcs/", n, "_wellBCs-hypen.csv", sep="")
  write.table(x = bc, file = output, quote=F, col.names = F, row.names = F)
}
```

Write unlabelled nuclei containing barcodes.

```{r}
sce_l <- readRDS("/vast/scratch/users/brown.d/S000322/S000322.manual_umi.sces.rds")
sce <- sce_l[["Unlabelled"]]
# saveRDS(sce, "/vast/scratch/users/brown.d/S000322/G000178_unlabelled_nuclei_basic-PDX.rds")

cell_bcs <- colnames(sce) 
write.table(x = cell_bcs, file = "OvarianCarcinosarcoma/data/S000322/cell_bcs/Unlabelled_validBCs-hypen.csv", quote=F, col.names = F, row.names = F)
```


Write the top 1000 cell barcodes by library size for the Unlabelled sample. Try to get Cell-SNP-Lite to run.

```{r, eval=FALSE, include=FALSE}
sce_unlab <- sces_filter[["Unlabelled"]]
cd_unlab <- colData(sce_unlab)
cd_unlab <- cd_unlab[order(cd_unlab$lib_size, decreasing = TRUE),]
cb_unlab <- cd_unlab[c(1:1000),2]
write.table(x = cb_unlab, file = "OvarianCarcinosarcoma/data/S000322/cell_bcs/unlabelled_top1k_wellBCs-hypen.csv", quote=F, col.names = F, row.names = F)
```

## Session Info

```{r}
sessionInfo()
```
