---
title: "Ambient RNA removal ovarian carcinosarcoma"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/ambient_rna/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
library(DropletUtils)
library(tidyverse)
library(patchwork)
library(scran)
library(scater)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim
I previous compared Total-Seq A hashtag, MULTI-Seq and CellPlex lipid on human PBMCs. All protocols worked pretty well but human PBMCs are the easiest sample to work with.

Then I tried mouse embryonic brain E18.5 which didn't work that well. CellPlex diluted 1:10 worked best here.

## This sample
Run a more challenging sample = Ovarian carcinosarcome PDX nuclei.  
Four PDXs were processed with 10x Genomics nuclei isolation kit. A capture was immdiately performed which is the unlabelled.  
Three more aliquots were labelled in singlicate with cell multiplexing reagents and captured later.

## Comparisons

* Fresh unlabelled nuclei after nuclei isolation with 10x Genomics kit
* Fixed nuclei nuclei isolation with 10x Genomics kit
* Fresh nuclei CellPlex 1:10 diluted
* Fresh nuclei custom MULTI-Seq CMO (Nextera mod)
* Fresh nuclei 1ug nucleus hashtag Mab414

# Load Data

Find the paths to the cell ranger output. From cell_calling notebook.  
Start by removing ambient RNA from the unlabelled capture only to check if it is usful or not.

```{r loadData}
sces <- readRDS("/vast/scratch/users/brown.d/S000322/S000322.manual_umi.sces.rds")
names(sces)

current <- sces[["Unlabelled"]]
```

## Empty drops calls

From cell_calling notebook

```{r loadEmptyDrops}
empties <- readRDS("/vast/scratch/users/brown.d/S000322/S000322_empties.rds")
e.out <- empties[["Unlabelled"]]
```

## Quick cluster

Need cluster labels for th ambient RNA removal. Performed first at cluster level then propagated back to single cells.

```{r norm_cluster}
current <- logNormCounts(current)
set.seed(1000)
clusters <- quickCluster(current, block=NULL)
colLabels(current) <- factor(clusters)
```

# Removing ambient contamination {.tabset}

Follow the OSCA tutorial. 
http://bioconductor.org/books/3.15/OSCA.advanced/droplet-processing.html#removing-ambient-contamination

Rather than attempting to remove contamination from individual cells, a more measured approach is to operate on clusters of related cells. The removeAmbience() function from DropletUtils will remove the contamination from the cluster-level profiles and propagate the effect of those changes back to the individual cells. 

```{r ambient_rm}
# Not all genes are reported in the ambient profile from emptyDrops,
# as genes with counts of zero across all droplets are just removed.
# So for convenience, we will restrict our analysis to genes with 
# non-zero counts in at least one droplet (empty or otherwise).
amb <- metadata(e.out)$ambient[,1]
stripped <- current[names(amb),]

out <- removeAmbience(counts(stripped), ambient=amb, groups=colLabels(stripped))
dim(out)

counts(stripped, withDimnames=FALSE) <- out
stripped <- logNormCounts(stripped)
```

## Visualise high expressors

The mouse gene should bee a good choice to check ambient RNA as it is a xenograft expereiment and mouse genes should be only in a few cells.

```{r highExpr}
ph <- plotHighestExprs(current, n=10, colour_cells_by = "label")

ph
```

## Visualise mouse gene counts before and after ambient removal

ENSMUSG00000098178 = Yam1 which is a lincRNA. 
There does not appear to be a large difference. The distribution of Yam1 is pushed down a bit but not at all by zero.

```{r mouse_expr_ambient}
gene <- "mm10___ENSMUSG00000098178"

p1 <- plotExpression(current, x="label", colour_by="label", features=gene) + 
        ggtitle("Before ambient removal")


p2 <- plotExpression(stripped, x="label", colour_by="label", features=gene) + 
        ggtitle("After ambient removal")

p1 / p2
```

Save the ambient RNA corrected untreated sample back into the list of sces

```{r save}
sces[["Unlabelled Ambient remove"]] <- stripped
saveRDS(sces, "/vast/scratch/users/brown.d/S000322/S000322.rm_ambient.sces.rds")
```

## Session Info

```{r}
sessionInfo()
```
