---
title: "Mung SingleCellExperiment object"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/mung_SCE/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
suppressPackageStartupMessages({
  library(DropletUtils)
  library(tidyverse)
  library(patchwork)
  library(scran)
  library(scater)
  })

source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim
I previous compared Total-Seq A hashtag, MULTI-Seq and CellPlex lipid on human PBMCs. All protocols worked pretty well but human PBMCs are the easiest sample to work with.

Then I tried mouse embryonic brain E18.5 which didn't work that well. CellPlex diluted 1:10 worked best here.

## This sample
Run a more challenging sample = Ovarian carcinosarcome PDX nuclei.  
Four PDXs were processed with 10x Genomics nuclei isolation kit. A capture was immdiately performed which is the unlabelled.  
Three more aliquots were labelled in singlicate with cell multiplexing reagents and captured later.

## Comparisons

* Fresh nuclei CellPlex 1:10 diluted
* Fresh nuclei custom MULTI-Seq CMO (Nextera mod)
* Fresh nuclei 1ug nucleus hashtag Mab414
* Fresh unlabelled nuclei after nuclei isolation with 10x Genomics kit
* Fresh unlabelled nuclei with ambient RNA removal

# Load Data

Find the paths to the cell ranger output. From cell_calling notebook.  
SNP la

```{r loadData}
sces <- readRDS("/vast/scratch/users/brown.d/S000322/S000322.snp-join.rds")
names(sces)
```

## Split the human and mouse counts
Alt experiments enables ERCC, CITE-Seq hashtag data etc to be split from the main gene expression matrix.

```{r}
for (i in 1:length(names(sces))) {
  current <- sces[[i]]
  is.mouse <- grepl("^mm10___", rownames(current))
  current <- splitAltExps(current, ifelse(is.mouse, "Mouse", "gene"))
  sces[[i]] <- current
}

```

## Add gene names

Human

```{r}
for (i in 1:length(names(sces))) {
  current <- sces[[i]]
  gene_symbol <- rowData(current)$Symbol
  gene_symbol <- gsub(pattern = "^GRCh38_", 
                       replacement = "", x = gene_symbol)
  row.names(current) <- gene_symbol
  sces[[i]] <- current
}
```

Mouse

```{r}
for (i in 1:length(names(sces))) {
  current <- sces[[i]]
  gene_symbol <- rowData(altExp(current, "Mouse"))$Symbol
  gene_symbol <- gsub(pattern = "^mm10___", 
                       replacement = "", x = gene_symbol)
  row.names(altExp(current, "Mouse")) <- gene_symbol
}
```

## Add cell QC

```{r}
for (i in 1:length(names(sces))) {
  current <- sces[[i]]
  is.mito <- grepl("^MT", rownames(current))
  current <- addPerCellQCMetrics(current, subsets=list(Mito=is.mito))
  current <- addPerFeatureQC(current)
  sces[[i]] <- current
}

```

## Add cellhashR calls

### CellPlex

```{r chr_cellplex}
tb <- read_tsv(here::here(
  "OvarianCarcinosarcoma/data/cellhashR", "cmo_tag_callFile_cellHashR.txt"
))

cd <- as_tibble(colData(sces[["CellPlex"]]))
cd <- cbind(cd, tb)

cd <- cd %>% 
  select(-cellbarcode)

colData(sces[["CellPlex"]]) <- DataFrame(cd)

rm(tb, cd)
```

### MULTI-Seq CMO

```{r chr_multiseq}
tb <- read_tsv(here::here(
  "OvarianCarcinosarcoma/data/cellhashR", "NXT_tag_callFile_cellHashR.txt"
))

cd <- as_tibble(colData(sces[["MULTI-Seq CMO"]]))
cd <- cbind(cd, tb)

cd <- cd %>% 
  select(-cellbarcode)

colData(sces[["MULTI-Seq CMO"]]) <- DataFrame(cd)

rm(tb, cd)
```

### Hashtag Ab

```{r chr_hashtag}
tb <- read_tsv(here::here(
  "OvarianCarcinosarcoma/data/cellhashR", "hto_tag_callFile_cellHashR.txt"
))

cd <- as_tibble(colData(sces[["Hashtag Ab"]]))
cd <- cbind(cd, tb)

cd <- cd %>% 
  select(-cellbarcode)

colData(sces[["Hashtag Ab"]]) <- DataFrame(cd)

rm(tb, cd)
```

## Extract colData for plotting

```{r}
col_data_multi <- rbind(
  colData(sces[[1]]),
  colData(sces[[2]]),
  colData(sces[[3]])
)

col_data_utd <- colData(sces[[4]])

drops <- c("sizeFactor","label")

# Need to combine col_data_multi and col_data_utd
colnames(col_data_multi)
colnames(col_data_utd)

keep_cols <- intersect(
  colnames(col_data_multi),
  colnames(col_data_utd)
)

col_data <- as_tibble(rbind(
  col_data_multi[,keep_cols],
                  col_data_utd[,keep_cols]
  ))
```

## Save objects

```{r}
saveRDS(sces, "/stornext/Projects/score/Analyses/R010_multiplexing/SCEs/PDX/S000322.qcmetrics.sces.rds")

write_csv(col_data, here::here(
  "OvarianCarcinosarcoma/data/Matrices", "quality_control_unified.csv.gz"
))
write_csv(as_tibble(col_data_multi), here::here(
  "OvarianCarcinosarcoma/data/Matrices", "quality_control_multi.csv.gz"
))
```

## Session Info

```{r}
sessionInfo()
```
