---
title: "Quality control and dimension reduction"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=1, 
                      fig.path='./figures/qc_plots/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
suppressPackageStartupMessages({
  library(tidyverse)
  library(patchwork)
  library(scuttle)
  library(scater)
  library(scran)
  library(PCAtools)
  })

source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication(base_size = 18))
```

# Aim

Remove the low quality cells and genes that are not expressed.  
Perform dimension reduction with PCA and UMAP

## Samples

* Fresh nuclei CellPlex 1:10 diluted
* Fresh nuclei custom MULTI-Seq CMO (Nextera mod)
* Fresh nuclei 1ug nucleus hashtag Mab414
* Fresh unlabelled nuclei after nuclei isolation with 10x Genomics kit
* Fresh unlabelled nuclei with ambient RNA removal
# Load Data

From the 1C_mung_SCE notebook.

```{r loadData}
sce_list <- readRDS((
  "/stornext/Projects/score/Analyses/R010_multiplexing/SCEs/PDX/S000322.qcmetrics.sces.rds"
))
names(sce_list)

for (i in 1:length(names(sce_list))) {
  current <- sce_list[[i]]
  colnames(current) <- current$cell
  #is.mito <- grepl("^MT", rownames(current))
  #current <- addPerCellQCMetrics(current, subsets=list(Mito=is.mito))
  #current <- addPerFeatureQC(current)
  sce_list[[i]] <- current
}

sce_list_orig <- sce_list
```

# Identifying outliers by each metric

In calling cells from empty droplets, we have already removed cells with very low library sizes or (by association) low numbers of expressed genes. Thus, further filtering on these metrics is not strictly necessary.

Filtering on the mitochondrial proportion provides the most additional benefit in this situation and so we seek to identify droplets with unusually large mitochondrial proportions (i.e. outliers). Outlier thresholds are defined based on the median absolute deviation (MADs) from the median value of the metric across all cells

```{r}
mito_drop <- list()
mouse_drop <- list()

for (n in names(sce_list)) {
  current <- sce_list[[n]]
  mito_drop[[n]] <- isOutlier(
    metric = current$subsets_Mito_percent, 
    nmads = 3, type = "higher")
  current <- current[,!mito_drop[[n]]]
  
  mouse_drop[[n]] <- isOutlier(
    metric = current$altexps_Mouse_percent, 
    nmads = 3)
  current <- current[,!mouse_drop[[n]]]
  
  sce_list[[n]] <- current
}

```

Measure how many cells dropped

```{r outlier_detect}
qc_df <- list()
reasons <- list()

for (n in names(sce_list)) {
  current <- sce_list[[n]]
  qc_df[[n]] <- colData(current)
  reasons[[n]] <- as.matrix(perCellQCFilters(
    x=qc_df[[n]], 
    sub.fields=c("subsets_Mito_percent", "altexps_Mouse_percent")))
  discard <- reasons[[n]]
  discard <- discard[,4]
  current <- current[,!discard]
  sce_list[[n]] <- current
}

colnames(reasons[[1]])
lapply(reasons, colSums2)
```

Summarise how many cells left

```{r}
cell_drop_tb <- rbind(
  cbind(length(colnames(sce_list_orig[[1]])), length(colnames(sce_list[[1]]))),
  cbind(length(colnames(sce_list_orig[[2]])), length(colnames(sce_list[[2]]))),
  cbind(length(colnames(sce_list_orig[[3]])), length(colnames(sce_list[[3]]))),
  cbind(length(colnames(sce_list_orig[[4]])), length(colnames(sce_list[[4]])))
)

colnames(cell_drop_tb) <- c("Before_Filter", "After_filter")
row.names(cell_drop_tb) <- names(sce_list)
cell_drop_tb
```

# Filtering out low-abundance genes

Low-abundance genes are problematic as zero or near-zero counts do not contain much information for reliable statistical inference. These genes typically do not provide enough evidence to reject the null hypothesis during testing, yet they still increase the severity of the multiple testing correction. In addition, the discreteness of the counts may interfere with statistical procedures. Thus, low-abundance genes are often removed in many RNA-seq analysis pipelines before the application of downstream methods.

Several metrics can be used to define low-abundance genes. The most obvious is the average count for each gene, computed across all cells in the data set. We typically observe a peak of moderately expressed genes following a plateau of lowly expressed genes.

The distributions look the same between captures.

```{r average_count_p}
ave_counts <- list()

for (n in names(sce_list)) {
  current <- sce_list[[n]]
  ave_counts[[n]] <- calculateAverage(current)
  to_keep <- ave_counts[[n]] > 0
  current <- current[to_keep, ]
  sce_list[[n]] <- current
}

par(mfrow = c(2, 2))
hist(
  x = log10(ave_counts[[1]]), 
  breaks = 100, 
  main = "CellPlex", 
  col = "grey",
  xlab = expression(Log[10] ~ "average count"))

hist(
  x = log10(ave_counts[[3]]), 
  breaks = 100, 
  main = "MULTI-Seq CMO", 
  col = "grey",
  xlab = expression(Log[10] ~ "average count"))

hist(
  x = log10(ave_counts[[2]]), 
  breaks = 100, 
  main = "Hashtag Ab", 
  col = "grey",
  xlab = expression(Log[10] ~ "average count"))

hist(
  x = log10(ave_counts[[4]]), 
  breaks = 100, 
  main = "Unlabelled", 
  col = "grey",
  xlab = expression(Log[10] ~ "average count"))
```

Subset the count matrices for genes that are not expressed. About 12,000 genes are removed. That still leaves 24,000 genes.

```{r filter_gene}

# Summarise how many cells left
gene_drop_tb <- rbind(
  cbind(length(rownames(sce_list_orig[[1]])), length(rownames(sce_list[[1]]))),
  cbind(length(rownames(sce_list_orig[[2]])), length(rownames(sce_list[[2]]))),
  cbind(length(rownames(sce_list_orig[[3]])), length(rownames(sce_list[[3]]))),
  cbind(length(rownames(sce_list_orig[[4]])), length(rownames(sce_list[[4]])))
)

colnames(gene_drop_tb) <- c("Before_Filter", "After_filter")
row.names(gene_drop_tb) <- names(sce_list)
gene_drop_tb
```

# GEX Normalization

## Using the deconvolution method to deal with zero counts

We pool counts from many cells to increase the count size for accurate size factor estimation (A. T. Lun, Bach, and Marioni 2016). Pool-based size factors are then ‘deconvolved’ into cell-based factors for cell-specific normalization. This removes scaling biases associated with cell-specific differences in capture efficiency, sequencing depth and composition biases.

```{r normalise}
set.seed(666)

clusters <- list()
for (n in names(sce_list)) {
  current <- sce_list[[n]]
  clusters[[n]] <- quickCluster(current)
  current <- computeSumFactors(current, clusters = clusters[[n]], min.mean = 0.1)
  current <- logNormCounts(current)
  print(summary(sizeFactors(current)))
  sce_list[[n]] <- current
}
```

# Selecting highly variable genes (HVGs)

## Fit variance
```{r var_genes_p, fig.cap="Highly variable genes. Top=CellPlex, Middle=MULTI-Seq, Bottom=Hashtag"}
var_fit <- list()
hvg <- list()

for (n in names(sce_list)) {
  current <- sce_list[[n]]
  var_fit[[n]] <- modelGeneVarByPoisson(current)
  hvg[[n]] <- getTopHVGs(var_fit[[n]], var.threshold = 0)
}
```

```{r}
p1 <- plotExpression(object = sce_list[[1]], features = hvg[[1]][1:10])
p2 <- plotExpression(object = sce_list[[2]], features = hvg[[2]][1:10])
p3 <- plotExpression(object = sce_list[[3]], features = hvg[[3]][1:10])
p4 <- plotExpression(object = sce_list[[4]], features = hvg[[4]][1:10])

p1 + p2 + p3 + p4 + 
  plot_layout(guides = "collect", ncol = 2)
```

# Dimension reduction {.tabset}

## PCA

```{r PCA}
set.seed(666)

for (n in names(sce_list)) {
  current <- sce_list[[n]]
  current <- denoisePCA(current, var_fit[[n]], subset.row = hvg[[n]])
  sce_list[[n]] <- current
}
```

### Number of PCA dimensions to retain

Based on these percent variance plots retain 20 PCA dimensions

```{r}
percent.var <- list()

for (n in names(sce_list)) {
  current <- sce_list[[n]]
  # Percentage of variance explained is tucked away in the attributes.
  percent.var[[n]] <- attr(reducedDim(current), "percentVar")
}

par(mfrow = c(2, 2))
for (i in 1:4) {
  plot(percent.var[[i]], xlab="PC", ylab="Variance explained (%)")
}

```

### Visualise PCA

```{r pca_plts, fig.height=12, fig.width=16}
pca_plt <- list()

for (n in names(sce_list)) {
  pca_plt[[n]] <- plotReducedDim(sce_list[[n]], dimred="PCA", colour_by = "detected")
}

pca_plt[[1]] + pca_plt[[2]] + pca_plt[[3]] +
  pca_plt[[4]] + plot_layout(ncol=2)
```

Look at the PCA of only the unlabelled sample

```{r pca_plt_unlabelled, fig.width=8, fig.height=2}
unlabelled <- sce_list[["Unlabelled"]]


pca_plt <- plotReducedDim(unlabelled, dimred="PCA", colour_by = "detected") + theme_Publication(base_size = 16)
pca_plt2 <- plotReducedDim(unlabelled, dimred="PCA", colour_by = "donor_id") + theme_Publication(base_size = 16)

pca_plt + pca_plt2
```

## Run UMAP

```{r umap}
set.seed(100)

for (n in names(sce_list)) {
  current <- sce_list[[n]]
  current <- runUMAP(current, dimred = "PCA", n_dimred=20)
  sce_list[[n]] <- current
}

```

## Visualise UMAP

```{r umap_plts, fig.height=12, fig.width=16}
umap_plt <- list()

for (n in names(sce_list)) {
  current <- sce_list[[n]]
  umap_plt[[n]] <- plotReducedDim(current, dimred="UMAP", colour_by="detected")
}

umap_plt[[1]] + umap_plt[[2]] + umap_plt[[3]] +
  umap_plt[[4]] + plot_layout(ncol=2)
```

## Save SCE object

```{r saveSCEs}
saveRDS(sce_list, here::here(
  "/vast/scratch/users/brown.d/S000322/S000322.dimred.rds"
))
saveRDS(hvg, here::here(
  "/vast/scratch/users/brown.d/S000322/S000322.hvg.rds"
))
```

## Session Info

```{r}
sessionInfo()
```
