---
title: "Oligo tag performance"
author: "Daniel Brown"
description: "Investigate the unassigned, doublet and multiplet calls"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/oligo_tag_performance/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())
library(scater)
library(patchwork)
library(knitr)
library(tidyverse)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim
I previous compared Total-Seq A hashtag, MULTI-Seq and CellPlex lipid on human PBMCs. All protocols worked pretty well but human PBMCs are the easiest sample to work with.

Then I tried mouse embryonic brain E18.5 which didn't work that well. CellPlex diluted 1:10 worked best here.

## This sample
Run a more challenging sample = Ovarian carcinosarcome PDX nuclei.  
Four PDXs were processed with 10x Genomics nuclei isolation kit. A capture was immdiately performed which is the unlabelled.  
Three more aliquots were labelled in singlicate with cell multiplexing reagents and captured later.

## Comparisons

* Fresh nuclei CellPlex 1:10 diluted
* Fresh nuclei custom MULTI-Seq CMO (Nextera mod)
* Fresh nuclei 1ug nucleus hashtag Mab414
* Fresh unlabelled nuclei after nuclei isolation with 10x Genomics kit
* Fresh unlabelled nuclei with ambient RNA removal

# Load Data

These SCE objects were generated in the 1C_mungSCE notebook.  
Subset the captures that were labelled with oligo tag reagents.  
The ceellranger multi did not produce calls for these datasets because the esignal to noise was so bad. Rely on cellhashR calls particulary BFF_Cluster.

```{r loadData}
sces <- readRDS("/vast/scratch/users/brown.d/S000322/S000322.qcmetrics.sces.rds")
sces <- sces[c(1:3)]
names(sces)

chr_sum <- read_csv(here::here(
  "OvarianCarcinosarcoma/data/cellhashR", "cellhashR_labels.csv.gz"
)) %>%
  select(Protocol:dropletutils) %>% 
  pivot_longer(bff_cluster:dropletutils) %>% 
  dplyr::count(Protocol, name, value) %>% 
  group_by(Protocol,name) %>% 
  mutate(pct = round(n/sum(n)*100, digits = 1))

chr_sum$value <- recode(chr_sum$value,
                        "Not Called" = "Negative")

chr_sum %>% 
  arrange(desc(n))
```

Extract the coldata and bind into a tibble.

```{r}
tb <- rbind(
  colData(sces[[1]]),
  colData(sces[[2]]),
  colData(sces[[3]])
)

tb <- as_tibble(tb)

tb %>% 
  select(-Sample, -altexps_Mouse_sum, -altexps_Mouse_detected, -altexps_Mouse_percent)
```

## Explain cell metadata

* Protocol
  + The cell multiplexing protocol I used. See comparisions section above.
* sum
  + Library size of gene expression (umi counts) per cell
* detected
  + Number of genes detected
* subsets_Mito_percent
  + Percent mitochondrial umi counts per cell
* altexps_Multiplexing Capture_sum
  + Library size of multiplexing oligo tags per cell
* bff_cluster : dropletutils
  + CellhashR https://bimberlab.github.io/cellhashR/ demultiplexing calls per cell
  + CellhashR runs many common demultiplexing algorithms
* consensuscall
  + The agreed demultiplexing call for the demultiplexing algorithms
  + Note this does not include cell ranger multi which is not included in cellhashR

# Visualizing the QC metrics based on Cell ranger Multi {.tabset}

Number of cells from each protocol

I was aiming for 20,000 cell-containing droplets for about 16,000 singlets. Due to the ambient RNA cellranger multi did not call cells accurately and I did it manually by DropletUtils.

The cell containing droplet numbers are much lower but similar between captures. There is evidence that Hashtag Ab had a less deterimental effect to the nucleus integrity than the lipid protocols and this is reflected in number of cells recovered.

```{r}
tb %>% 
  dplyr::count(Protocol) %>% 
  kable()
```

```{r}
scheme <- c("#A260DA", "chocolate", "darkseagreen", "cornflowerblue", "darkred", "#999999")
```

## Visulaise cell ranger multi confidence scores

I can't merge these with the main single cell experiment object because the cell calling by cell ranger was ineffective.  
A possible solution is running cell ranger again but supplying a whitelist of barcodes.

```{r cellranger_mung}
crm_cmo <- read_csv(here::here(
  "OvarianCarcinosarcoma/data/S000322/cellranger_outs", "G000225_PDX_CMO", "multi/multiplexing_analysis/assignment_confidence_table.csv"
))
crm_cmo$Protocol <- "CellPlex"

crm_nxt <- read_csv(here::here(
  "OvarianCarcinosarcoma/data/S000322/cellranger_outs", "G000225_PDX_NXT", "multi/multiplexing_analysis/assignment_confidence_table.csv"
))
crm_nxt$Protocol <- "MULTI-Seq CMO"

crm_hto <- read_csv(here::here(
  "OvarianCarcinosarcoma/data/S000322/cellranger_outs", "G000225_PDX_HTO", "multi/multiplexing_analysis/assignment_confidence_table.csv"
))
crm_hto$Protocol <- "Hashtag Ab"

crm <- rbind(
  crm_cmo[,c("Protocol","Assignment", "Assignment_Probability")],
  crm_nxt[,c("Protocol","Assignment", "Assignment_Probability")],
  crm_hto[,c("Protocol", "Assignment", "Assignment_Probability")]
)
```

Visulaise the assignment probability

```{r cellranger_assign}

ggplot(crm, aes(Assignment_Probability, fill = Protocol)) +
  geom_density(alpha = 0.2) +
  xlab("Cell Ranger Assignment\n Probability") + 
  scale_colour_brewer(palette = "Dark2") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

```

## Visualise the percent calls each tag

Based on these plots I will choose to proceed with Seurat HTO demux as it has the most consistent recovery.

Need to do somee rreecoding of the fields otherwise there are too many categorical variables to plot.

```{r tag_percent}
scheme <- c("chocolate", "#999999", "#008A15", "#1AD2DE", "#D0E532", "#006FB2")

chr_sum_tag <- chr_sum

chr_sum_tag$value <- recode(chr_sum_tag$value,
                            "CMO301" = "Tag1",
                            "CMO302" = "Tag2",
                            "CMO303" = "Tag3",
                            "CMO304" = "Tag4",
                            "HTO_1" = "Tag1",
                            "HTO_2" = "Tag2",
                            "HTO_3" = "Tag3",
                            "HTO_4" = "Tag4",
                            "Nxt_451" = "Tag1",
                            "Nxt_452" = "Tag2",
                            "Nxt_453" = "Tag3",
                            "Nxt_455" = "Tag4",
)
chr_sum_tag$Protocol <- recode(chr_sum_tag$Protocol,
                            "CMO" = "CellPlex",
                            "NXT" = "MULTI-Seq CMO",
                            "HTO" = "Hashtag Ab",

)

chr_sum_tag$name <- recode(chr_sum_tag$name,
                            "bff_cluster" = "BFF Cluster",
                            "dropletutils" = "DropletUtils",
                            "gmm_demux" = "GMM Demux",
                           "htodemux" = "srt HTO Demux",
                           "multiseq" = "deMULTIplex"

)


plt1 <- ggplot(chr_sum_tag, 
             aes(x = name, y= pct, fill = value)) + 
  geom_bar(position="stack", stat="identity") +
  ylab("Percent of capture") + 
  xlab("") + 
  scale_y_continuous(breaks=seq(0,100,20)) +
  scale_fill_manual(values=scheme, name = "Droplet call") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Protocol)

plt1
```

## Visualise the percent calls summary to singlet

```{r tag_percent_singlet}
chr_sum_tag_sing <- chr_sum_tag

chr_sum_tag_sing$value <- recode(chr_sum_tag_sing$value,
                                 "Tag1" = "Singlet",
                                 "Tag2" = "Singlet",
                                 "Tag3" = "Singlet",
                                 "Tag4" = "Singlet"

)

plt2 <- ggplot(chr_sum_tag_sing, 
             aes(x = name, y= pct, fill = value)) + 
  geom_bar(position="stack", stat="identity") +
  ylab("Percent of capture") + 
  xlab("") + 
  scale_y_continuous(breaks=seq(0,100,20)) +
   scale_fill_manual(values=c("#E69F00", "#999999", "#56B4E9"), name = "Droplet call") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~Protocol)

plt2
```

# TO DO: Visualise umi and gene detected metrics by droplet call {.tabset}

```{r}
tb$htodemuxTag <- recode(tb$htodemux,
                            "CMO301" = "Tag1",
                            "CMO302" = "Tag2",
                            "CMO303" = "Tag3",
                            "CMO304" = "Tag4",
                            "HTO_1" = "Tag1",
                            "HTO_2" = "Tag2",
                            "HTO_3" = "Tag3",
                            "HTO_4" = "Tag4",
                            "Nxt_451" = "Tag1",
                            "Nxt_452" = "Tag2",
                            "Nxt_453" = "Tag3",
                            "Nxt_455" = "Tag4",
)

p3 <- ggplot(tb, 
             aes(x = htodemuxTag, y= altexps_Multiplexing.Capture_sum+1, fill = htodemuxTag)) + 
  geom_violin(show.legend = FALSE) + 
  ylab("Library Size (UMIs)") + 
  xlab("") +
  scale_y_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "l") +
  scale_fill_brewer(palette = "Dark2") +
  xlab("srt HTO Demux Call") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size=16)) +
  facet_wrap(~Protocol)

p3
```


## Library size

Need to run this part if and when I can get genotyping to work.

```{r, eval=FALSE}
droplet_tb <- tb %>% 
  select(Barcode, Protocol,sum:altexps_Multiplexing.Capture_percent,Demuxafy_assignment:bff_droplet) %>% 
  pivot_longer(
    cols = c(Demuxafy_assignment:bff_droplet),
    names_to = "Algorithm",
    values_to = "DropletCall",)
```

```{r libSize_fAlgorithm, fig.cap="Library size in deduplicated UMIs", eval=FALSE}
plt1 <- ggplot(droplet_tb, 
             aes(x = DropletCall, y= sum+1, fill = Algorithm)) + 
  geom_violin(show.legend = T) + 
  ylab("Library Size (UMIs)") + 
  xlab("") +
  scale_y_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "l") +
  scale_fill_brewer(palette = "Dark2") +
  facet_wrap(~Protocol)

plt1
```

```{r libSize_fDropletCall, fig.cap="Library size in deduplicated UMIs", fig.height=6,fig.width=10, eval=FALSE}
plt1 <- ggplot(droplet_tb, 
             aes(x = Algorithm, y= sum+1, fill = DropletCall)) + 
  geom_violin(show.legend = T) + 
  ylab("Library Size (UMIs)") + 
  xlab("") +
  scale_y_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "l") +
  scale_fill_brewer(palette = "Set2", aesthetics = c("Multiplet", "Singlet", "Unassigned")) +
  facet_wrap(~Protocol)

plt1
```

## Multiplexing capture oligos

No clear difference in unassigned and singlets except for LMO.

```{r oligp_count_p, fig.cap="Library size of multiplexing capture sequencing", fig.height=6,fig.width=10, eval=FALSE}
plt4 <- ggplot(droplet_tb, 
             aes(x = DropletCall, y= altexps_Multiplexing.Capture_sum+1, fill = Algorithm)) + 
  geom_violin(show.legend = T) + 
  ylab("Library Size (UMIs)") + 
  xlab("") +
  scale_y_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "l") +
  scale_fill_brewer(palette = "Dark2") +
  facet_wrap(~Protocol)

plt4
```

## Put together

```{r qc_summary_p, fig.cap="Summary of QC metrics", fig.width=12, fig.height=8, eval=FALSE}
plt1 + plt2 + plt3 + plt4 + 
  plot_layout(guides = "collect", ncol = 2)
```

# Plot library size by dual cell ranger multi and demuxafy label

## BFF cluster

About the best visualization I can achieve in the absence of ground truth for genetic identical mouse embryo is to compare doublet calls from gene expression and oligo tag demultiplexing algorithm.

```{r mung_doublet_bff, eval=FALSE}
dob_tb <- tb %>% 
  select(Barcode,sum,detected,subsets_Mito_percent,Protocol,Demuxafy_assignment:bff_droplet)

# Combine columns and recode
dob_tb$class <- paste("bff",dob_tb$bff_droplet,
                      "gex", dob_tb$Demuxafy_assignment,
                      sep="-")

# remove unassigned for clarity of plotting
dob_tb <- dob_tb %>% 
  filter(bff_droplet != "unassigned") %>% 
  filter(Demuxafy_assignment != "unassigned")
```

Violin style

```{r libSize_bff, fig.cap="Library size in deduplicated UMIs", eval=FALSE}
unique(dob_tb$class)

p1 <- ggplot(dob_tb, 
             aes(x = class, y= sum+1, fill = class)) + 
  geom_violin(show.legend = FALSE) + 
  ylab("Library Size (UMIs)") + 
  xlab("") +
  scale_y_continuous(trans='log10') +
  scale_x_discrete(labels=c("bff-singlet-gex-singlet" = "S\n\nS", 
                            "bff-multiplet-gex-singlet" = "M\n\nS",
                            "bff-multiplet-gex-multiplet" = "M\n\nM",
                            "bff-singlet-gex-multiplet" = "S\n\nM")) +
  annotation_logticks(base = 10, sides = "l") +
  scale_fill_brewer(palette = "Set1") +
  theme(axis.text.x = element_text(angle = 0, hjust = 1, size=14)) +
  facet_wrap(~Protocol)

p1
```

Summarise number of cells in each class

```{r, eval=FALSE}
dob_tb %>% 
  dplyr::count(class,Protocol) %>% 
  pivot_wider(names_from = Protocol, values_from = n) %>% 
  kable()
```

Mitochondrial percentage.  
Not much difference here

```{r mito_bff, fig.cap="Library size in deduplicated UMIs", eval=FALSE}
unique(dob_tb$class)

p2 <- ggplot(dob_tb, 
             aes(x = class, y= subsets_Mito_percent, fill = class)) + 
  geom_violin(show.legend = FALSE) + 
  ylab("Mitochondrial Percent") + 
  xlab("") +
  scale_x_discrete(labels=c("bff-singlet-gex-singlet" = "S\n\nS", 
                            "bff-multiplet-gex-singlet" = "M\n\nS",
                            "bff-multiplet-gex-multiplet" = "M\n\nM",
                            "bff-singlet-gex-multiplet" = "S\n\nM")) +
  scale_fill_brewer(palette = "Set1") +
  theme(axis.text.x = element_text(angle = 0, hjust = 1, size=14)) +
  facet_wrap(~Protocol)

p2
```

## Cell ranger multi

```{r mung_doublet_cr, eval=FALSE}
dob_tb <- tb %>% 
  select(Barcode,sum,detected,subsets_Mito_percent,Protocol,Demuxafy_assignment:bff_droplet)

# Combine columns and recode
dob_tb$class <- paste("cr",dob_tb$multi_droplet,
                      "gex", dob_tb$Demuxafy_assignment,
                      sep="-")

# remove unassigned for clarity of plotting
dob_tb <- dob_tb %>% 
  filter(multi_droplet != "unassigned") %>% 
  filter(Demuxafy_assignment != "unassigned")
```

Violin style

```{r libSize_cr, fig.cap="Library size in deduplicated UMIs", eval=FALSE}
unique(dob_tb$class)

p3 <- ggplot(dob_tb, 
             aes(x = class, y= sum+1, fill = class)) + 
  geom_violin(show.legend = FALSE) + 
  ylab("Library Size (UMIs)") + 
  xlab("") +
  scale_y_continuous(trans='log10') +
  scale_x_discrete(labels=c("cr-singlet-gex-singlet" = "S\n\nS", 
                            "cr-multiplet-gex-singlet" = "M\n\nS",
                            "cr-multiplet-gex-multiplet" = "M\n\nM",
                            "cr-singlet-gex-multiplet" = "S\n\nM")) +
  annotation_logticks(base = 10, sides = "l") +
  scale_fill_brewer(palette = "Set1") +
  theme(axis.text.x = element_text(angle = 0, hjust = 1, size=14)) +
  facet_wrap(~Protocol)

p3
```

Summarise number of cells in each class

```{r, eval=FALSE}
dob_tb %>% 
  dplyr::count(class,Protocol) %>% 
  pivot_wider(names_from = Protocol, values_from = n) %>% 
  kable()
```

Mitochondrial percentage.  
Not much difference here

```{r mito_cr, fig.cap="Library size in deduplicated UMIs", eval=FALSE}
unique(dob_tb$class)

p4 <- ggplot(dob_tb, 
             aes(x = class, y= subsets_Mito_percent, fill = class)) + 
  geom_violin(show.legend = FALSE) + 
  ylab("Mitochondrial Percent") + 
  xlab("") +
  scale_x_discrete(labels=c("cr-singlet-gex-singlet" = "S\n\nS", 
                            "cr-multiplet-gex-singlet" = "M\n\nS",
                            "cr-multiplet-gex-multiplet" = "M\n\nM",
                            "cr-singlet-gex-multiplet" = "S\n\nM")) +
  scale_fill_brewer(palette = "Set1") +
  theme(axis.text.x = element_text(angle = 0, hjust = 1, size=14)) +
  facet_wrap(~Protocol)

p4
```

## Cell Ranger multi vs BFF Cluster same plot

This is hard to interpret due to bimodal distribution of the singlets.

```{r mung_doublet_both, eval=FALSE}
dob_tb <- tb %>% 
  select(Barcode,sum,detected,Protocol,Demuxafy_assignment:bff_droplet)

# Combine columns and recode
dob_tb$class <- paste("cr",dob_tb$multi_droplet,
                      "bff", dob_tb$bff_droplet,
                      sep="-")

# remove where unassigned for clarity of plotting
dob_tb <- dob_tb %>% 
  filter(class != "cr-unassigned-bff-unassigned") %>% 
  filter(class != "cr-unassigned-bff-multiplet")

```

Violin style

```{r libSize_both, fig.cap="Library size in deduplicated UMIs", eval=FALSE}
unique(dob_tb$class)

p3 <- ggplot(dob_tb, 
             aes(x = class, y= sum+1, fill = class)) + 
  geom_violin(show.legend = FALSE) + 
  ylab("Library Size (UMIs)") + 
  xlab("") +
  scale_y_continuous(trans='log10') +
  scale_x_discrete(labels=c("cr-multiplet-bff-multiplet" = "M\n\nM",
                            "cr-multiplet-bff-singlet" = "M\n\nS", 
                            "cr-singlet-bff-multiplet" = "S\n\nM",
                            "cr-singlet-bff-unassigned" = "S\n\nU",
                            "cr-unassigned-bff-singlet" = "U\n\nS",
                            "cr-unassigned-bff-multiplet" = "U\n\nM",
                            "cr-singlet-bff-singlet" = "S\n\nS")) +
  annotation_logticks(base = 10, sides = "l") +
  scale_fill_brewer(palette = "Set1") +
  theme(axis.text.x = element_text(angle = 0, hjust = 1, size=14)) +
  facet_wrap(~Protocol)

p3
```

Summarise number of cells in each class

```{r, eval=FALSE}
dob_tb %>% 
  dplyr::count(class,Protocol) %>% 
  pivot_wider(names_from = Protocol, values_from = n) %>% 
  kable()
```

## Combine the violin charts of demuxafy and oligo tag demultiplexing algorithm

```{r combine_gex_oligo, fig.width=8, fig.height=12, eval=FALSE}
p1 + p2 + p3 + p4
```

# Conclusion

To write.

## Session Info

```{r}
sessionInfo()
```
