---
title: "GEX marker genes"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/gex_markers/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
suppressPackageStartupMessages({
  library(tidyverse)
  library(patchwork)
  library(scuttle)
  library(scater)
  library(scran)
  library(viridis)
  })

source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

Perform differential expression test on clusters to find marker genes.

## Samples

* Fresh nuclei CellPlex 1:10 diluted
* Fresh nuclei custom MULTI-Seq CMO (Nextera mod)
* Fresh nuclei 1ug nucleus hashtag Mab414
* Fresh unlabelled nuclei after nuclei isolation with 10x Genomics kit
* Fresh unlabelled nuclei with ambient RNA removal
# Load Data

From the 2C_QC_dimred notebook

```{r loadData}
sce_list <- readRDS((
  "/vast/scratch/users/brown.d/S000322/S000322.cluster.rds"
))

capture <- names(sce_list)
capture
```

## Recap clusters in a UMAP

```{r umap_cluster, fig.height=12, fig.width=12}
umap_plt <- list()

for (i in 1:length(sce_list)) {
  current <- sce_list[[i]]
  umap_plt[[i]] <- plotReducedDim(current, dimred="UMAP", colour_by="cluster") + ggtitle(capture[i])
}

umap_plt[[1]] + umap_plt[[2]] + umap_plt[[3]] +
  umap_plt[[4]]
```

# Unlabelled marker gene detection

For each cluster, we rank candidate markers based on one of these effect size summaries.  
Use pairwise comparisons rather than comparing each cluster to the average of all other cells.

```{r}
current <- sce_list[["Unlabelled"]]

marker.info <- scoreMarkers(current, groups=current$cluster)
```

Compare each pair of clusters and compute scores quantifying the differences in the expression distributions between clusters.  
The scores for all pairwise comparisons involving a particular cluster are then consolidated into a single DataFrame for that cluster.

## Iterate over each cluster

```{r}
markers <- list()

for (i in 1:length(marker.info)) {
    chosen <- marker.info[[i]]
    ordered <- chosen[order(chosen$mean.AUC, decreasing=TRUE),]

    chosen <- chosen[order(chosen$mean.AUC, decreasing=TRUE),]
    markers[[i]] <- as.data.frame(chosen[c(1:5),])
}

names(markers) <- 1:length(markers)
```

## Mung the list of marker genes

Convert to a summarised and usable form for all clusters

```{r}
marker <- do.call(rbind, markers)
marker$gene <- row.names(marker)

marker <- as_tibble(marker) %>% 
  select(gene,self.average, mean.logFC.cohen, median.logFC.cohen, 
         mean.AUC, median.AUC, 
         mean.logFC.detected, median.logFC.detected) %>% 
  separate(gene, c("cluster", "gene.symbol"), sep = 2) %>% 
  filter(mean.logFC.cohen > 0.60)

marker
```

## Unlabelled visualise results

### Dot plot

```{r unlabelled_marker}
marker_genes <- unique(gsub("^\\.", "",marker$gene.symbol))

dot_unlabel<- plotDots(current, 
         features = marker_genes, 
         group="cluster", color=plasma(100)) +
  labs(title = "", y = "", x = "Cluster", color = "Log counts", size="Proportion of cells") +
  theme_Publication()

dot_unlabel
```

### Expression plots

```{r unlabelled_expression, eval=FALSE}

plotExpression(sce_list[["Unlabelled"]], features=marker_genes, 
    x="cluster", colour_by="cluster") +
  theme_Publication()
```

### Save markers

```{r}
protocol <- "unlabelled"

saveRDS(markers, here::here(
   "OvarianCarcinosarcoma/data/Matrices/marker_genes", paste(protocol,"marker_list.rds", sep="_")
))

write_csv(marker, here::here(
   "OvarianCarcinosarcoma/data/Matrices/marker_genes", paste(protocol,"top_marker_list.csv", sep="_")
))
```

# Hashtag antibody marker gene detection

For each cluster, we rank candidate markers based on one of these effect size summaries.  
Use pairwise comparisons rather than comparing each cluster to the average of all other cells.

```{r}
current <- sce_list[["Hashtag Ab"]]

marker.info <- scoreMarkers(current, groups=current$cluster)
```

Compare each pair of clusters and compute scores quantifying the differences in the expression distributions between clusters.  
The scores for all pairwise comparisons involving a particular cluster are then consolidated into a single DataFrame for that cluster.

## Iterate over each cluster

```{r}
markers <- list()

for (i in 1:length(marker.info)) {
    chosen <- marker.info[[i]]
    ordered <- chosen[order(chosen$mean.AUC, decreasing=TRUE),]

    chosen <- chosen[order(chosen$mean.AUC, decreasing=TRUE),]
    markers[[i]] <- as.data.frame(chosen[c(1:5),])
}

names(markers) <- 1:length(markers)
```

## Mung the list of marker genes

Convert to a summarised and usable form for all clusters

```{r}
marker <- do.call(rbind, markers)
marker$gene <- row.names(marker)

marker <- as_tibble(marker) %>% 
  select(gene,self.average, mean.logFC.cohen, median.logFC.cohen, 
         mean.AUC, median.AUC, 
         mean.logFC.detected, median.logFC.detected) %>% 
  separate(gene, c("cluster", "gene.symbol"), sep = 2) %>% 
  filter(mean.logFC.cohen > 0.60)

marker
```

## Hashtag antibody visualise results

### Dot plot

```{r haashtag_marker}
marker_genes <- marker$gene.symbol

dot_unlabel<- plotDots(current, 
         features = marker_genes, 
         group="cluster", color=plasma(100)) +
  labs(title = "", y = "", x = "Cluster", color = "Log counts", size="Proportion of cells") +
  theme_Publication()

dot_unlabel
```

### Expression plots

```{r hashtag_expression, eval=FALSE}

plotExpression(current, features=marker_genes, 
    x="cluster", colour_by="cluster") +
  theme_Publication()
```

### Save markers

```{r}
protocol <- "hashtag"

saveRDS(markers, here::here(
   "OvarianCarcinosarcoma/data/Matrices/marker_genes", paste(protocol,"marker_list.rds", sep="_")
))

write_csv(marker, here::here(
   "OvarianCarcinosarcoma/data/Matrices/marker_genes", paste(protocol,"top_marker_list.csv", sep="_")
))
```

# MULTI-Seq CMO marker gene detection

For each cluster, we rank candidate markers based on one of these effect size summaries.  
Use pairwise comparisons rather than comparing each cluster to the average of all other cells.

```{r}
current <- sce_list[["MULTI-Seq CMO"]]

marker.info <- scoreMarkers(current, groups=current$cluster)
```

Compare each pair of clusters and compute scores quantifying the differences in the expression distributions between clusters.  
The scores for all pairwise comparisons involving a particular cluster are then consolidated into a single DataFrame for that cluster.

## Iterate over each cluster

```{r}
markers <- list()

for (i in 1:length(marker.info)) {
    chosen <- marker.info[[i]]
    ordered <- chosen[order(chosen$mean.AUC, decreasing=TRUE),]

    chosen <- chosen[order(chosen$mean.AUC, decreasing=TRUE),]
    markers[[i]] <- as.data.frame(chosen[c(1:5),])
}

names(markers) <- 1:length(markers)
```

## Mung the list of marker genes

Convert to a summarised and usable form for all clusters

```{r}
marker <- do.call(rbind, markers)
marker$gene <- row.names(marker)

marker <- as_tibble(marker) %>% 
  select(gene,self.average, mean.logFC.cohen, median.logFC.cohen, 
         mean.AUC, median.AUC, 
         mean.logFC.detected, median.logFC.detected) %>% 
  separate(gene, c("cluster", "gene.symbol"), sep = 2) %>% 
  filter(mean.logFC.cohen > 0.60)

marker
```

## Multi-Seq CMO visualise results

### Dot plot

```{r multiseq_marker}
marker_genes <- marker$gene.symbol
marker_genes <- gsub("\\.", "",marker_genes)

marker_genes <- marker_genes[marker_genes %in% row.names(current)]

dot_unlabel<- plotDots(current, 
         features = marker_genes, 
         group="cluster", color=plasma(100)) +
  labs(title = "", y = "", x = "Cluster", color = "Log counts", size="Proportion of cells") +
  theme_Publication()

dot_unlabel
```

### Expression plots

```{r muliseq_expression, eval=FALSE}

plotExpression(current, features=marker_genes, 
    x="cluster", colour_by="cluster") +
  theme_Publication()
```

### Save markers

```{r}
protocol <- "multiseq"

saveRDS(markers, here::here(
   "OvarianCarcinosarcoma/data/Matrices/marker_genes", paste(protocol,"marker_list.rds", sep="_")
))

write_csv(marker, here::here(
   "OvarianCarcinosarcoma/data/Matrices/marker_genes", paste(protocol,"top_marker_list.csv", sep="_")
))
```

# CellPlex marker gene detection

For each cluster, we rank candidate markers based on one of these effect size summaries.  
Use pairwise comparisons rather than comparing each cluster to the average of all other cells.

```{r}
current <- sce_list[["CellPlex"]]

marker.info <- scoreMarkers(current, groups=current$cluster)
```

Compare each pair of clusters and compute scores quantifying the differences in the expression distributions between clusters.  
The scores for all pairwise comparisons involving a particular cluster are then consolidated into a single DataFrame for that cluster.

## Iterate over each cluster

```{r}
markers <- list()

for (i in 1:length(marker.info)) {
    chosen <- marker.info[[i]]
    ordered <- chosen[order(chosen$mean.AUC, decreasing=TRUE),]

    chosen <- chosen[order(chosen$mean.AUC, decreasing=TRUE),]
    markers[[i]] <- as.data.frame(chosen[c(1:5),])
}

names(markers) <- 1:length(markers)
```

## Mung the list of marker genes

Convert to a summarised and usable form for all clusters

```{r}
marker <- do.call(rbind, markers)
marker$gene <- row.names(marker)

marker <- as_tibble(marker) %>% 
  select(gene,self.average, mean.logFC.cohen, median.logFC.cohen, 
         mean.AUC, median.AUC, 
         mean.logFC.detected, median.logFC.detected) %>% 
  separate(gene, c("cluster", "gene.symbol"), sep = 2) %>% 
  filter(mean.logFC.cohen > 0.60)

marker
```

## CellPlex visualise results

### Dot plot

```{r cellplex_marker}
marker_genes <- marker$gene.symbol

dot_unlabel<- plotDots(current, 
         features = marker_genes, 
         group="cluster", color=plasma(100)) +
  labs(title = "", y = "", x = "Cluster", color = "Log counts", size="Proportion of cells") +
  theme_Publication()

dot_unlabel
```

### Expression plots

```{r cellplex_expression, eval=FALSE}

plotExpression(current, features=marker_genes, 
    x="cluster", colour_by="cluster") +
  theme_Publication()
```

### Save markers

```{r}
protocol <- "cellplex"

saveRDS(markers, here::here(
   "OvarianCarcinosarcoma/data/Matrices/marker_genes", paste(protocol,"marker_list.rds", sep="_")
))

write_csv(marker, here::here(
   "OvarianCarcinosarcoma/data/Matrices/marker_genes", paste(protocol,"top_marker_list.csv", sep="_")
))
```

## Session Info

```{r}
sessionInfo()
```
