---
title: "Oligo Visualisation based on GEX labels"
author: "Daniel Brown"
description: "After annotating cells based on gene exprssion labels visualise tag clusters"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      cache.lazy = FALSE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/oligo_cluster_vis/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())
suppressPackageStartupMessages({
  library(scater)
  library(scran)
  library(scuttle)
  library(pheatmap)
  library(patchwork)
  library(knitr)
  library(tidyverse)
})
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

Follow annotation of cells and clusters based on gene expression labels visualise the cells dimension reduced based on oligo tag counts.

## This sample
Run a more challenging sample = Ovarian carcinosarcome PDX nuclei.  
Four PDXs were processed with 10x Genomics nuclei isolation kit. A capture was immdiately performed which is the unlabelled.  
Three more aliquots were labelled in singlicate with cell multiplexing reagents and captured later.

## Comparisons

* Fresh unlabelled nuclei after nuclei isolation with 10x Genomics kit
* Fixed nuclei nuclei isolation with 10x Genomics kit
* Fresh nuclei CellPlex 1:10 diluted
* Fresh nuclei custom MULTI-Seq CMO (Nextera mod)
* Fresh nuclei 1ug nucleus hashtag Mab414

## Load Data

From the 3A_GEX_marker_genes notebook

I remove the unlabelled capture that does not contain any oligo tag information as these were not labelled.

```{r loadData}
sce_list <- readRDS((
  "/vast/scratch/users/brown.d/S000322/S000322.cluster.rds"
))
names(sce_list)

# Remove the unlabelled captures which do not have a altExps slot
sce_list <- sce_list[c(1:3)]

# Swap out the tag oligo slot for the gex slot as the primary count object
for (n in names(sce_list)) {
    current <- swapAltExp(sce_list[[n]], name="Multiplexing Capture", saved="GEX")
    sce_list[[n]] <- current
}
```

# Oligo Normalization

## Median normalization

Use the most basic normalisation as composition biases and other differences are of interest to my study. 

```{r normalise}
set.seed(666)

for (n in names(sce_list)) {
    current <- sce_list[[n]] 
    sf.geo <- geometricSizeFactors(current)
    sizeFactors(current) <- sf.geo
    current <- logNormCounts(current)
    sce_list[[n]] <- current
}

assayNames(current)
```

# Dimension reduction
We can apply downstream procedures like clustering and visualization on the log-normalized abundance matrix for the ADTs (Figure 12.7). Alternatively, if we had generated a matrix of PCs, we could use that as well.

```{r umap}
set.seed(999)

for (n in names(sce_list)) {
    current <- sce_list[[n]] 
    current <- runUMAP(current)
    sce_list[[n]] <- current
}
```

# Plot UMAP colour by gene expression clusters {.tabset}

The gene expression cluster labels are used to colour the UMAP plots were the dimension reduction is done on olgio tag counts.
All dimension reduction plots are relatively poor compared to the PBMC and mouse brain experiment.

## CellPlex 1:10 dilution

```{r umap_tagCluster_gexLabel_cmo, fig.cap="CellPlex"}
p1 <- plotUMAP(sce_list[["CellPlex"]], 
         colour_by="cluster") +
         ggtitle("CellPlex") + theme_Publication()

p1
```

## Hashtag antibody

```{r umap_tagCluster_gexLabel_hto, fig.cap="HashtagAb"}
p2 <- plotUMAP(sce_list[["Hashtag Ab"]], 
         colour_by="cluster") +
         ggtitle("Hashtag Ab") + theme_Publication()
p2
```

## Custom MULTI-Seq CMO

```{r umap_tagCluster_gexLabel_nxt, fig.cap="MULTI-Seq"}
p3 <- plotUMAP(sce_list[["MULTI-Seq CMO"]], 
         colour_by="cluster") +
         ggtitle("MULTI-Seq CMO") + theme_Publication()

p3
```

# UMAP together

```{r umap_combine_gexLabel, fig.asp=1, fig.cap="Combined UMAP"}
p1 + theme_Publication() +
  p2 + theme_Publication() + 
  p3 + theme_Publication() +
  plot_layout(guides = "collect", ncol = 2)

```

# Plot UMAP colour by oligo tag calls {.tabset}

This is the more useful visualization and can go in the manuscript. I chose Seurat HTO demux because this returned the most calls where the proporrtions were consistent with the cell labelling.

## CellPlex 1:10 dilution

```{r umap_tagCluster_tagLabel_cmo, fig.cap="CellPlex"}
p1 <- plotUMAP(sce_list[["CellPlex"]], 
         colour_by="htodemux") +
         ggtitle("CellPlex") + 
  theme_Publication()

p1
```

## Hashtag antibody

```{r umap_tagCluster_tagLabel_hto, fig.cap="HashtagAb"}
p2 <- plotUMAP(sce_list[["Hashtag Ab"]], 
         colour_by="htodemux") +
         ggtitle("Hashtag Ab") + theme_Publication()
p2
```

## Custom MULTI-Seq CMO

```{r umap_tagCluster_tagLabel_nxt, fig.cap="MULTI-Seq"}
p3 <- plotUMAP(sce_list[["MULTI-Seq CMO"]], 
         colour_by="htodemux") +
         ggtitle("MULTI-Seq CMO") + theme_Publication()

p3
```

# UMAP together

```{r umap_combine_taglabel, fig.asp=0.8, fig.cap="Combined UMAP"}
p1 + theme(legend.text = element_text(size = 12), legend.title=element_blank()) +
  p2 + theme(legend.text = element_text(size = 12), legend.title=element_blank()) + 
  p3 + theme(legend.text = element_text(size = 12), legend.title=element_blank()) +
  plot_layout(ncol = 2)
```

# Heatmaps {.tabset}

Heatmap is looking like the most useful visualization. Here I order the columns (cells) by the Seurat HTOdemux calls instead of euclidean clustering of the oligo tag counts.

```{r sumcounts}
all_ave <- list()

for (n in names(sce_list)) {
    current <- sce_list[[n]]
    se.averaged <- sumCountsAcrossCells(current,
                                        ids=current$htodemux, # Change to donor identity when info is available
    exprs_values="logcounts", average=TRUE)
    all_ave[[n]] <- assay(se.averaged)
}

names(all_ave) <- names(sce_list)

fontSize=12
fontsizeRow = 20
legendFontSize = 6

# Set colour scheme to sequential
col_fun <- colorRampPalette(c("#FFFFFF", "#FF0000"))(50)
```

## CellPlex

```{r cellplex_heat, fig.asp=0.95}
current <- "CellPlex"
set.seed(666)
cells <- sample(2500)

normcounts <- logcounts(sce_list[[current]][,cells])

# Update my_sample_col data when genotypes are available
my_sample_col <- as.data.frame(colData(
  sce_list[[current]][,cells]
  )[,c("cluster","htodemux", "consensuscall.global", "donor_id")]
  )

my_sample_col <- my_sample_col[order(my_sample_col$htodemux),]
colnames(my_sample_col) <- c("GEX cluster", "srt Demux", "Consensus call", "SNP_ID")

pheatmap(normcounts[,row.names(my_sample_col)],
         annotation_col = my_sample_col,
         cluster_rows=FALSE,
         cluster_cols = FALSE,
         treeheight_col=0,
         show_rownames=TRUE,
         show_colnames = FALSE,
         fontsize=fontSize,
         color = col_fun,
         row.center = TRUE,
         legend_font_size = legendFontSize,
         fontsize_row = fontsizeRow,
         labels_col=row.names(current)
         )
  
```

## MULTI-Seq CMO

```{r MULTI-Seq_heat, fig.asp=0.95}
current <- "MULTI-Seq CMO"
set.seed(666)
cells <- sample(2500)

normcounts <- logcounts(sce_list[[current]][,cells])

# Update my_sample_col data when genotypes are available
my_sample_col <- as.data.frame(colData(
  sce_list[[current]][,cells]
  )[,c("cluster","htodemux", "consensuscall.global", "donor_id")]
  )

my_sample_col <- my_sample_col[order(my_sample_col$htodemux),]

colnames(my_sample_col) <- c("GEX cluster", "srt Demux", "Consensus call", "SNP_ID")

pheatmap(normcounts[,row.names(my_sample_col)],
         annotation_col = my_sample_col,
         cluster_rows=FALSE,
         cluster_cols = FALSE,
         treeheight_col=0,
         show_rownames=TRUE,
         show_colnames = FALSE,
         fontsize=fontSize,
         color = col_fun,
         row.center = TRUE,
         fontsize_row = fontsizeRow,
         legend_font_size = legendFontSize,
         labels_col=row.names(current)
         )
```

## Hashtag Ab

```{r hashtag_heat, fig.asp=0.95}
current <- "Hashtag Ab"
set.seed(666)
cells <- sample(2500)

normcounts <- logcounts(sce_list[[current]][,cells])

# Update my_sample_col data when genotypes are available
my_sample_col <- as.data.frame(colData(
  sce_list[[current]][,cells]
  )[,c("cluster","htodemux", "consensuscall.global", "donor_id")]
  )

my_sample_col <- my_sample_col[order(my_sample_col$htodemux),]

colnames(my_sample_col) <- c("GEX cluster", "srt Demux", "Consensus call", "SNP_ID")

pheatmap(normcounts[,row.names(my_sample_col)],
         annotation_col = my_sample_col,
         cluster_rows=FALSE,
         cluster_cols = FALSE,
         treeheight_col=0,
         show_rownames=TRUE,
         show_colnames = FALSE,
         fontsize=fontSize,
         color = col_fun,
         row.center = TRUE,
         fontsize_row = fontsizeRow,
         legend_font_size = legendFontSize,
         labels_col=row.names(current)
         )
```

# Save SCE Objects

```{r saveSCEs, eval=FALSE}
for (n in names(sce_list)) {
    sce_list[[n]] <- swapAltExp(sce_list[[n]], "GEX")
}

saveRDS("/vast/scratch/users/brown.d/S000322/S000322.tag_dimred.rds")

```

## Conclusion

* Using the gene expression information was not particularly useful.
* SNP demultiplexing by vireo was also not useful as far too few cells were called

## Session Info

```{r}
sessionInfo()
```
