---
title: "GEX check annotation Seurat map reference"
author: "Daniel Brown"
description: "Check the cell annotations with Seurat map reference to determine the difference"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='../figures/checkSRTAnno/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())
library(patchwork)
library(scuttle)
library(scater)
library(knitr)
library(Seurat)
library(tidyverse)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication(base_size = 18))
```

## Recap
The purpose of this experiment was to compare the performance of jumpcode CRISPRclean which is the commercialised version of DASH. 

In the notebook GEX_cluster_annotate I used SingleR to provide cell annotations. Especially in the Untreated they are not really consistent.

## Aim

Check the cell annotations more comprehensively and choose the best one for presentation.

## Sample

I restrict my analysis to R010_MULTI-Seq_GEX in PBMCs. This sample had the best signal to noise.

I sequenced this particular sample much deeper as part of NN265.  
https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/MzQ3LjF8NjgxOTcvMjY3LTE0NzgvVHJlZU5vZGUvMjMwMTMwNzYyN3w4ODEuMDk5OTk5OTk5OTk5OQ==

I sequenced both CRISPRclean treated and untreated.  
To enable sequencing of both I had William repeat the libary prep of the untreated:

https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/MzQ1Ljh8NjgxOTcvMjY2L1RyZWVOb2RlLzM4OTk3NzUzNTd8ODc3Ljg=

# Load Data

## Seurat

The object is a list of Seurat objects. The first is the CRISPRclean and the second the untreated.  
These objects are generated in the *seurat_ref_map* notebook.

```{r loadSRT}
#all.srt <- readRDS(file = here::here(
#  "data/genomics_data/SCEs/PBMCs", "mapped.srt.rds"
#))

all.srt <- readRDS(file = "/stornext/Projects/score/Analyses/R010_multiplexing/SCEs/PBMCs/cell_annotation.srt.rds"
)

names(all.srt)
```

## SCE

```{r loadSCE}
cmo_sce <- readRDS(here::here(
  "PBMC/data/PBMCs", "cmo-annotate.rds"))

lmo_sce <- readRDS(here::here(
  "PBMC/data/PBMCs", "lmo-annotate.rds"))

hto_sce <-  readRDS(here::here(
  "PBMC/data/PBMCs", "hto-annotate.rds"))


all.sce <- list(cmo_sce, hto_sce, lmo_sce)
names(all.sce) <- names(all.srt)
```

## Join Seurat ref map result to SCE

```{r}
for (n in names(all.sce)) {
  curr_sce <- all.sce[[n]]
  curr_srt <- all.srt[[n]]
  colData(curr_sce) <- DataFrame(curr_srt[[]])
  all.sce[[n]] <- curr_sce
}

```

# Visualise on UMAP {.tabset}

## Seurat map reference

```{r cluster_anno_srt}
all.umap <- list()
for (n in names(all.sce)) {
    all.umap[[n]] <- plotUMAP(all.sce[[n]], 
                              colour_by="predicted.celltype.l1",
                              text_by="predicted.celltype.l1", text_size=3) +
      ggtitle(n)
}

all.umap[[1]] + 
  all.umap[[2]] + theme(legend.position='none') +
  all.umap[[3]] + theme(legend.position='none') +
  plot_layout(guides = "collect", ncol = 3)
```

Check for a doublet cluster

## MULTI-Seq annotation

```{r cluster_anno_multiseq}
all.umap <- list()
for (n in names(all.sce)) {
    all.umap[[n]] <- plotUMAP(all.sce[[n]], 
                              colour_by="mul_Assignment",
                              text_by="predicted.celltype.l1", text_size=3) +
      ggtitle(n)
}

all.umap[[1]] + 
  all.umap[[2]] + theme(legend.position='none') +
  all.umap[[3]] + theme(legend.position='none') +
  plot_layout(guides = "collect", ncol = 3)
```

# Explore marker gene expression in PBMCs for each of these annotations {.tabset}

I got these labels from https://satijalab.org/seurat/articles/pbmc3k_tutorial.html

| Cluster ID | Markers       | Cell Type    |
|------------|---------------|--------------|
| 0          | IL7R, CCR7    | Naive CD4+ T |
| 1          | CD14, LYZ     | CD14+ Mono   |
| 2          | IL7R, S100A4  | Memory CD4+  |
| 3          | MS4A1         | B            |
| 4          | CD8A          | CD8+ T       |
| 5          | FCGR3A, MS4A7 | FCGR3A+ Mono |
| 6          | GNLY, NKG7    | NK           |
| 7          | FCER1A, CST3  | DC           |
| 8          | PPBP          | Platelet     |

```{r}
markers <- c("IL7R", "CCR7","CD14", "LYZ",
             "S100A4", "MS4A1", "CD8A",
             "FCGR3A", "MS4A7", "GNLY",
             "NKG7", "FCER1A", "CST3", "PPBP"
)
```

# Calculate cell type proportions compared to expected PBMCs

## Mung colData

Recode the various columns to be consistent with other figures.

```{r extr_tb_anno}
tb <- as_tibble(
  rbind(
  colData(all.sce[[1]]),
  colData(all.sce[[2]]),
  colData(all.sce[[3]])
))

tb$Protocol <- dplyr::recode(tb$Protocol,
                             CMO = "CellPlex",
                             HTO = "Hashtag Ab",
                             LMO = "MULTI-Seq")

tb$vir_Donor <- recode(tb$vir_Donor,
                      "VBDR1186" = "donor0",
                      "VBDR1206" = "donor1",
                      "VBDR1216" = "donor2",
                      "VBDR1233" = "donor3"
)

tb$mul_Donor <- recode(tb$mul_Donor,
                      "VBDR1186" = "donor0",
                      "VBDR1206" = "donor1",
                      "VBDR1216" = "donor2",
                      "VBDR1233" = "donor3"
)
```


```{r}
tb %>% 
  select(cell_id, Protocol,
         predicted.celltype.l1, predicted.celltype.l2,
         mul_Donor, vir_Donor) %>% 
  head() %>% 
  kable()
```

## Calculate percentage of cell type in each protocol

### Seurat map reference

```{r cal-prop_azi}
cellProp_srt <- tb %>% 
  filter(mul_Donor != "empty") %>% 
  filter(mul_Donor != "multiplet") %>% 
  filter(mul_Donor != "unassigned") %>% 
  select(cell_id, predicted.celltype.l1, 
         Protocol, mul_Donor) %>%
  dplyr::count(predicted.celltype.l1, Protocol, mul_Donor) %>%
  group_by(Protocol, mul_Donor) %>% 
  mutate(percent = n / sum(n) * 100) %>%
  arrange(desc(percent))

cellProp_srt
```

# Plot the PBMC cell type proportions {.tabset}

## Donor colour, protocol shape

```{r per_anno_srt, fig.cap="Immune subset proportion from each oligo tag multiplexing protocol"}
p1 <- ggplot(cellProp_srt, 
             aes(x = predicted.celltype.l1, y= percent, shape = Protocol, colour = mul_Donor)) + 
  geom_jitter(size=3, width = 0.15, alpha=0.75) +
  ylab("Percent of capture") + 
  xlab("") + ggtitle("") + 
    scale_y_continuous(breaks=seq(0,50,10)) +
  scale_color_brewer(palette = "Dark2", name = "Donor") +
  scale_shape(name="Protocol") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

p1
```

## Protocol colour

Make the same plot but without the donor encoded in the shape

```{r per_anno_srt_plain}
p2 <- ggplot(cellProp_srt, 
             aes(x = predicted.celltype.l1, y= percent, colour = Protocol)) + 
  geom_jitter(size=3, width = 0.15, alpha=0.75) +
  ylab("Percent of capture") + 
  xlab("") + ggtitle("") + 
    scale_y_continuous(breaks=seq(0,50,10)) +
  scale_color_brewer(palette = "Dark2", name = "Protocol") +
  scale_shape(name="Depletion") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

p2
```

## Donor shape, protocol colour

Reverse the plot annotation by having shape be the donor and colour the protocol.  
I find this less clear.

```{r per_anno_invert}
p3 <- ggplot(cellProp_srt, 
             aes(x = predicted.celltype.l1, y= percent, colour = Protocol, shape = mul_Donor)) + 
  geom_jitter(size=3, width = 0.15, alpha=0.75) +
  ylab("Percent of capture") + 
  xlab("") + ggtitle("") + 
    scale_y_continuous(breaks=seq(0,50,10)) +
  scale_color_brewer(palette = "Set1", name = "Protocol") +
  scale_shape(name="Donor") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

p3
```

* B cells (5-10%)
* CD4+ T cells (25-60%)
* CD8+ T cells (5-30%)
* NK cells (10-30%)
* Monocytes (5-10%)
* Dendritic cells (1-2%)
* Progenitors (0.1-0.2%)

# Plot level 2 annotations

## Seurat map reference

```{r cal-prop_l2}
cellProp_srt2 <- tb %>% 
  filter(mul_Donor != "empty") %>% 
  filter(mul_Donor != "multiplet") %>% 
  filter(mul_Donor != "unassigned") %>% 
  select(cell_id, predicted.celltype.l2, 
         Protocol, mul_Donor) %>%
  dplyr::count(predicted.celltype.l2, Protocol, mul_Donor) %>%
  group_by(Protocol, mul_Donor) %>% 
  mutate(percent = n / sum(n) * 100) %>%
  arrange(desc(percent))

cellProp_srt2
```

## Plot level 2 annotation

```{r per_anno_srt2, fig.cap="Level 2 immune subset proportion from each oligo tag multiplexing protocol"}
p4 <- ggplot(cellProp_srt2, 
             aes(x = predicted.celltype.l2, y= percent, shape = Protocol, colour = mul_Donor)) + 
  geom_jitter(size=3, width = 0.15, alpha=0.75) +
  ylab("Percent of capture") + 
  xlab("") + ggtitle("") + 
    scale_y_continuous(breaks=seq(0,50,10)) +
  scale_color_brewer(palette = "Dark2", name = "Donor") +
  scale_shape(name="Protocol") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size=12))

p4
```

# Conclusion

Seurat map reference matches the expected frequencies of PBMCs closely.  
There is not much difference between cell type annotations at least visually.

## Session Info

```{r}
sessionInfo()
```
