---
title: "Compare Tag and SNP assignments"
description: "Statistically compare tag oligos and SNP assignments"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6,
                      fig.path='../figures/tag_snp/',
                      layout="l-body-outset")
knitr::opts_knit$set(root.dir = here::here())
library(here)
library(tidyverse)
library(knitr)
library(edgeR)
library(patchwork)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Compare Tag and SNP assignments

## Aim

Statistically compare SNP and oligo tag assignments and see if there is a difference in proportions.

## Recap
The purpose of this experiment was to compare the performance of 3 different cell multiplexing reagents in PBMCs.  
To obtain the ground truth of performance, the genotypes may be used to demultiplex donors.
In this experiment there are 4 donors. The cell multiplexing was done in technical duplicate.  

# Read data

Homotypic doublets are removed from this tibble

```{r}
tb <- read_csv("/stornext/Projects/score/Analyses/R010_multiplexing/Matrices/PBMC/vireo/pbmc_calls_rmHomoDoublets.csv.gz")

tb <- tb %>%
  select(cell_id, Protocol, vir_Donor, mul_Assignment, mul_Donor, chr_bff_cluster)

tb$Protocol <- recode(tb$Protocol,
                       HTO = "Hashtag Ab",
                       LMO = "MULTI-Seq",
                       CMO = "CellPlex")

homo_doublets <- read_csv("/stornext/Projects/score/Analyses/R010_multiplexing/Matrices/PBMC/vireo/pbmc_homo_doublets.csv")

remove_cells <- paste(homo_doublets$cell_barcode, homo_doublets$Protocol, sep = "_")

```


```{r}
ocm <- read_csv("/stornext/Projects/score/Analyses/R010_multiplexing/Matrices/PBMC/vireo/ocm/oca_cm.csv")

ocm$Protocol <- recode(ocm$Protocol,
                       HTO = "Hashtag Ab",
                       LMO = "MULTI-Seq",
                       CMO = "CellPlex")

ocm
```

# Statistical tests {.tabset}

## CellPlex

Based on Peter Hickey's advice this statstical test is for a difference in abundances in response to treatment is explored in the  [OSCA book](https://bioconductor.org/books/3.16/OSCA.multisample/differential-abundance.html).

Our DA analysis will again be performed with the edgeR package. This allows us to take advantage of the NB GLM methods to model overdispersed count data in the presence of limited replication - except that the counts are not of reads per gene, but of cells per label (Lun, Richard, and Marioni 2017). The aim is to share information across labels to improve our estimates of the biological variability in cell abundance between replicates.

```{r}
# Create the table for stats test. Need the mean of the observations where the replicate is the PBMC donors
current <- "CellPlex"
abundances <- ocm

abundances <- abundances %>%
  filter(Protocol == current) %>% 
  select(-percent) %>%
  select(-Protocol) %>%
  pivot_wider(names_from = vir_Donor, values_from = n)
abundances <- as.data.frame(abundances)

# Attaching some column metadata.
# The extra.info data frame should have columns of the form
# sample, group
extra.info <- abundances[,1]
counts <- t( abundances[,c(2:7)])
y.ab <- DGEList(counts = counts, samples=extra.info)
y.ab
```

Unlike DE analyses, we do not perform an additional normalization step with calcNormFactors(). This means that we are only normalizing based on the “library size”, i.e., the total number of cells in each sample. Any changes we detect between conditions will subsequently represent differences in the proportion of cells in each cluster.

We formulate the design matrix with a blocking factor for the batch of origin for each sample and an additive term for the td-Tomato status (i.e., injection effect). Here, the log-fold change in our model refers to the change in cell abundance after injection, rather than the change in gene expression.

```{r}
design <- model.matrix(~samples, data =  y.ab$samples)
y.ab <- estimateDisp(y.ab, design, trend="none")
summary(y.ab$common.dispersion)
fit.ab <- glmQLFit(y.ab, design, robust=TRUE, abundance.trend=FALSE)
```

### Session Info

```{r}
sessionInfo()
```
