# Manuscript_Sample-Multiplexing-Oligo-Comparison
Manuscript: A Risk-reward Examination of Sample Multiplexing Reagents for Single Cell RNA-Seq

## Abstract

Single-cell RNA sequencing (scRNA-Seq) has emerged as a powerful tool for understanding cellular heterogeneity and function. However, the choice of sample multiplexing reagents can impact data quality and experimental outcomes. In this study, we compared various multiplexing reagents, including MULTI-Seq, Hashtag antibody, and CellPlex, across diverse sample types such as human peripheral blood mononuclear cells (PBMCs), mouse embryonic brain and patient-derived xenografts (PDXs).
We found that all multiplexing reagents worked well in cell types robust to \textit{ex vivo} manipulation but suffered from signal-to-noise issues in more delicate sample types. We compared multiple cell demultiplexing algorithms which differed in performance depending on data quality. We find that minor improvements to laboratory workflows such as titration and rapid processing are critical to optimal performance.
We also compared the performance of fixed scRNA-Seq kits and highlight the advantages of the Parse Bioscience kit for fragile samples. Highly multiplexed scRNA-Seq experiments require more sequencing resources, therefore we evaluated CRISPR-based destruction of non-informative genes to enhance sequencing value. Our comprehensive analysis provides insights into the selection of appropriate sample multiplexing reagents and protocols for scRNA-Seq experiments,  facilitating more accurate and cost-effective studies.

## Data availability

The count matrices and metadata are available as SingleCellExperiment objects at DOI: 10.5281/zenodo.8031078 https://zenodo.org/record/8031079

Single-cell | RNA-seq | Sample Multiplexing | Fixed | CRISPRclean
Correspondence: brown.d@wehi.edu.au
