---
title: "QC dimred Donor1 = PDX117"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='../figures/donor1/qc_dimred/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
suppressPackageStartupMessages({
  library(tidyverse)
  library(patchwork)
  library(scuttle)
  library(scater)
  library(scran)
  library(viridis)
  })

source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

Quality control and dimension reduction for donor1 which is PDX117

# Recap

When I extracted nuclei and fixed with the 10x Genomics kit the sample became increasingly clumped and ended up blocking the microfluidics chip. Fix with the Parse Bio version 2 kit. 

Additionally try a 10x Genomics protocol where I fix tumor pieces first then extract nuclei.

## This sample
Run a more challenging sample = Ovarian carcinosarcome PDX nuclei.  
Four PDXs were processed with 10x Genomics nuclei isolation kit. A capture was immdiately performed which is the unlabelled.  
Three more aliquots were labelled in singlicate with cell multiplexing reagents and captured later.

## Comparisons

* Fresh unlabelled nuclei after nuclei isolation with 10x Genomics kit

# Read SCE

From the **../3B_manual_annotation** notebook

```{r recap_umap_donor, fig.asp=1}
sce <- readRDS("/vast/scratch/users/brown.d/S000322/S000322.cluster.rds")[["Unlabelled"]]
sce$pool_cluster <- sce$cluster

umap_orig <- plotReducedDim(sce, dimred="UMAP", colour_by="donor_id")
umap_orig_cluster <- plotReducedDim(sce, dimred="UMAP", colour_by="pool_cluster")

umap_orig / umap_orig_cluster
```

Recap in the existing UMAP which I will later overwrite.  

I may have to remove these cells which are far removed from the main cluster of donor 1 cells.

```{r umap_focus_donor1}
sce <- sce[,sce$donor_id == "donor1"]
plotReducedDim(sce, dimred="UMAP", colour_by="donor_id") 
```

# Quality control

## Identifying outliers by each metric

Sine a quarter of the cells have been removed in subsetting this sample perform the low quality cell removal again.

Include removal of cells with too high library size. There is a group of cells with extremely high library size which are likely cell doublets or clumps. This was a very sticky sample.

```{r}
sce_orig <- sce

mito_drop <- isOutlier(
    metric = sce$subsets_Mito_percent, 
    nmads = 3, type = "higher")
```

Measure how many cells dropped.

```{r outlier_detect}
qc_df <- colData(sce)
reasons <- as.matrix(perCellQCFilters(
    x=qc_df, 
    sub.fields="subsets_Mito_percent"))

discard <- reasons
discard <- discard[,4]

sce <- sce[,!discard]

colnames(reasons)
colSums(reasons)
```

Summarise how many cells left.

216 cells have been dropped. Very few cells are dropped for mitochondrial percentage. Most are due to having a library size of greater than 3 MADs. I set this cutoff because I am worried about the clumpiness of the sample.

```{r}
cell_drop_tb <- rbind(
  cbind(length(colnames(sce_orig)), length(colnames(sce)))
)

colnames(cell_drop_tb) <- c("Before_Filter", "After_filter")
cell_drop_tb
```

## Filtering out low-abundance genes

Low-abundance genes are problematic as zero or near-zero counts do not contain much information for reliable statistical inference.  
We drop another ~5000 genes combined to the overall Parse object

```{r average_count_p}
ave_counts <- calculateAverage(sce)
to_keep <- ave_counts > 0
sce <- sce[to_keep, ]

hist(
  x = log10(ave_counts), 
  breaks = 100, 
  main = "10x Genomics v3.1 donor 1", 
  col = "grey",
  xlab = expression(Log[10] ~ "average count"))
```


```{r filter_gene}
# Summarise how many genes left
gene_drop_tb <- rbind(
  cbind(length(rownames(sce_orig)), length(rownames(sce)))
)

colnames(gene_drop_tb) <- c("Before_Filter", "After_filter")
gene_drop_tb
```

# GEX Normalization

## Using the deconvolution method to deal with zero counts

We pool counts from many cells to increase the count size for accurate size factor estimation (A. T. Lun, Bach, and Marioni 2016). Pool-based size factors are then ‘deconvolved’ into cell-based factors for cell-specific normalization. This removes scaling biases associated with cell-specific differences in capture efficiency, sequencing depth and composition biases.

```{r normalise}
set.seed(666)

clusters <- quickCluster(sce)
sce <- computeSumFactors(sce, clusters = clusters, min.mean = 0.1)
sce <- logNormCounts(sce)
print(summary(sizeFactors(sce)))

```

# Selecting highly variable genes (HVGs)

## Fit variance


```{r var_genes_p}
var_fit <- modelGeneVarByPoisson(sce)
hvg <- getTopHVGs(var_fit, var.threshold = 0)
```

```{r}
p1 <- plotExpression(object = sce, features = hvg[1:10])
p1
```

# Dimension reduction {.tabset}

## PCA

```{r PCA}
set.seed(666)

sce <- denoisePCA(sce, var_fit, subset.row = hvg)
```

### Number of PCA dimensions to retain

Based on these percent variance plots retain 30 PCA dimensions.  
The variation is relatively low with PC1 being 4% only.

```{r}
# Percentage of variance explained is tucked away in the attributes.
percent.var <- attr(reducedDim(sce), "percentVar")

plot(percent.var, xlab="PC", ylab="Variance explained (%)")
```

### Visualise PCA

Very little variation present in the dataset.

```{r pca_plts}
pca_plt <- plotReducedDim(sce, dimred="PCA", colour_by = "detected")

pca_plt + theme_Publication()
```

## Run UMAP

```{r umap}
set.seed(100)
sce <- runUMAP(sce, dimred = "PCA", n_dimred=30)
```

# Clustering

We build a shared nearest neighbour graph and use the Louvain algorithm to identify clusters. We build the graph using the principal components.

## Initial clustering

Smaller 'k' usually yields finer clusters.

```{r clustering_ks}
set.seed(911)

snn_gr_cmo <- buildSNNGraph(sce, use.dimred = "PCA", k = 5) 
clust.5 <- igraph::cluster_louvain(snn_gr_cmo)
table(clust.5$membership)

snn_gr_cmo <- buildSNNGraph(sce, use.dimred = "PCA", k = 10) 
clust.10 <- igraph::cluster_louvain(snn_gr_cmo)
table(clust.10$membership)

snn_gr_cmo <- buildSNNGraph(sce, use.dimred = "PCA", k = 15) 
clust.15 <- igraph::cluster_louvain(snn_gr_cmo)
table(clust.15$membership)

snn_gr_cmo <- buildSNNGraph(sce, use.dimred = "PCA", k = 20) 
clust.20 <- igraph::cluster_louvain(snn_gr_cmo)
table(clust.20$membership)
```

Based on trying k from 5 - 20 I use k of 15 as all clusters have more than 100 cells. k of 10 does reflect what thee UMAP looks like but cluster 8, the group at the bottom center of the plot has 45 cells.

```{r k_10}
set.seed(911)

snn_gr_cmo <- buildSNNGraph(sce, use.dimred = "PCA", k = 10) 
clusters <- igraph::cluster_louvain(snn_gr_cmo)
sce$cluster <- factor(clusters$membership)
plotReducedDim(sce, dimred="UMAP", colour_by="cluster") + ggtitle("k=10") + theme_Publication()
```

Of course the true test will be when ascribing biological meaning.

```{r clustering_k15}
set.seed(911)

snn_gr_cmo <- buildSNNGraph(sce, use.dimred = "PCA", k = 15) 
clusters <- igraph::cluster_louvain(snn_gr_cmo)
sce$cluster <- factor(clusters$membership)

table(clusters$membership)
```

# Visualise UMAP {.tabset}

## Cluster number only

```{r umap_cluster}
umap_cluster <- plotReducedDim(sce, dimred="UMAP", colour_by="cluster") + theme_Publication()

umap_cluster
```

## Original cluster number when I had all donors in the SCE

The original cluster number which contained the majority of donor 1 cells was cluster 2. 
Cluster 7 is an outlier cluster in the original pooled sample but no evidence to suggest it is technical yet.  
Cluster 8 is a satielle of cluster 2 in the original pooled clustering, this could have a biological basis.

```{r umap_orig_cluster, fig.asp=1}
umap_orig_cluster <- plotReducedDim(sce, dimred="UMAP", colour_by="pool_cluster") + theme_Publication()

umap_cluster / umap_orig_cluster
```

## With QC parameters

Not a great deal of structure but also no obvious distortions to do with libarry size or mitochondrial gene percentage

```{r umap_plts, fig.dim=c(12,12)}
umap_sum <- plotReducedDim(sce, dimred="UMAP", colour_by="sum") + theme_Publication()
umap_detected <- plotReducedDim(sce, dimred="UMAP", colour_by="detected") + theme_Publication()
umap_cluster <- plotReducedDim(sce, dimred="UMAP", colour_by="cluster") + theme_Publication()
umap_mito <- plotReducedDim(sce, dimred="UMAP", colour_by="subsets_Mito_percent") + theme_Publication()


umap_cluster + umap_sum + umap_detected + umap_mito + plot_layout(ncol=2)
```

# Conclusion

Investigate marker genes in a subsequent notebook.

```{r saveSCEs}
saveRDS(sce,
  "/vast/scratch/users/brown.d/S000322/S000322_unlabelled_donor1-cluster.rds"
)
 
```

## Session Info

```{r}
sessionInfo()
```
