---
title: "PDX117 marker gene detection"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.8, 
                      fig.path='../figures/donor1/marker_genes/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
suppressPackageStartupMessages({
  library(tidyverse)
  library(patchwork)
  library(scuttle)
  library(scater)
  library(scran)
  library(viridis)
  })

source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

Perform differential expression test on clusters to find marker genes.

# Aim

Quality control and dimension reduction for donor1 which is PDX117

# Recap

When I extracted nuclei and fixed with the 10x Genomics kit the sample became increasingly clumped and ended up blocking the microfluidics chip. Fix with the Parse Bio version 2 kit. 

Additionally try a 10x Genomics protocol where I fix tumor pieces first then extract nuclei.

## This sample
Run a more challenging sample = Ovarian carcinosarcome PDX nuclei.  
Four PDXs were processed with 10x Genomics nuclei isolation kit. A capture was immdiately performed which is the unlabelled.  
Three more aliquots were labelled in singlicate with cell multiplexing reagents and captured later.

## Comparisons

* Fresh unlabelled nuclei after nuclei isolation with 10x Genomics kit

# Read SCE

From the **1A_qc_dimred_donor1** notebook.

```{r loadData}
sce <- readRDS(
 "/vast/scratch/users/brown.d/S000322/S000322_unlabelled_donor1-cluster.rds"
)
```

## Recap clusters in a PCA

```{r pca_cluster, fig.asp=0.8}
pca_plt <- plotReducedDim(sce, dimred="PCA", colour_by="cluster") + theme_Publication(base_size=22)

pca_plt
```

## Recap clusters in a UMAP

Cluster 3 is the very distinct one.  
Cluster 4 also of interest.

```{r umap_cluster, fig.asp=0.8}
umap_plt <- plotReducedDim(sce, dimred="UMAP", colour_by="cluster") + theme_Publication(base_size=22)

umap_plt
```

## Check high expessed genes

Lots of mitochondrial genes.

```{r}
her <- plotHighestExprs(sce, n=15, colour_cells_by = "cluster") + theme_Publication()
her
```

# Marker gene detection

For each cluster, we rank candidate markers based on one of these effect size summaries.  
Use pairwise comparisons rather than comparing each cluster to the average of all other cells.

```{r}
marker.info <- scoreMarkers(sce, groups=sce$cluster)
```

Compare each pair of clusters and compute scores quantifying the differences in the expression distributions between clusters.  
The scores for all pairwise comparisons involving a particular cluster are then consolidated into a single DataFrame for that cluster.

## Iterate over each cluster

```{r}
markers <- list()

for (i in 1:length(marker.info)) {
    chosen <- marker.info[[i]]
    ordered <- chosen[order(chosen$mean.AUC, decreasing=TRUE),]

    chosen <- chosen[order(chosen$mean.AUC, decreasing=TRUE),]
    markers[[i]] <- as.data.frame(chosen[c(1:5),])
}

names(markers) <- 1:length(markers)
```

## Mung the list of marker genes

Convert to a summarised and usable form for all clusters

```{r}
marker <- do.call(rbind, markers)
marker$gene <- row.names(marker)

marker <- as_tibble(marker) %>% 
  select(gene,self.average, mean.logFC.cohen, median.logFC.cohen, 
         mean.AUC, median.AUC, 
         mean.logFC.detected, median.logFC.detected) %>%
  separate(gene, c("cluster", "gene.symbol"), sep = 2) %>% 
  filter(mean.logFC.cohen > 0.50) %>% 
  filter(self.average > 1.5)

# Fix up the cluster number
marker$cluster <- gsub("\\.", "",marker$cluster)
marker <- marker[!duplicated(marker$gene.symbol),]

marker
```

Add some custom markers in there based on EMT

```{r}
custom_marker <- c("MKI67", "VIM", "EPCAM", "DACH1", "MET")
```

## Visualise results {.tabset}

### Dot plot

* Custer 3 the distinct one has a lot of MALAT1. Normally a telltale sign of low quality cells where only the high expressing gene remains.
  + MIR99AHG also highest here

```{r dot_marker, fig.width=8, fig.height=20}
dot <- plotDots(sce, 
         features = c(marker$gene.symbol, custom_marker), 
         group="cluster", color=plasma(100)) +
  labs(title = "", y = "", x = "Cluster", color = "Log counts", size="Proportion of cells") +
  theme_Publication(base_size = 14)

dot
```

Write a subset of the most discriminatory markers for further investigation.

```{r}
marker_subset <- c(
  "DACH1", "VIM", "NSD2", "KCNQ1OT1", "MALAT1",
  "CNTNAP2", "SDK1", "NRXN1", "MT-ATP6"
)
```

### Expression plots

```{r expression, fig.dim=c(8,8)}
exprs <- plotExpression(sce, features=marker_subset, 
    x="cluster", colour_by="cluster", ncol = 3) +
  theme_Publication(base_size = 14)

exprs
```

### UMAP MALAT1

```{r umap_malat1, fig.asp=0.8}
umap_malat <- plotReducedDim(sce, dimred="UMAP", colour_by="MALAT1") + theme_Publication(base_size=22)

umap_malat
```

### UMAP SDK1

```{r umap_sdk1, fig.asp=1}
umap_sdk1 <- plotReducedDim(sce, dimred="UMAP", colour_by="SDK1") + theme_Publication(base_size=18)

umap_plt / umap_sdk1
```


# Conclusion

* Cluster 3 is the very distinct one and is marked by high MALAT1 expression. 
* Cluster 4 is more interesting and is marked by NRXN1, CNTNAP2 and SDK1

## Session Info

```{r}
sessionInfo()
```
