---
title: "CRISPRclean depletion performance"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/PBMC/crisprclean_performance/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
library(DropletUtils)
library(tidyverse)
library(patchwork)
library(scuttle)
library(scater)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

Evaluate jumpcode CRISPRclean single-cell boost kit. This adds more content on top of ribosomal and mitochondrial genes.  
Including Pseudogenes and non variablee genes

## This sample

I restrict my analysis to R010_MULTI-Seq_GEX in PBMCs. This sample had the best signal to noise.

I sequenced this particular sample much deeper as part of NN265.  
[ELN entry](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/MzQ3LjF8NjgxOTcvMjY3LTE0NzgvVHJlZU5vZGUvMjMwMTMwNzYyN3w4ODEuMDk5OTk5OTk5OTk5OQ==)

## Comparisons

Based on the CRISPRclean advertising I downsampled the Jumpcode sample to 50% of the reads for untreated sample.

# Load Data

Find the paths to the cell ranger output

```{r loadData}
sces <- readRDS(here::here(
  "CRISPRclean/data/Jumpcode/SCEs", "PBMC_CRISPR_colData.sces.rds"
))

tb <- read_csv(here::here(
  "CRISPRclean/data/Jumpcode/Matrices", "PBMC_CRISPR_colData.csv.gz"
))

tb$Workflow <- paste(tb$Protocol, tb$Sequencing, sep="_")
```

For this visualisation it is not that informative to have the 50% downsampled sample because as expected all the metrics are lower. We remove it.

```{r}
tb_orig <- tb
tb <- tb %>% 
  filter(Sequencing == "Full")
```

Load the jumpcode panel

```{r}
jpc <- read_csv(here::here(
  "CRISPRclean/metadata/Jumpcode", "CRISPRclean_Single_Cell_RNA_Boost_Kit_depletion_content_set.csv"
))
```

# Visualizing the QC metrics {.tabset}

Curious that slightly fewer cells were called by cellranger 7 for the CRIPSRclean boost sample.

```{r}
tb %>% 
  dplyr::count(Workflow)

```

## Library size

The untreated sample when mapped to full reference transcriptome that includes mitochondrial and ribosomal genes has the expected larger library size.

Downsampling has the effect of compressing the 2 modes of the full sequencing dataset.

```{r libSize_p, fig.cap="Library size in deduplicated UMIs"}
plt1 <- ggplot(tb, 
             aes(x = Protocol, y= sum+1, fill = Protocol)) + 
  geom_boxplot(show.legend = FALSE) + 
  ylab("Library Size (UMIs)") + 
  xlab("") +
  scale_y_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "l") +
  scale_fill_brewer(palette = "Set1") 

plt1
```

## Detected genes

Slightly more genes detected in Untreated in line with UMI counts.
Those outliers remain when I remove doublets.

```{r detected_genes_p, fig.cap="Number of detected genes"}
plt2 <- ggplot(tb, 
             aes(x = Protocol, y= sum+1, fill = Protocol)) +  
  geom_boxplot(show.legend = FALSE) + 
  ylab("Detected genes") + 
  xlab("") +
  scale_fill_brewer(palette = "Set1") 

plt2
```

## Mitochondiral percentage

The depletion has worked as intended.

```{r mito_per_p, fig.cap="Mitochondrial percentage of total counts"}
plt3 <- ggplot(tb, 
             aes(x = Protocol, y= subsets_Mito_percent, fill = Sequencing)) + 
  geom_boxplot(show.legend = FALSE, outlier.size = 0.2, outlier.alpha = 0.25) + 
  ylab("Mitochondrial\n percentage") + 
  xlab("") +
  scale_fill_brewer(palette = "Set1")

plt3
```

## Genes in jumpcode panel

* The median percent of counts in CRISPRclean panel is `r median(tb$subsets_Jumpcode_percent[tb$Workflow == "CRISPRclean Boost_Whole"])`
* The median percent of counts in CRISPRcleean panel is `r median(tb$subsets_Jumpcode_percent[tb$Workflow == "Undepleted_Whole"])`


```{r, eval=FALSE}
median(tb$subsets_Jumpcode_percent[tb$Protocol == "CRISPRclean Boost"])
median(tb$subsets_Jumpcode_percent[tb$Protocol == "Undepleted"])
```

Plot both the dataset mapped to the masked reference and the normal reference.

```{r panel_count_p, fig.cap="Library size of multiplexing capture sequencing"}
plt4 <- ggplot(tb, 
             aes(x = Protocol, y= subsets_Jumpcode_percent, fill = Protocol)) + 
  geom_boxplot(show.legend = FALSE, outlier.size = 0.2, outlier.alpha = 0.25) + 
  ylab("Jumpcode panel\n percentage") + 
  xlab("") +
  scale_y_continuous(breaks=seq(0,100,10), limits = c(0,60)) +
  scale_fill_brewer(palette = "Set1") +
  theme_Publication(base_size=20)

plt4
```

# Put QC plots together

```{r qc_summary_p, fig.cap="Summary of QC metrics", fig.asp=1}
plt1 +  plt2 + plt3 + plt4 + 
  plot_layout(guides = "collect", ncol = 2)
```

# UMI counts vs genes {.tabset}

The left shift means umis are spreead across more diverse genes.

## Log scale

Clear improvement

```{r umi_genes_log}
plt5 <- ggplot(tb, 
             aes(x = sum, y= detected, colour = Protocol)) + 
  geom_point(alpha = 0.25, size=0.5) + 
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1))) +
  xlab("Library Size (UMIs)") + 
  ylab("Detected Genes") +
  scale_x_continuous(trans='log10') +
  scale_y_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "bl") +
  scale_colour_brewer(palette = "Set1") +
  theme_Publication(base_size=20)

plt5
```

## Natural scale

```{r umi_genes_nat}
plt5 <- ggplot(tb, 
             aes(x = sum, y= detected, colour = Protocol)) + 
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1))) +
  geom_point(alpha = 0.25, size=0.5) + 
  xlab("Library Size (UMIs)") + 
  ylab("Detected Genes") +
  scale_colour_brewer(palette = "Set1")

plt5
```

## Natural scale downsampled crisprclean

```{r umi_genes_nat_ds}
plt6 <- ggplot(tb_orig[!tb_orig$Workflow %in% "CRISPRclean Boost_Full",], 
             aes(x = sum, y= detected, colour = Workflow)) + 
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1))) +
  geom_point(alpha = 0.25, size=0.5) + 
  xlab("Library Size (UMIs)") + 
  ylab("Detected Genes") +
  scale_colour_brewer(palette = "Set1")

plt6
```

# Look at high expressing genes {.tabset}

```{r}
names(sces)
sces_whole <- sces[c(1,2)] 
```

## Combine 1 figure

Best to calculate and plot this separately. Large size

```{r highExpr_big}
sce_jmp <- sces[["Jumpcode_full"]]
sce_utd <- sces[["Untreated"]]

p1 <- plotHighestExprs(sce_utd, n = 15) + ggtitle("Undepleted") +
  theme_Publication(base_size=16)
p2 <- plotHighestExprs(sce_jmp, n = 15) + ggtitle("CRISPRclean") +
  theme_Publication(base_size=16)

p1 + p2 + plot_layout(guides = "collect")
```

Small size

```{r highExpr_small}
p3 <- plotHighestExprs(sce_utd, n = 20) + ggtitle("Undepleted") +
  theme_Publication()
p4 <- plotHighestExprs(sce_jmp, n = 20) + ggtitle("CRISPRclean") +
  theme_Publication()

p3 + p4 + plot_layout(guides = "collect")
```

## Untreated

```{r untreated_highExpr}
p1
```

## Depleted

```{r jumpcode_highExpr}
p2
```

## Session Info

```{r}
sessionInfo()
```
