---
title: "GEX compare cell anotations CRIPSRclean sc boost"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/compare_anno/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())
library(patchwork)
library(scran)
library(scater)
library(scuttle)
library(knitr)
library(edgeR)
library(DropletUtils)
library(tidyverse)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

Evaluate jumpcode CRISPRclean single-cell boost kit. This adds more content on top of ribosomal and mitochondrial genes.  
Including Pseudogenes and non variable genes

## This sample

I restrict my analysis to R010_MULTI-Seq_GEX in PBMCs. This sample had the best signal to noise.

I sequenced this particular sample much deeper as part of NN265.  
[ELN entry](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/MzQ3LjF8NjgxOTcvMjY3LTE0NzgvVHJlZU5vZGUvMjMwMTMwNzYyN3w4ODEuMDk5OTk5OTk5OTk5OQ==)

## Comparisons

I mapped both the undepleted and CRISPRclean depleted datasets to GRCh38 reference provided by cellranger and the masked reference provided by jumpcode. Jumpcode says only the masked reference will highlight the differences in depleted.

I follow the vignette of [Seurat reference map](https://satijalab.org/seurat/articles/multimodal_reference_mapping.html)

## Aim

Visualise the cell type annotations from Azimuth and compute proportions. 

# Load Data

Read in data. Was generated in 1C_Seurat_ref_map notebook.

```{r loadData}
sces <- readRDS(here::here(
  "CRISPRclean/data/Jumpcode/SCEs", "PBMC_CRISPR_dimred.sces.rds"
))
names(sces)
```

# Clustering

We build a shared nearest neighbour graph and use the Louvain algorithm to identify clusters. We build the graph using the principal components.

## Initial clustering

Smaller 'k' usually yields finer clusters.
The number of clusters stabilise aroundd k=10

```{r cluster_1}
set.seed(911)

for (n in names(sces)) {
  current <- sces[[n]]
  snn_gr <- buildSNNGraph(current, use.dimred = "PCA", k = 10) 
  clusters <- igraph::cluster_louvain(snn_gr)
  current$cluster <- factor(clusters$membership)
  sces[[n]] <- current
}

```

# Visualise in dimension reduction {.tabset}

## Cluster identity

```{r vis_cluster, fig.height=4, fig.width=10}
umap_cluster <- list()

for (n in names(sces)) {
  current <- sces[[n]]
  umap_cluster[[n]] <- plotUMAP(sces[[n]], colour_by="cluster") +
    ggtitle(n) +
    theme(legend.text=element_text(size=6)) +
    theme_Publication()
}

umap_cluster[[1]] + umap_cluster[[2]] + 
  umap_cluster[[3]]
```

## Cell type annotation Seurat map reference

```{r vis_anno, fig.height=3, fig.width=12}
umap_cellType1 <- list()

for (n in names(sces)) {
  current <- sces[[n]]
  umap_cellType1[[n]] <- plotUMAP(sces[[n]], colour_by="predicted.celltype.l1") +
    ggtitle(n) +
    theme_Publication()
}

umap_cellType1[[1]] + umap_cellType1[[2]] + umap_cellType1[[3]] +
  plot_layout(guides = "collect")
```

Plot separately so I can save as png.

```{r vis_anno_untreated}
umap_cellType1[[1]]
```

```{r vis_anno_jumpcodeFull}
umap_cellType1[[2]]
```

```{r vis_anno_jumpcodeHalf}
umap_cellType1[[3]]
```

## Represent cluster ID and cell type annotation on the sample visualisation

```{r vis_cluster_anno, fig.height=4, fig.width=10}
(umap_cluster[[1]] + umap_cluster[[2]] +  umap_cluster[[3]]) /
  (umap_cellType1[[1]] + umap_cellType1[[2]] + umap_cellType1[[3]] +
     plot_layout(guides = "collect"))
```

## Library size

```{r vis_libsize, fig.height=4, fig.width=10}
umap_plt <- list()

for (n in names(sces)) {
  current <- sces[[n]]
  umap_plt[[n]] <- plotUMAP(sces[[n]], colour_by="sum") +
  theme(legend.text=element_text(size=10)) +
    ggtitle(n) +
    theme_Publication()
}

umap_plt[[1]] + umap_plt[[2]] + 
  umap_plt[[3]] +
  plot_layout(guides = "collect")
```

## Mitochondrial percent

```{r vis_mito, fig.height=4, fig.width=10}
umap_plt <- list()

for (n in names(sces)) {
  current <- sces[[n]]
  umap_plt[[n]] <- plotUMAP(sces[[n]], colour_by="subsets_Mito_percent") +
  theme(legend.text=element_text(size=10)) +
    ggtitle(n) +
    theme_Publication()
}

umap_plt[[1]] + umap_plt[[2]] + 
  umap_plt[[3]] +
  plot_layout(guides = "collect")
```

# Closer look CD14 and CD16 monocytes

In the Jumpcode [python documation](https://github.com/JumpcodeGenomics/Single-Cell-RNA-Boost/blob/main/Final_GitHub_SP.ipynb) they write the following text:

1. CD14+ monocytes are split into two clusters in depleted.
2. DCs were not clustered separately from DCs (even though distant in UMAP) in the control. But were in the depleted
3. Depletion revealed a MKI67+ and CDK4/CD34+ progenitor populations
4. CD16+ monocytes are more transcriptionally distinct in the depleted cells

In my analysis the undepleted has a better separation of CD16 monocyctes to my visual inspection.

```{r cd14_cd16, fig.height=4, fig.width=10}
umap_cd16 <- list()
umap_cd14 <- list()

for (n in names(sces)) {
  umap_cd16[[n]] <- plotUMAP(sces[[n]], colour_by="FCGR3A") +
  theme(legend.text=element_text(size=10)) +
    ggtitle(n) +
    theme_Publication()
}

for (n in names(sces)) {
  umap_cd14[[n]] <- plotUMAP(sces[[n]], colour_by="CD14") +
  theme(legend.text=element_text(size=10)) +
    ggtitle(n) + 
    theme_Publication()
}

# Keep the legend for only the last sample for the plot to look compact
(umap_cd16[[1]] + theme(legend.position = "none") + 
    umap_cd16[[2]] + theme(legend.position = "none") + 
    umap_cd16[[3]]) /
(umap_cd14[[1]] + theme(legend.position = "none") + 
   umap_cd14[[2]] + theme(legend.position = "none") + 
   umap_cd14[[3]])

```

# Calculate percentage of cell type

Convert the metadata to tibble.

```{r}
tb <- as_tibble(rbind(
  colData(sces[[1]]),
  colData(sces[[2]]),
  colData(sces[[3]])
))

tb <- tb %>% 
  select(-Sample)

tb <- tb %>% rename("Broad cell type" = "predicted.celltype.l1",
                    "Finer cell type" = "predicted.celltype.l2"
                    )
```

Calculate proportions

```{r cal-prop}
cellProp_prot <- tb %>% 
  dplyr::rename("Donor" = mul_Donor) %>% 
  select(cluster, Donor, `Broad cell type`, Protocol, Sequencing) %>%
  dplyr::count(`Broad cell type`, Donor, Protocol, Sequencing) %>%
  group_by(Donor, Protocol, Sequencing) %>%
  mutate(percent = n / sum(n) * 100) %>%
  unite("Workflow", Protocol:Sequencing, remove = FALSE, sep= "_") %>%
  arrange(desc(percent))

cellProp_prot %>% 
  head()
```

## Broad annotation

```{r per_anno_prop, fig.cap = "Within each capture (Protocol) the proportion of the different calls is reprersented as a percent of all cells", fig.asp=0.6}
p1 <- ggplot(cellProp_prot, 
             aes(x = `Broad cell type`, y= percent, colour = Workflow, shape=Donor)) + 
  geom_jitter(size=3, width = 0.15, alpha=0.75) +
  ylab("Percent of capture") + 
  xlab("") +
  scale_color_brewer(palette = "Dark2", name = "Depletion", label=c("CRISPRclean", "CRISPRclean 50% reads", "Undepleted")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) 

p1
```

## Finer annotation

```{r cal-prop-fine}
cellProp_fine_prot <- tb %>% 
  dplyr::rename("Donor" = mul_Donor) %>% 
  select(cluster, Donor, `Finer cell type`, Protocol, Sequencing) %>%
  dplyr::count(`Finer cell type`, Donor, Protocol, Sequencing) %>%
  group_by(Donor, Protocol, Sequencing) %>%
  mutate(percent = n / sum(n) * 100) %>%
  filter(percent >= 0.5 ) %>% 
  unite("Workflow", Protocol:Sequencing, remove = FALSE, sep= "_") %>%
  arrange(percent)

cellProp_fine_prot %>% 
  head()
```

```{r per_anno_fine, fig.cap = "Within each capture (Protocol) the proportion of the different calls is reprersented as a percent of all cells", fig.asp=0.6}
p2 <- ggplot(cellProp_fine_prot, 
             aes(x = `Finer cell type`, y= percent, colour = Workflow, shape=Donor)) + 
  geom_jitter(size=3, width = 0.15, alpha=0.75) +
  ylab("Percent of capture") +  
  xlab("") +
  scale_color_brewer(palette = "Dark2", name = "Depletion", label=c("CRISPRclean", "CRISPRclean 50% reads", "Undepleted")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) 

p2
```

# Statistical tests {.tabset}

## Broad cell type category

Based on Peter Hickey's advice this statstical test is for a difference in abundances in response to treatment is explored in the  [OSCA book](https://bioconductor.org/books/3.16/OSCA.multisample/differential-abundance.html).

Our DA analysis will again be performed with the edgeR package. This allows us to take advantage of the NB GLM methods to model overdispersed count data in the presence of limited replication - except that the counts are not of reads per gene, but of cells per label (Lun, Richard, and Marioni 2017). The aim is to share information across labels to improve our estimates of the biological variability in cell abundance between replicates.

```{r}
# Create the table for stats test. Need the mean of the observations where the replicate is the PBMC donors
abundances <- cellProp_prot %>% 
  filter(Sequencing == "Full") %>%
  select(Protocol, `Broad cell type`, Donor, n)

abundances <- abundances[,c(2:5)]
abundances <- abundances %>% 
  pivot_wider(names_from = `Broad cell type`, values_from = n)
abundances <- as.data.frame(abundances)

# Attaching some column metadata.
# The extra.info data frame should have columns of the form
# sample, group
extra.info <- abundances[,c(1,2)]
counts <- t( abundances[,c(3:10)])
y.ab <- DGEList(counts = counts, samples=extra.info)
y.ab

```

Unlike DE analyses, we do not perform an additional normalization step with calcNormFactors(). This means that we are only normalizing based on the “library size”, i.e., the total number of cells in each sample. Any changes we detect between conditions will subsequently represent differences in the proportion of cells in each cluster.

We formulate the design matrix with a blocking factor for the batch of origin for each sample and an additive term for the td-Tomato status (i.e., injection effect). Here, the log-fold change in our model refers to the change in cell abundance after injection, rather than the change in gene expression.

```{r}
design <- model.matrix(~factor(Donor) + factor(Protocol), y.ab$samples)
y.ab <- estimateDisp(y.ab, design, trend="none")
summary(y.ab$common.dispersion)
fit.ab <- glmQLFit(y.ab, design, robust=TRUE, abundance.trend=FALSE)
```

Perform the differential abundance test.  
As expected there is no significant difference in cell type abundance when comparing undepeleted and depleted samples.

```{r}
res <- glmQLFTest(fit.ab, coef=ncol(design))
summary(decideTests(res))
topTags(res, n = 5)
```

## Fine cell type category

```{r}
# Create the table for stats test. Need the mean of the observations where the replicate is the PBMC donors
abundances <- cellProp_fine_prot %>% 
  filter(Sequencing == "Full") %>%
  select(Protocol, `Finer cell type`, Donor, n)

abundances <- abundances[,c(2:5)]
abundances <- abundances %>% 
  pivot_wider(names_from = `Finer cell type`, values_from = n)
abundances <- as.data.frame(abundances)

# Need to recode NA
abundances[is.na(abundances)] <- 0

# Attaching some column metadata.
# The extra.info data frame should have columns of the form
# sample, group
extra.info <- abundances[,c(1,2)]
counts <- t( abundances[,c(3:10)])
y.ab <- DGEList(counts = counts, samples=extra.info)
y.ab

```

Unlike DE analyses, we do not perform an additional normalization step with calcNormFactors(). This means that we are only normalizing based on the “library size”, i.e., the total number of cells in each sample. Any changes we detect between conditions will subsequently represent differences in the proportion of cells in each cluster.

We formulate the design matrix with a blocking factor for the batch of origin for each sample and an additive term for the td-Tomato status (i.e., injection effect). Here, the log-fold change in our model refers to the change in cell abundance after injection, rather than the change in gene expression.

```{r}
design <- model.matrix(~factor(Donor) + factor(Protocol), y.ab$samples)
y.ab <- estimateDisp(y.ab, design, trend="none")
summary(y.ab$common.dispersion)
fit.ab <- glmQLFit(y.ab, design, robust=TRUE, abundance.trend=FALSE)
```

Perform the differential abundance test.  
As expected there is no significant difference in cell type abundance when comparing undepeleted and depleted samples.

```{r}
res <- glmQLFTest(fit.ab, coef=ncol(design))
summary(decideTests(res))
topTags(res, n = 5)
```

# Conclusion

* I do not find subclustering of CD14 and CD16 monocytes in the Jumopcode data like in their vignette
* There is no broad (level 1) difference in cell type abundance with CRISPRclean treatment
* There is no finer (level 2) difference in cell type abundance with CRISPRclean treatment

## Save SCE

```{r saveSCE}
saveRDS(sces, here::here(
  "CRISPRclean/data/Jumpcode/SCEs", "PBMC_CRISPR_clustered.sces.rds"
))
```


## Session Info

```{r}
sessionInfo()
```
