---
title: "Investigate which cells dropped jumpcode"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/cellsDrop/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())
library(scater)
library(scran)
library(scuttle)
library(edgeR)
library(patchwork)
library(knitr)
library(tidyverse)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

Evaluate jumpcode CRISPRclean single-cell boost kit. This adds more content on top of ribosomal and mitochondrial genes.  
Including Pseudogenes and non variablee genes.

In this notebook perform the DE testing between treated and untreated.

## This sample

I restrict my analysis to R010_MULTI-Seq_GEX in PBMCs. This sample had the best signal to noise.

I sequenced this particular sample much deeper as part of NN265.  
[ELN entry](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/MzQ3LjF8NjgxOTcvMjY3LTE0NzgvVHJlZU5vZGUvMjMwMTMwNzYyN3w4ODEuMDk5OTk5OTk5OTk5OQ==)

## This analysis

I will perform both the OSCA recomended guidelines for low quaality cell rremoval, mstly based on medin absolute deviation.  
Also the Jumpcode recommended [pipeline](https://github.com/JumpcodeGenomics/Single-Cell-RNA-Boost/blob/main/jumpcode_Rcode.md)

## Comparisons

I mapped both the undepleted and CRISPRclean depleted datasets to GRCh38 reference provided by cellranger and the masked reference provided by jumpcode. Jumpcode says only the masked reference will highlight the differences in depleted.

# Load Data

Read in data

```{r loadData}
sces <- readRDS(here::here(
  "CRISPRclean/data/Jumpcode/SCEs", "PBMC_CRISPR_colData.sces.rds"
))
names(sces)

tb <- read_csv(here::here(
  "CRISPRclean/data/Jumpcode/Matrices", "PBMC_CRISPR_colData.csv.gz"
))

sce_jmp <- sces[["Jumpcode_full"]]
sce_utd <- sces[["Untreated"]]
```

Load the jumpcode panel

```{r}
jpc <- read_csv(here::here(
  "CRISPRclean/metadata/Jumpcode", "CRISPRclean_Single_Cell_RNA_Boost_Kit_depletion_content_set.csv"
))
```

# Visualise library size veruses mito percent

There remains a relationship between library size and mitochondrial percentage indicative of low quality cells.

```{r libsize_mito-per, fig.cap="x-axis is log10"}
plt1 <- ggplot(tb, 
             aes(x = sum, y= subsets_Mito_percent, colour = Protocol)) + 
  geom_point(alpha = 0.2, size=0.75, show.legend = FALSE) + 
  xlab("Library Size (UMIs)") + 
  ylab("Mitochondrial percent") +
  scale_x_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "b") +
  scale_colour_brewer(palette = "Set1") +
  facet_wrap(~Protocol, scales = "free_y")

plt1
```

# GEX: Quality control

## More manual approach to outlier removal

I can summarise by cell between the 2 treatments and check by tibble if the same cell is removed or not.

* For sum.field, small outliers are detected. These are considered to represent low-quality cells that have not been insufficiently sequenced.
* For detected.field, small outliers are detected. These are considered to represent low-quality cells with low-complexity libraries. Detection is performed on the log-scale to adjust for a heavy right tail. 
* For each column specified by sub.fields, large outliers are detected. This aims to remove cells with high mitochondrial content, usually corresponding to damaged cells.

```{r}
filtered <- list()

for (n in names(sces)) {
    current <- sces[[n]]
    is.mito <- grep("^MT", row.names(current))
    filtered[[n]] <- quickPerCellQC(current, subsets=list(Mito=is.mito), sub.fields="subsets_Mito_percent", filter=FALSE)
    filtered[[n]] <- as_tibble(filtered[[n]])
}

names(filtered) <- names(sces)

summary_stats <- do.call(rbind,filtered)
summary_stats$Protocol <- str_split(row.names(summary_stats), "\\.", simplify=T)[,1]
```

Examine the number of cells dropped by CRISPRclean and untreated.  
655 more cells are dropped in the Untreated dataset. Slightly more than twice as many cells as CRISPRclean.

```{r}
summary_stats %>% 
  group_by(Protocol) %>% 
  count(discard) %>% 
  pivot_wider(names_from = Protocol, values_from = n)
```

Inspect which metric is most different.  
Its all from the mitochondrial percentage.

```{r}

summary_stats %>% 
  group_by(Protocol) %>% 
  count(low_lib_size, low_n_features, high_subsets_Mito_percent, discard)

```

# Plot all cells with highlight dropped cells

Find those cells that are unique to CRISPRclean or untreated.  
See where they lie on the quality control plots.

```{r}
tb$discard <- "Retained"
tb$discard[tb$`CSP-discard`] <- "CRISPR"
tb$discard[tb$`UTD-discard`] <- "Untreated"
both <- tb$`CSP-discard` == TRUE & tb$`UTD-discard` == TRUE
tb$discard[both] <- "Both"

tb %>% 
  count(discard) %>% 
  kable()
```

# UP TO HERE

## Plot quality control metrics {.tabset}

### Mitochondiral percentage

This is the key plot to view

```{r mito_per_p, fig.cap="Mitochondrial percentage of total counts in cells unique to each protocol"}
tb_wide <- tb %>% 
  pivot_wider(names_from = deplete, values_from = subsets_Mito_percent, id_cols = c(cellBC, discard))

plt3 <- ggplot(tb_wide,
             aes(x = Untreated, y= CRISPRClean, colour=discard)) + 
  geom_point(size=1, alpha=0.75) +
  ylab("CRISPRclean") + 
  xlab("Untreated") +
  ggtitle("Mitochondrial percentage") +  scale_fill_brewer(palette = "Set2")

plt3
```

## Mitochondrial with linear model

```{r mito_lm, fig.cap="Mitochondrial percentage of total counts in cells unique to each protocol", fig.width=12, fig.height=6}
tb_wide <- tb %>% 
  pivot_wider(names_from = deplete, values_from = subsets_Mito_percent, id_cols = c(cellBC, discard))

plt4 <- ggplot(tb_wide,
             aes(x = Untreated, y= CRISPRClean)) + 
  geom_point(size=1, alpha=0.75) +
  ylab("CRISPRclean") + 
  xlab("Untreated") +
  ggtitle("Mitochondrial percentage") +
  geom_smooth(method = "lm") +
  scale_fill_brewer(palette = "Set2")

plt3 + plt4
```

### Library size

```{r libSize_p, fig.cap="Library size in cells unique to each protocol"}
tb_wide <- tb %>% 
  pivot_wider(names_from = deplete, values_from = sum, id_cols = c(cellBC, discard))


plt1 <- ggplot(tb_wide,
             aes(x = Untreated, y= CRISPRClean, colour=discard)) + 
  geom_point(size=0.75, alpha=0.5) +
  ylab("CRISPRclean") + 
  xlab("Untreated") +
  ggtitle("Library size (UMIs)") +
  scale_y_continuous(trans='log10') + scale_x_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "bl") +
  scale_colour_brewer(palette = "Set2") 

plt1
```

### Detected genes

No obvious increase in genes detected

```{r detected_genes_p, fig.cap="Number of detected genes in cells unique to each protocol"}
tb_wide <- tb %>% 
  pivot_wider(names_from = deplete, values_from = detected, id_cols = c(cellBC, discard))

plt2 <- ggplot(tb_wide,
             aes(x = Untreated, y= CRISPRClean, colour=discard)) + 
  geom_point(size=1, alpha=0.75) +
  ylab("CRISPRclean") + 
  xlab("Untreated") +
  ggtitle("Genes detected") +
  scale_colour_brewer(palette = "Set2") 

plt2
```

## Session Info

```{r}
sessionInfo()
```
