---
title: "PDX117 marker gene detection"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='../figures/PDX117/marker_genes/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
suppressPackageStartupMessages({
  library(tidyverse)
  library(patchwork)
  library(scuttle)
  library(scater)
  library(scran)
  library(viridis)
  })

source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

Perform differential expression test on clusters to find marker genes.

# Recap

When I extracted nuclei and fixed with the 10x Genomics kit the sample became increasingly clumped and ended up blocking the microfluidics chip. Fix with the Parse Bio version 2 kit. 

Additionally try a 10x Genomics protocol where I fix tumor pieces first then extract nuclei.

## Samples
Human ovarian carcinosarcoma PDX tissue pieces. Has been kept frozen as rice sized pieces.

I received 4 samples from Clare Scottâ€™s lab:

| Sample | Count preFix | Count postfix |
|--------|--------------|---------------|
| PDX117 | 780,000      | 308,250       |
| PDX105 | 373,000      | 237,750       |

I passed PDX1 = PDX117 and PDX3 = PDX105 to Zac Moore for processing

Link to the [ELN entry](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/NDE0Ljd8NjgxOTcvMzE5L1RyZWVOb2RlLzE2NzI5MzY5MzV8MTA1Mi43)

# Read SCE

From the **1A_qc_dimred** notebook.

```{r loadData}
sce <- readRDS(
 "/vast/scratch/users/brown.d/Parse_PDX_Zac/SCEs/parse_dimred.pdx117.rds"
)
```

## Recap clusters in a UMAP

```{r umap_cluster}
umap_plt <- plotReducedDim(sce, dimred="UMAP", colour_by="cluster") + theme_Publication()

umap_plt
```

## Check high expessed genes

Nothing jumps out to me

```{r}
her <- plotHighestExprs(sce, n=15, colour_cells_by = "cluster") + theme_Publication()
her
```

# Marker gene detection

For each cluster, we rank candidate markers based on one of these effect size summaries.  
Use pairwise comparisons rather than comparing each cluster to the average of all other cells.

```{r}
marker.info <- scoreMarkers(sce, groups=sce$cluster)
```

Compare each pair of clusters and compute scores quantifying the differences in the expression distributions between clusters.  
The scores for all pairwise comparisons involving a particular cluster are then consolidated into a single DataFrame for that cluster.

## Iterate over each cluster

```{r}
markers <- list()

for (i in 1:length(marker.info)) {
    chosen <- marker.info[[i]]
    ordered <- chosen[order(chosen$mean.AUC, decreasing=TRUE),]

    chosen <- chosen[order(chosen$mean.AUC, decreasing=TRUE),]
    markers[[i]] <- as.data.frame(chosen[c(1:5),])
}

names(markers) <- 1:length(markers)
```

## Mung the list of marker genes

Convert to a summarised and usable form for all clusters

```{r}
marker <- do.call(rbind, markers)
marker$gene <- row.names(marker)

marker <- as_tibble(marker) %>% 
  select(gene,self.average, mean.logFC.cohen, median.logFC.cohen, 
         mean.AUC, median.AUC, 
         mean.logFC.detected, median.logFC.detected) %>% 
  separate(gene, c("cluster", "gene.symbol"), sep = 2) %>% 
  filter(mean.logFC.cohen > 0.90) %>% 
  filter(self.average > 3)

# Fix up the cluster number and gene IDs
marker$gene.symbol <- unique(gsub("^\\.", "",marker$gene.symbol))
marker$cluster <- gsub("\\.", "",marker$cluster)

marker
```

Add some custom markers in there based on EMT

```{r}
custom_marker <- c("MKI67", "VIM", "EPCAM", "DACH1", "MET")
```

## Visualise results {.tabset}

* MUC16
  + Cluster 1 and 6
  + A marker of epithelial ovarian cancer. The extracellular portion of MUC16 can be cleaved off and released into the serum, becoming the circulating marker CA125, which is a well-known marker for the recurrence.
* HSPH1
  + Cluster 3
  + Heat shock protein 70 family of proteins. Naively I would say cell stress
* PEX5L
  + Cluster 4
  + Peroxisome matrix targeting signal-1 binding activity and small GTPase binding activity. Predicted to be involved in protein import into peroxisome matrix
* NRXN1
  + Cluster 4
  + cell adhesion molecule function in synaptic transmission
* ADD3
  + Cluster 7
  + Membrane-cytoskeleton-associated protein that promotes the assembly of the spectrin-actin network
  + Related to cell migration

Cluster 1 and 6 don't seem to have different markers. Clusters 5 - 7 very similar.

### Dot plot

```{r dot_marker, fig.width=8, fig.height=20}
dot <- plotDots(sce, 
         features = c(marker$gene.symbol, custom_marker), 
         group="cluster", color=plasma(100)) +
  labs(title = "", y = "", x = "Cluster", color = "Log counts", size="Proportion of cells") +
  theme_Publication(base_size = 14)

dot
```

Write a subset of the most discriminatory markers for further investigation.

```{r}
marker_subset <- c(
  "NEAT1", "PEX5L", "NRXN1", "HSPH1", "KMT2E",
  "MUC16", "ADD3", "PKNOX2",
  "MKI67", "VIM", "EPCAM", "DACH1"
)
```

### Expression plots

There are no markers that discriminate cluster 2 and 6. I think they are the same cell type.

```{r expression, fig.dim=c(8,8)}
exprs <- plotExpression(sce, features=marker_subset, 
    x="cluster", colour_by="cluster", ncol = 3) +
  theme_Publication(base_size = 14)

exprs
```

### Expression plots technical factors

```{r tech_expression_parse, fig.asp=0.9}
marker_subset <- c(
  "MALAT1", "SDK1", "NRXN1", "MIR99AHG"
)

exprs <- plotExpression(sce, features=marker_subset, 
    x="cluster", colour_by="cluster", ncol = 2) +
  theme_Publication(base_size = 20) +
  theme(legend.position = "none", axis.text.x = element_text(angle = 0))

exprs
```

### UMPA KI67

```{r umap_ki67}
umap_plt <- plotReducedDim(sce, dimred="UMAP", colour_by="MKI67") + theme_Publication()

umap_plt
```

# Conclusion

* Cluster 4 is the major distinct cluster
* There are no markers that discriminate cluster 2 and 6. I think they are the same cell type.
* Suspect cluster 2 is proliferating
* Need to find difference between cluster 1 and 6


## Session Info

```{r}
sessionInfo()
```
