---
title: "Generate Single-cell experiment Jumpcode Ovarian PDX"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/gn_sce/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
library(DropletUtils)
library(here)
library(tidyverse)
library(patchwork)
library(scuttle)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim 

Evaluate jumpcode CRISPRclean single-cell boost kit. This adds more content on top of ribosomal and mitochondrial genes.  
Including Pseudogenes and non variable genes

## This sample

Ovarian carcinosarcoma PDX. The unlabelled capture only.

[ELN entry](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/NDA1LjZ8NjgxOTcvMzEyL1RyZWVOb2RlLzczMDU3MTIzN3wxMDI5LjY=)

## Comparisons

Jumpcode single-cell boost versus undepleted

## Sequncing metrics

| **    Sample** | **    Untreated** | **    Jumpcode** |
|---|---|---|
|        Estimated Number of   Cells |        85,838 |        84,909 |
|        Mean Reads per Cell |        2,416 |        2,442 |
|        Number of Reads |        207,384,474 |        207,384,474 |
|        Valid Barcodes |        93.80% |        96.80% |
|        Sequencing Saturation |        54.30% |        51.30% |
|        Reads Mapped to Genome |        87.60% |        97.90% |
|        GRCh38 Reads Mapped to Genome |        84.70% |        94.90% |
|        mm10 Reads Mapped to Genome |        2.90% |        3.00% |
|        Reads Mapped Confidently to Genome |        81.70% |        95.40% |
|        GRCh38 Reads Mapped Confidently to Genome |        79.30% |        92.50% |
|        mm10 Reads Mapped Confidently to Genome |        2.40% |        2.80% |
|        Reads Mapped Confidently to Intergenic   Regions |        7.20% |        6.50% |
|        GRCh38 Reads Mapped Confidently to Intergenic   Regions |        6.90% |        6.40% |
|        mm10 Reads Mapped Confidently to Intergenic   Regions |        0.30% |        0.10% |
|        Reads Mapped Confidently to Intronic Regions |        47.00% |        52.30% |
|        GRCh38 Reads Mapped Confidently to Intronic   Regions |        46.10% |        50.90% |
|        mm10 Reads Mapped Confidently to Intronic   Regions |        0.90% |        1.30% |
|        Reads Mapped Confidently to Exonic Regions |        27.40% |        36.60% |
|        GRCh38 Reads Mapped Confidently to Exonic   Regions |        26.30% |        35.20% |
|        mm10 Reads Mapped Confidently to Exonic   Regions |        1.10% |        1.40% |
|        Reads Mapped Confidently to Transcriptome |        47.30% |        64.50% |
|        GRCh38 Reads Mapped Confidently to   Transcriptome |        45.80% |        62.50% |
|        mm10 Reads Mapped Confidently to   Transcriptome |        1.50% |        2.00% |
|        Reads Mapped Antisense to Gene |        26.50% |        23.50% |
|        GRCh38 Estimated Number of Cell Partitions |        85,799 |        84,862 |
|        mm10 Estimated Number of Cell Partitions |        1,788 |        1,734 |
|        Fraction Reads in Cells |        95.70% |        95.60% |
|        GRCh38 Fraction Reads in Cells |        95.70% |        95.60% |
|        mm10 Fraction Reads in Cells |        27.70% |        38.40% |
|        GRCh38 Median Genes per Cell |        305 |        460 |
|        mm10 Median Genes per Cell |        133 |        278 |
|        GRCh38 Total Genes Detected |        32,081 |        32,366 |
|        mm10 Total Genes Detected |        15,907 |        17,216 |
|        GRCh38 Median UMI Counts per Cell |        334 |        490 |
|        mm10 Median UMI Counts per Cell |        151 |        336 |
|        GEMs with >0 Cells |        85,838 |        84,909 |
|        GEMs with >1 Cell |        85,838 |        84,909 |
|        Fraction GEMs with >1 Cell |        100.00% |        100.00% |
|        Fraction GEMs with >1 Cell (Lower Bound) |        100.00% |        100.00% |
|        Fraction GEMs with >1 Cell (Upper Bound) |        100.00% |        100.00% |
|        Mean UMI Count Purity    |        97.30%    |        97.70%    |

# Load Data

Find the paths to the cell ranger output

```{r dataPaths}
head_path <- "/stornext/Projects/score/GenomicsRnD/DB/S000331/cellranger_outs/"
tail_path <- "/outs/raw_feature_bc_matrix"

UTD_DataPath <- paste(head_path, "G000225_PDX_UTD", tail_path, sep="")

UTD_mask_DataPath <- paste(head_path, "G000225_PDX_UTD-mask", tail_path, sep="")

JMP_DataPath <-  paste(head_path, "G000225_PDX_JMPds", tail_path, sep="")

JMP_mask_DataPath <-paste(head_path, "G000225_PDX_JMPds-mask", tail_path, sep="")

```

Load data using DropletUtils

```{r loadData}
utdData <- read10xCounts(UTD_DataPath,version='3', col.names = T)
jmpData <- read10xCounts(JMP_DataPath,version='3', col.names = T)

utdData$Protocol <- "Undepleted"
jmpData$Protocol <- "CRISPRclean-Boost"
```

# Convert to list of sces

```{r saveIntRDS}
sces <- list(utdData, jmpData)
names(sces) <- c("Untreated", "Jumpcode")

sces_orig <- sces
```

## Subset sces for valid cell barcodes

Use the valid cell barcodes from the analysis performed in *OvarianCarcinosarcoma/notebooks/S000322/1D_mungSCE.Rmd*. 
Doublets and cells that are unassigned to any donor are included

```{r}
manual_cell_call <- readRDS((
  "/stornext/Projects/score/Analyses/R010_multiplexing/SCEs/PDX/S000322.qcmetrics.sces.rds"
))[["Unlabelled"]]

valid_bc <- colnames(manual_cell_call)

for (n in names(sces)) {
  current <- sces[[n]]
  sces[[n]] <- current[,valid_bc]
}
```

## Write basic object to file

SCE object

```{r}
saveRDS(sces, here::here(
  "OvarianCarcinosarcoma/data/SCEs/CRISPRclean_score", "PDX_CRISPR_basic.sces.rds"
))

# sces <- readRDS(here::here(
#  "OvarianCarcinosarcoma/data/SCEs/CRISPRclean_score", "PDX_CRISPR_basic.sces.rds"
#))
```

# Attach donor metadata

Add doublet and donor information

```{r}
mtd <- as_tibble(colData(manual_cell_call))

mtd <- mtd %>% 
  select(cell, donor_id, prob_doublet)
```

Add rownames as gene ids and attach donor labels

```{r}
# sces <- sces_orig

for (n in names(sces)) {
  current <- sces[[n]]
  row.names(current) <- rowData(current)$Symbol
  col_data <- as.data.frame(colData(current))
  col_data <- left_join(x = col_data, y = mtd, 
                      by= c("Barcode" = "cell"))
  colData(current) <- DataFrame(col_data)
  colnames(current) <- col_data$Barcode
  sces[[n]] <- current
}
```

# Split species and fix gene names

Alt experiments enables ERCC, CITE-Seq hashtag data etc to be split from the main gene expression matrix.

```{r}
for (i in 1:length(names(sces))) {
  current <- sces[[i]]
  is.mouse <- grepl("^mm10___", rownames(current))
  current <- splitAltExps(current, ifelse(is.mouse, "Mouse", "gene"))
  sces[[i]] <- current
}

```

## Add gene names

Human

```{r}
for (i in 1:length(names(sces))) {
  current <- sces[[i]]
  gene_symbol <- rowData(current)$Symbol
  gene_symbol <- gsub(pattern = "^GRCh38_", 
                       replacement = "", x = gene_symbol)
  row.names(current) <- gene_symbol
  sces[[i]] <- current
}
```

Mouse

```{r}
for (i in 1:length(names(sces))) {
  current <- sces[[i]]
  gene_symbol <- rowData(altExp(current, "Mouse"))$Symbol
  gene_symbol <- gsub(pattern = "^mm10___", 
                       replacement = "", x = gene_symbol)
  row.names(altExp(current, "Mouse")) <- gene_symbol
}
```

# Calculate jumpcode metrics

Load the jumpcode panel

```{r}
jpc <- read_csv(here::here(
  "CRISPRclean/metadata/Jumpcode", "CRISPRclean_Single_Cell_RNA_Boost_Kit_depletion_content_set.csv"
))
```

Calculate Metrics

```{r}
for (n in names(sces)) {
    current <- sces[[n]]
    mito_set <- grep("^MT-" ,row.names(current), value = T)
    is_mito <- rownames(current) %in% mito_set
    crispClean_set <- jpc$GeneID
    is_crc <- rownames(current) %in% crispClean_set

    current <- scater::addPerCellQC(current, 
                                    subsets = list(Mito=is_mito, Jumpcode=is_crc)
                 )
    sces[[n]] <- current
}

```

Write col data to tibble and save

```{r}
tb <- as_tibble(rbind(
  colData(sces[[1]]),
  colData(sces[[2]])
))

tb <- tb %>% 
  select(-Sample)
```

# Save outputs

```{r}
saveRDS(sces,
  "/vast/scratch/users/brown.d/S000331/SCEs/PDX_CRISPR_colData.sces.rds"
)

write_csv(tb, here::here(
  "CRISPRclean/data/PDX/Matrices", "PDX_CRISPR_colData.csv.gz"
))
```

## Session Info

```{r}
sessionInfo()
```
