---
title: "CRISPRclean depletion performance PDX"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/crisprclean_performance/',
                      warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
library(DropletUtils)
library(tidyverse)
library(patchwork)
library(scuttle)
library(scater)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

Evaluate jumpcode CRISPRclean single-cell boost kit. This adds more content on top of ribosomal and mitochondrial genes.  
Including Pseudogenes and non variablee genes

## This sample

Ovarian carcinosarcoma PDX. The unlabelled capture only.
[ELN entry](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/NDA1LjZ8NjgxOTcvMzEyL1RyZWVOb2RlLzczMDU3MTIzN3wxMDI5LjY=)

## Sequencing metrics

| **    Sample** | **    Untreated** | **    Jumpcode** |
|---|---|---|
|        Estimated Number of   Cells |        85,838 |        84,909 |
|        Mean Reads per Cell |        2,416 |        2,442 |
|        Number of Reads |        207,384,474 |        207,384,474 |
|        Valid Barcodes |        93.80% |        96.80% |
|        Sequencing Saturation |        54.30% |        51.30% |
|        Reads Mapped to Genome |        87.60% |        97.90% |
|        Reads Mapped Confidently to Genome |        81.70% |        95.40% |
|        Reads Mapped Confidently to Intergenic   Regions |        7.20% |        6.50% |
|        Reads Mapped Confidently to Intronic Regions |        47.00% |        52.30% |
|        Reads Mapped Confidently to Exonic Regions |        27.40% |        36.60% |
|        Reads Mapped Confidently to Transcriptome |        47.30% |        64.50% |
|        Fraction Reads in Cells |        95.70% |        95.60% |
|        GRCh38 Total Genes Detected |        32,081 |        32,366 |

# Load Data

This SCE list is generated in **1A_generateSCE.Rmd**

```{r loadData}
sces <- readRDS(
 "/vast/scratch/users/brown.d/S000331/SCEs/PDX_CRISPR_colData.sces.rds"
)

tb <- read_csv(here::here(
   "CRISPRclean/data/PDX/Matrices", "PDX_CRISPR_colData.csv.gz"
))

```

Load the jumpcode panel

```{r}
jpc <- read_csv(here::here(
  "CRISPRclean/metadata/Jumpcode", "CRISPRclean_Single_Cell_RNA_Boost_Kit_depletion_content_set.csv"
))
```

# Visualizing the QC metrics {.tabset}

I subset the SCE object for valid cells in the unlabelled capture

```{r}
tb %>% 
  dplyr::count(Protocol)
```

## Library size

The CRISPRclean treated sample has a slightly higher library size. This must be related to the mapping rrate being higher in CRISPR sample. Probably the rmoval of pseudogenes has caused this.

```{r libSize_p, fig.cap="Library size in deduplicated UMIs"}
plt1 <- ggplot(tb, 
             aes(x = Protocol, y= sum+1, fill = Protocol)) + 
  geom_boxplot(show.legend = FALSE) + 
  ylab("Library Size (UMIs)") + 
  xlab("") +
  scale_y_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "l") +
  scale_fill_brewer(palette = "Set1") 

plt1
```

## Detected genes

Consistent with library size.

```{r detected_genes_p, fig.cap="Number of detected genes"}
plt2 <- ggplot(tb, 
             aes(x = Protocol, y= sum+1, fill = Protocol)) +  
  geom_boxplot(show.legend = FALSE) + 
  scale_y_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "l") +
  ylab("Detected genes") + 
  xlab("") +
  scale_fill_brewer(palette = "Set1") 

plt2
```

## Mitochondrial percentage

The background is low, this being a nuclei sample and all. The depletion has worked as intended.  
There does not appear to be any outliers in CRISPR sample

```{r mito_per_p, fig.cap="Mitochondrial percentage of total counts"}
plt3 <- ggplot(tb, 
             aes(x = Protocol, y= subsets_Mito_percent, fill = Protocol)) + 
  geom_boxplot(show.legend = FALSE, outlier.size = 0.5, outlier.alpha = 0.5) + 
  ylab("Mitochondrial\n percentage") + 
  xlab("") +
  scale_fill_brewer(palette = "Set1")

plt3
```

## Genes in jumpcode panel

* The median percent of counts in CRISPRclean panel is `r median(tb$subsets_Jumpcode_percent[tb$Workflow == "CRISPRclean Boost_Whole"])`
* The median percent of counts in CRISPRcleean panel is `r median(tb$subsets_Jumpcode_percent[tb$Workflow == "Undepleted_Whole"])`


```{r, eval=FALSE}
median(tb$subsets_Jumpcode_percent[tb$Protocol == "CRISPRclean Boost"])
median(tb$subsets_Jumpcode_percent[tb$Protocol == "Undepleted"])
```

Much like the mitochondiral percent this is reduced to zero

```{r panel_count_p, fig.cap="Library size of multiplexing capture sequencing"}
plt4 <- ggplot(tb, 
             aes(x = Protocol, y= subsets_Jumpcode_percent, fill = Protocol)) + 
  geom_boxplot(show.legend = FALSE, outlier.size = 0.5, outlier.alpha = 0.5) + 
  ylab("Jumpcode panel\n percentage") + 
  xlab("") +
  scale_fill_brewer(palette = "Set1") +
  theme_Publication(base_size=20)

plt4
```

# Put QC plots together

```{r qc_summary_p, fig.cap="Summary of QC metrics", fig.asp=1}
plt1 +  plt2 + plt3 + plt4 + 
  plot_layout(guides = "collect", ncol = 2)
```

# UMI counts vs genes {.tabset}

No visible shift, only the magnitude of CRISPRclean is higher due to higher library size.  
This is consistent with the sequencing saturation being similar for both samples.

## Log scale


```{r umi_genes_log}
plt5 <- ggplot(tb, 
             aes(x = sum, y= detected, colour = Protocol)) + 
  geom_point(alpha = 0.25, size=0.5) + 
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1))) +
  xlab("Library Size (UMIs)") + 
  ylab("Detected Genes") +
  scale_x_continuous(trans='log10') +
  scale_y_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "bl") +
  scale_colour_brewer(palette = "Set1") +
  theme_Publication(base_size=20)

plt5
```

## Natural scale

```{r umi_genes_nat}
plt5 <- ggplot(tb, 
             aes(x = sum, y= detected, colour = Protocol)) + 
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1))) +
  geom_point(alpha = 0.25, size=0.5) + 
  xlab("Library Size (UMIs)") + 
  ylab("Detected Genes") +
  scale_colour_brewer(palette = "Set1")

plt5
```

# Look at high expressing genes {.tabset}

plotHighestExprs of scater package: Plot the features with the highest average expression across all cells, along with their expression in each individual cell.

The CRISPR treated sample has worked as expected with genes in the panel being removed.  

```{r}
names(sces)
sces_whole <- sces[c(1,2)] 
```

## Combine 1 figure

Best to calculate and plot this separately. Large size

```{r highExpr_big}
sce_jmp <- sces[["Jumpcode"]]
sce_utd <- sces[["Untreated"]]

p1 <- plotHighestExprs(sce_utd, n = 15) + ggtitle("Undepleted") +
  theme_Publication(base_size=16)
p2 <- plotHighestExprs(sce_jmp, n = 15) + xlim(0,20) + ggtitle("CRISPRclean") +
  theme_Publication(base_size=16)

p1 + p2 + plot_layout(guides = "collect")
```

Small size

```{r highExpr_small}
p3 <- plotHighestExprs(sce_utd, n = 20) + ggtitle("Undepleted") +
  theme_Publication(base_size=12)
p4 <- plotHighestExprs(sce_jmp, n = 20) + xlim(0,20) + ggtitle("CRISPRclean") +
  theme_Publication(base_size=12)

p3 + p4 + plot_layout(guides = "collect")
```

## Untreated

```{r untreated_highExpr}
p1
```

## Depleted

```{r jumpcode_highExpr}
p2
```

## Session Info

```{r}
sessionInfo()
```
