---
title: "CRISPR clean DE PDX"
author: "Daniel Brown"
description: "Differential expression between CRISPRclean and untreated"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/DEtest_crisprclean/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())
library(scater)
library(scran)
library(edgeR)
library(patchwork)
library(knitr)
library(tidyverse)
library(plotly)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

Evaluate jumpcode CRISPRclean single-cell boost kit. This adds more content on top of ribosomal and mitochondrial genes.  
Including Pseudogenes and non variablee genes.

In this notebook perform the DE testing between treated and untreated.

## This sample

Ovarian carcinosarcoma PDX. The unlabelled capture only.
[ELN entry](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/NDA1LjZ8NjgxOTcvMzEyL1RyZWVOb2RlLzczMDU3MTIzN3wxMDI5LjY=)

## Comparisons

I mapped both the undepleted and CRISPRclean depleted datasets to GRCh38 reference provided by cellranger and the masked reference provided by jumpcode. Jumpcode says only the masked reference will highlight the differences in depleted.

# Load Data

Read in data

```{r loadData}
sces <- readRDS(
 "/vast/scratch/users/brown.d/S000331/SCEs/PDX_CRISPR_colData.sces.rds"
)
names(sces)

sce_jmp <- sces[["Jumpcode"]]
sce_utd <- sces[["Untreated"]]
```

Load the jumpcode panel

```{r}
jpc <- read_csv(here::here(
  "CRISPRclean/metadata/Jumpcode", "CRISPRclean_Single_Cell_RNA_Boost_Kit_depletion_content_set.csv"
))
```

Combine into a single object

```{r rejoin}
sce <- cbind(sce_jmp, sce_utd)
rm(sces, sce_utd, sce_jmp)
```

# Differential expression test CRISPRclean vs untreated

Here I will follow the example of *Chapter 4 DE analyses between conditions
*http://bioconductor.org/books/3.14/OSCA.multisample/multi-sample-comparisons.html*

Keep only the singlets assigned to donors for simplicity

```{r}
keep_cells <- c("donor0", "donor1", "donor2", "donor3")
sce_singlet <- sce[,sce$donor_id %in% keep_cells]

table(sce_singlet$Protocol, sce_singlet$donor_id)
```

## Creating pseudo-bulk samples

Sum counts together for all cells with the same combination of label and sample.

```{r pseudo-bulk}
summed <- aggregateAcrossCells(sce_singlet, 
    id=colData(sce_singlet)[,c("Protocol", "donor_id")])

```

## Performing the DE analysis

Use quasi-likelihood (QL) methods from the edgeR package

```{r}
dge <- DGEList(counts(summed), samples=colData(summed))

```

Remove genes that are lowly expressed.

```{r}
keep <- filterByExpr(dge, group=dge$Protocol)
dge <- dge[keep,]
summary(keep)
```

Correct for composition biases by computing normalization factors with the trimmed mean of M-values method.

```{r}
dge <- calcNormFactors(dge)
dge$samples

in_crispr <- row.names(dge) %in% jpc$GeneID
```

## Generate a mean-difference (MD) plot for each normalized pseudo-bulk profile

```{r md_plot}
par(mfrow=c(2,2))
for (i in seq_len(ncol(dge))) {
    plotMD(dge, column=i, status = in_crispr,
           xlab="Average logCPM",
           main ="",
           ylab="log ratio (Untreated vs CRISPRClean)",
           legend = FALSE)
}
```

## Generate a multi-dimensional scaling (MDS) plot for the pseudo-bulk profiles

The major difference between samples is the CRISPRclean treatment, less so for the donor ID

```{r mds}
par(mfrow=c(1,1))

plotMDS(cpm(dge, log=TRUE),
        col=c(rep("red",4), rep("blue",4)), 
        labels= dge$samples$donor_id)
```

# Perform differential expression testing

## Model matrix

```{r}
design <- model.matrix(~factor(Protocol) + factor(donor_id), dge$samples)

dge <- estimateDisp(dge, design)
summary(dge$trended.dispersion)

plotBCV(dge)
```

## Fit BCV

```{r}
fit <- glmQLFit(dge, design, robust=TRUE)
summary(fit$var.prior)
plotQLDisp(fit)
```

## Perform DE test

```{r}
res <- glmQLFTest(fit, coef="factor(Protocol)Undepleted")
summary(decideTests(res))

topTags <- topTags(res, n=20000)
head(topTags, 20)
```

```{r volcano}
tb <- as_tibble(topTags)[[1]]
tb$In_Panel <- row.names(tb) %in% jpc$GeneID

# add a column of NAs
tb$DElabel <- FALSE
# if log2Foldchange > 1 or < -1 and pvalue < 0.05, set as "UP"
tb$DElabel[tb$logFC > 2 & tb$FDR < 0.05] <- "DE"
tb$DElabel[tb$logFC < -2 & tb$FDR < 0.05] <- "DE"
tb$DElabel[tb$DElabel == "DE"] <- row.names(tb)[tb$DElabel == "DE"]
tb$DElabel[tb$DElabel == FALSE] <- NA

plt1 <- ggplot(data=tb, aes(x=logFC, y=-log10(FDR), colour=In_Panel, label=row.names(tb))) + 
  geom_point(alpha=0.33, size=1.5) +
  geom_text(size=1, alpha=0) +
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1))) +
  scale_color_manual(values = c("blue", "red")) +
  theme_Publication()

plt1
```

Plot MA plot

```{r maPlot}

plt2 <- ggplot(data=tb, aes(x=logCPM, y=logFC, colour=In_Panel, label=row.names(tb))) + 
  geom_point(alpha=0.25, size=1.5) +
  geom_text(size=1, alpha=0) +
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1))) +
  ylab("Log fold chnage") + xlab("Log counts per million") +
  scale_color_manual(values = c("blue", "red"), name = "In CRIPSRclean \npanel") +
  theme_Publication()

plt2
```

# Extract DE genes for further investigation

Remove genes not in the panel.

```{r}
tb_de <- tb[!is.na(tb$DElabel),]
tb_de <- as_tibble(tb_de[!tb_de$In_Panel,])

tb_de %>% 
  arrange(desc(logFC))
```

write output to file

```{r}
write_csv(tb, here::here(
  "CRISPRclean/data/PDX/Matrices", "PDX_DEgenes.csv.gz"
))
```

# Conclusion

The most differently expressed genes are present in the Jumpcode panel or can be reasonably be expected to cross react through sequence similarity.

## Session Info

```{r}
sessionInfo()
```
