---
title: "CRISPR clean offtarget effects PDX"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/off_target/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())
library(scater)
library(scran)
library(patchwork)
library(ggvenn)
library(tidyverse)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

Evaluate jumpcode CRISPRclean single-cell boost kit. This adds more content on top of ribosomal and mitochondrial genes.  
Including Pseudogenes and non variablee genes.

In this notebook perform the DE testing between treated and untreated.

## This sample

Ovarian carcinosarcoma PDX. The unlabelled capture only.
[ELN entry](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/NDA1LjZ8NjgxOTcvMzEyL1RyZWVOb2RlLzczMDU3MTIzN3wxMDI5LjY=)

## JUMPcode scCLEAN publication

The following quotes from the preprint

*Pandey, Amitabh C., Jon Bezney, Dante DeAscanis, Ethan Kirsch, Farin Ahmed, Austin Crinklaw, Kumari Sonal Choudhary et al. "A CRISPR/Cas9-based enhancement of high-throughput single-cell transcriptomics." bioRxiv (2022): 2022-09.*

* While 33 genes passed the significance threshold, only 21 were determined to be off-target and 18 were direct targets of scCLEAN guides (Supplementary Fig. 1d-e).
* Distinguishing between genes that are a product of random sampling from two libraries of fundamentally different composition and true off-target effects. 
  + Genes that donâ€™t follow a linear trend in log2 fold change reduction (negative values represent a decrease in scCLEAN expression relative to 10x-V3) are considered offtarget (n=21) and are automatically detected by a knee bend algorithm. 
* e, Quantifying whether the 21 selected genes in this hierarchical selection process are targets of scCLEAN guides or overlap with regions of on-target genes (18/21)

# Load Data

Read in data

```{r loadData, eval=FALSE}
sces <- readRDS(here::here(
 "/vast/scratch/users/brown.d/S000331/SCEs/PDX_CRISPR_colData.sces.rds"
))
names(sces)

sce_jmp <- sces[["Jumpcode"]]
sce_utd <- sces[["Untreated"]]

sce <- cbind(sce_jmp, sce_utd)
rm(sces, sce_utd, sce_jmp)
```

Read the differential expressed genes between Jumpcode and untreated from *2B_DEtest_crisprClean* notebook.  
Subset for the significant off-target elements.

```{r}
tb <- read_csv(here::here(
  "CRISPRclean/data/PDX/Matrices", "PDX_DEgenes.csv.gz"
))

db_expt <- tb %>% 
  filter(logFC >= 1 & PValue < 0.05) %>% 
  filter(In_Panel == FALSE) %>% 
  select(DElabel)

db_expt <- db_expt$DElabel
db_expt <- db_expt[!is.na(db_expt)]
```

Remove the novel transcripts which are unannotated genes, pseduogenes, miRNAs, lncRNAs etc They start with "AC" pr "AL"

```{r}
db_expt <- db_expt[grep("^AC" ,db_expt, invert = TRUE)]
db_expt <- db_expt[grep("^AL" ,db_expt, invert = TRUE)]
```

Load the jumpcode panel

```{r}
jpc <- read_csv(here::here(
  "CRISPRclean/metadata/Jumpcode", "CRISPRclean_Single_Cell_RNA_Boost_Kit_depletion_content_set.csv"
))
```

Load the off target panel from 

```{r}
offtarget <- read_csv(here::here(
  "CRISPRclean/metadata/Jumpcode", "Paandey_2022_offtarget.csv"
))
```

# Create venn digram

```{r mung_venn, fig.dim=c(10,14)}
jpc_expt <- offtarget$GeneID

venn_list <- list(
  "scBoost Brown" = db_expt, 
  "Pandey2022" = jpc_expt
  )

ggvenn(venn_list,
       show_elements = T, label_sep = "\n",
       fill_color = c("blue", "green")
       )

```

# Conclusion

There are only 5 genes overlapping between my PDX dataset and Pandey 2022.

## Session Info

```{r}
sessionInfo()
```
