---
title: "Oligo and gene expression library size"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=1, 
                      fig.path='../figures/oligo_gex_size/',
                      warning=FALSE, message=FALSE,
                      layout="l-body-outset")
library(here)
knitr::opts_knit$set(root.dir = here::here())
library(tidyverse)
library(knitr)
library(SingleCellExperiment)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication(base_size = 16))
```

# Recap
The purpose of this experiment was to compare the performance of 3 different cell multiplexing reagents in PBMCs.  
To obtain the ground truth of performance, the genotypes may be used to demultiplex donors.
In this experiment there are 4 donors. The cell multiplexing was done in technical duplicate.  

# Aim

Plot summary statistics for Cell multiplexing oligos

## Compare:

Reagents
1. Total-Seq A hashtag
2. MULTI-seq LMO
3. CellPlex reagent

Donors
1. VBDR1233
2. VBDR1206
3. VBDR1186
4. VBDR1216

# Read data

```{r loadSCE}
cmo_sce <- readRDS(here::here(
  "PBMC/data/PBMCs", "cmo-annotate.rds"))

lmo_sce <- readRDS(here::here(
  "PBMC/data/PBMCs", "lmo-annotate.rds"))

hto_sce <-  readRDS(here::here(
  "PBMC/data/PBMCs", "hto-annotate.rds"))


all.sce <- list(cmo_sce, hto_sce, lmo_sce)
names(all.sce) <- c("CellPlex", "Hashtag Ab", "MULTI-Seq LMO")
```

Convert to tibble

```{r}
cmo_tb <- as_tibble(colData(cmo_sce))
hto_tb <- as_tibble(colData(hto_sce))
lmo_tb <- as_tibble(colData(lmo_sce))

tb <- rbind(
  cmo_tb,
  hto_tb,
  lmo_tb
)

tb$CellRangerCall <- tb$mul_Assignment
```

Mung the tibble.

```{r mung_tb}
tb$CellRangerCall <- recode(tb$CellRangerCall,
                      CMO301="Singlet",
                      CMO302="Singlet",
                      CMO303="Singlet",
                      CMO304="Singlet",
                      CMO305="Singlet",
                      CMO306="Singlet",
                      CMO307="Singlet",
                      CMO308="Singlet",
                      HTO_1="Singlet",
                      HTO_2="Singlet",
                      HTO_3= "Singlet",
                      HTO_4= "Singlet",
                      HTO_5= "Singlet",
                      HTO_6= "Singlet",
                      HTO_7= "Singlet",
                      HTO_8= "Singlet",
                      MULTI_2= "Singlet",
                      MULTI_3= "Singlet",
                      MULTI_4= "Singlet",
                      MULTI_5= "Singlet",
                      MULTI_6= "Singlet",
                      MULTI_7= "Singlet",
                      MULTI_8= "Singlet",
                      MULTI_9= "Singlet"
)

tb$Protocol <- recode(tb$Protocol,
                      CMO="CellPlex",
                      HTO="Hashtag Ab",
                      LMO="MULTI-Seq LMO"
)

```

# Plot oligo and gene exprression library size

```{r lib_size_xy}
p1 <- ggplot(tb, 
             aes(x = altexps_Multiplexing.Capture_sum, y= sum, colour = CellRangerCall)) + 
  geom_point(size=0.5, alpha=0.33) +
  ylab("Gene Expression \nLibrary Size (UMI)") + 
  xlab("Oligo Library Size (UMI)") + 
  scale_color_brewer(palette = "Dark2", name = "Droplet call") +
  scale_y_continuous(trans='log10') + scale_x_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "bl") +
  guides(color = guide_legend(override.aes = list(size = 2, alpha=1))) +
  facet_wrap(~Protocol, ncol=1)

p1
```

## Session info
```{r sessionInfo}
sessionInfo()
```