---
title: "Build SCE 10x Genomics example data"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
library(scater)
library(tidyverse)
library(DropletUtils)
```

# Inputs/Library

The following is from https://www.10xgenomics.com/resources/datasets/30-k-mouse-e-18-combined-cortex-hippocampus-and-subventricular-zone-cells-multiplexed-12-cm-os-3-1-standard-6-0-0

Cells from four E18 mouse combined cortex, hippocampus and ventricular zone tissues (C57BL/6 strain) were obtained by 10x Genomics from BrainBits LLC. Cells from the four E18 mouse tissues were multiplexed at equal proportions with 3 CMOs per tissue, resulting in a pooled sample labeled with 12 CMOs. Cells from the non-multiplexed sample were used as one of the four sample types that composed the multiplexed sample.

The config.csv input file was submitted with four sample IDs (correlating to the four E18 mouse tissues) and with three respective CMO IDs assigned per sample ID.

## Libraries

Libraries were prepared following the Chromium Next GEM Single Cell 3ʹ Reagent Kits v3.1 (Dual Index) with Feature Barcode technology for Cell Multiplexing User Guide (CG000388) and sequenced on Illumina NovaSeq 6000.

Single Cell 3’ Cell Multiplexing v3.1 Dual Index Library

## Metrics

* Sequencing Depth: 25,417 read pairs per cell
Paired-end, dual indexing Read 1: 28 cycles (16 bp barcode, 12 bp UMI); i5 index: 10 cycles (sample index); i7 index: 10 cycles (sample index); Read 2: 90 cycles (transcript)
Pooled Multiplexed Sample - Key Metrics

* Estimated Number of Cells: 33,197
* Cells Assigned to a Sample: 23,338

# Load Data {.tabset}

Find the paths to the cell ranger output

```{r dataPaths}
path = here::here(
  'data/10xGenomicsExample/raw_feature_bc_matrix'
)
```

```{r cmo}
cmoAnno <- read.csv(here::here(
  "data/10xGenomicsExample", "SC3_v3_NextGem_DI_CellPlex_Neurons_30K_Multiplex_multiplexing_analysis_assignment_confidence_table.csv"
))
# Keep only those cells called by cell ranger
keep_cells <- cmoAnno$Barcodes

cmoData <- read10xCounts(path, col.names = T)
cmoData <- cmoData[,keep_cells]

cmoData$Protocol = "CellPlex"
cmoData$mul_Assignment <- cmoAnno$Assignment
cmoData$mul_Assignment_Probability <- cmoAnno$Assignment_Probability

cmoData <- splitAltExps(
  cmoData,
  rowData(cmoData)$Type,
  "Gene Expression")
```

## Add gene IDs as row names

``` {r rownames}
row.names(cmoData) <- rowData(cmoData)$Symbol
```

## Mitochondrial content

```{r mito_qc}

mito_set <- grep("^mt-" ,row.names(cmoData), value = T)
is_mito <- rownames(cmoData) %in% mito_set
cmoData <- scater::addPerCellQC(cmoData, 
                                subsets = list(Mito = which(is_mito)))
```

# Save SCE Objects

Save the basic sce objects

```{r saveSCEs}
saveRDS(cmoData, file = here::here(
  "data/10xGenomicsExample/SCEs", "cmo-basic.rds"
  ))
```

## Save tag counts

I need the oligo tag counts to run cellHashR.

```{r savetagcounts, eval=F, include=F}
write10xCounts(path = here::here(
  "data/10xGenomicsExample/Matrices", "cmo_tag_counts"
  ), x = counts(altExp(cmoData)))
```

# Attach colData

```{r colData}
col_data <- as.data.frame(cbind(
  colData(cmoData),
  cmoAnno))

head(col_data)

col_data <- col_data %>% 
  select(Barcode:total,Multiplet,Blanks)
```

## Add additional cell metadata

Code in singlets and doublets

```{r singletDoublet}
col_data$Droplet <- "singlet"
col_data$Droplet[col_data$mul_Assignment == "Multiplet"] <- "multiplet"
col_data$Droplet[col_data$mul_Assignment == "Unassigned"] <- "unassigned"
```

```{r attachColData}
colData(cmoData) <- DataFrame(col_data)
```

# Save SCE Objects

Save the basic sce objects

```{r saveSCEs}
saveRDS(cmoData, file = here::here(
  "data/10xGenomicsExample/SCEs", "cmo-anno.rds"
  ))
```

## Session Info

```{r}
sessionInfo()
```
