---
title: "Oligo metrics summary after downsampling"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='../../figures/downsample_oligo_metrics/',
                      warning=FALSE, message=FALSE,
                      layout="l-body-outset")
library(here)
knitr::opts_knit$set(root.dir = here::here())
library(tidyverse)
library(knitr)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Recap
The purpose of this experiment was to compare the performance of 3 different cell multiplexing reagents in PBMCs.  
To obtain the ground truth of performance, the genotypes may be used to demultiplex donors.
In this experiment there are 4 donors. The cell multiplexing was done in technical duplicate.  

# Aim

Plot summary statistics for Cell multiplexing oligos. I performed downsampling of the multiplexing tags fastq files.

There is no way I know of getting a fixed reads per cell for tag oligos.  
The best I can do is get the cell number of each capture and multiply by the number of reads.  
This is described in *scripts/downsampling/downsample_tags*

* CMO = 19,943 cells
* LMO = 21,419 cells
* HTO = 21,131 cells

## Compare:

Reagents
1. Total-Seq A hashtag
2. MULTI-seq LMO
3. CellPlex reagent

Donors
1. VBDR1233
2. VBDR1206
3. VBDR1186
4. VBDR1216

## Data munging

The original notebook in R010_multiplexing has the correct paths.  
I have a copy in the folder PBMC/notebooks/downsampling/oligo_summary_metrics.Rmd

For the time being it is too time consuming to fix all the path names. Read in the processedd metrics for plotting.  
Will need to split in to a data generation Rscript at some point.

### Read data

```{r}
tag_call <- read_csv(here::here(
  "PBMC/data/Matrices/downsampling", "multiplexing_summary_stats.csv"
))
```

The category No tag assigned includes both cells that were considered Blanks and cells considered Unassigned

## Plot broad classification {.tabset}

* There is a jump between 100 and 500 reads per cell but beyond that there is not much difference.
* The difference between 1,000 and 5,000 is almost nothing.  
* Only the multiplet and unassigned seems to be changing.

### By Label
```{r Label}
plt1 <- ggplot(data=tag_call) +
  geom_point(aes(y=pct_cells, x=reads, colour=call), 
              size=3) +
  geom_line(aes(y=pct_cells, x=reads, colour=call)) +
  scale_y_continuous(breaks=seq(0,100,20), limits = c(0,100)) + 
  #scale_colour_brewer(type="qualitative", palette = "Dark2") +
  scale_colour_manual(values=c("#E69F00", "#56B4E9", "#999999"), name = "Droplet call") +
  xlab("Downsampled oligo tag reads per cell") + ylab("Percent of capture") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_wrap(~protocol)

plt1
```

## Session info
```{r sessionInfo}
sessionInfo()
```