---
title: "Check demuxafy stats"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/demuxafy/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())
library(patchwork)
library(knitr)
library(tidyverse)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Aim

Because I don't have distinct genotypes for the mouse embryo I have no ground truth to compare multiplexing tag oligo results.
I tried demuxafy which is a wrapper for many common tools that call doublets from gene expression data https://demultiplexing-doublet-detecting-docs.readthedocs.io/en/latest/Background.html

## Take the voting majority demuxafy {.tabset}

This are matrices created with demuxafy *CombineResults.R* in the singularity image

Explanation of how I do a majority vote:

* Count the agreement of of the different doublet calling algorithms
* Filter for the majority calls

### Cellplex

```{r cellplex}
current <- "cellplex"

tb <- read_delim(file = here::here(
  "data/NN265/cellranger_multi_out/demuxafy", current ,"combined_out"
  ))

# Count number of votes
cellplex_tb <- demux_tb %>% 
  select(Barcode, DoubletFinder_DropletType, scDblFinder_DropletType, scds_DropletType, scrublet_DropletType) %>% 
  pivot_longer(!Barcode, names_to = "Algorithm", values_to = "count" )%>% 
  count(Barcode,count) %>% 
  filter(n > 2)

# Join and recode cell barcodes with fewer than 3 votes as NA
tb <- left_join(tb, cellplex_tb,
          by = c("Barcode" = "Barcode")) %>% 
  select(Barcode, count) %>% 
  replace_na(list(Barcode = "Unknown", count = "unassigned"))

colnames(tb) <- c("Barcode", "Demuxafy_assignment")
tb
cellplex <- tb
```

### MULTI-Seq LMO

```{r multiseqLMO}
current <- "multiseqLMO"

tb <- read_delim(file = here::here(
  "data/NN265/cellranger_multi_out/demuxafy", current ,"combined_out"
  ))

# Count number of votes
long_tb <- tb %>% 
  select(Barcode, DoubletFinder_DropletType, scDblFinder_DropletType, scds_DropletType, scrublet_DropletType) %>% 
  pivot_longer(!Barcode, names_to = "Algorithm", values_to = "count" )%>% 
  count(Barcode,count) %>% 
  filter(n > 2)

# Join and recode cell barcodes with fewer than 3 votes as NA
tb <- left_join(tb, long_tb,
          by = c("Barcode" = "Barcode")) %>% 
  select(Barcode, count) %>% 
  replace_na(list(Barcode = "Unknown", count = "unassigned"))

colnames(tb) <- c("Barcode", "Demuxafy_assignment")
tb
multiseqLMO <- tb
```

### MULTI-Seq CMO

```{r multiseqCMO}
current <- "multiseqCMO"

tb <- read_delim(file = here::here(
  "data/NN265/cellranger_multi_out/demuxafy", current ,"combined_out"
  ))

# Count number of votes
long_tb <- tb %>% 
  select(Barcode, DoubletFinder_DropletType, scDblFinder_DropletType, scds_DropletType, scrublet_DropletType) %>% 
  pivot_longer(!Barcode, names_to = "Algorithm", values_to = "count" )%>% 
  count(Barcode,count) %>% 
  filter(n > 2)

# Join and recode cell barcodes with fewer than 3 votes as NA
tb <- left_join(tb, long_tb,
          by = c("Barcode" = "Barcode")) %>% 
  select(Barcode, count) %>% 
  replace_na(list(Barcode = "Unknown", count = "unassigned"))

colnames(tb) <- c("Barcode", "Demuxafy_assignment")
tb
multiseqCMO <- tb
```

## Next steps

Join to the single cell experiment object in the *_create_SCE.Rmd*

# Write outputs

```{r}
write_tsv(cellplex, file = here::here(
  "data/demuxafy", "cellplex", "cellplex_demuxafy_calls.tsv"
))

write_tsv(multiseqLMO, file = here::here(
  "data/demuxafy", "multiseqLMO", "multiseqLMO_demuxafy_calls.tsv"
))

write_tsv(multiseqCMO, file = here::here(
  "data/demuxafy", "multiseqCMO", "multiseqCMO_demuxafy_calls.tsv"
))
```

## Session Info

```{r}
sessionInfo()
```
